{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "UZZqPBS0UppM",
    "outputId": "281abfa6-b935-4d99-a267-e237900aa5dc"
   },
   "source": [
    "!pip install numpy pandas requests requests_cache SetSimilaritySearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "uG7JfetZjcXH"
   },
   "outputs": [],
   "source": [
    "from copy import copy,deepcopy\n",
    "from itertools import combinations\n",
    "import json\n",
    "import textwrap\n",
    "\n",
    "\n",
    "from fisher import pvalue_npy\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "rzoUUSDeVz0a"
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import requests_cache\n",
    "\n",
    "\n",
    "requests_cache.install_cache(\n",
    "    \"pfocr_cache\",\n",
    "    allowable_methods=['GET', 'POST'],\n",
    "    stale_if_error=False,\n",
    ")\n",
    "#requests_cache.remove_expired_responses()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "j9DiSC9NU_Fa"
   },
   "source": [
    "# BTE: Cluster TRAPI Results by Selected PFOCR Figures\n",
    "\n",
    "From discussion with Andrew, we want to first identify a set of figures that can serve to cluster the TRAPI results. We can call these the \"cluster figures\".\n",
    "\n",
    "**Identify Cluster Figures**\n",
    "\n",
    "We identify the top cluster figures by iteratively running Fisher's Exact Test to identify PFOCR figures that are most similar to all the CURIEs from the BTE TRAPI results. To get each cluster figure, we do the following:\n",
    "\n",
    "1. Run Fisher's Exact Test for the CURIEs from each figure vs. the set of the unique CURIEs found from all the TRAPI results, limiting each  set of CURIEs to those with categories found in both PFOCR and BTE TRAPI, e.g., \"disease\" is found in PFOCR and \"biolink:Disease\" is found in BTE TRAPI\n",
    "2. Select the figure with the lowest p-value as a cluster figure\n",
    "3. Exclude that figure and the CURIEs in it\n",
    "4. Repeat to get the next cluster figure, continuing until:\n",
    "    - no more figures match the remaining BTE TRAPI results\n",
    "    - we've found some set number of figures (20)\n",
    "    - we run out of BTE TRAPI results\n",
    "    - we run out of PFOCR figures\n",
    "\n",
    "**Cluster BTE TRAPI Results by Selected PFOCR Figures (Cluster Figures)**\n",
    "\n",
    "Once we have a set of cluster figures, we use them to cluster the BTE TRAPI results. We do this by running Fisher's Exact Test to identify which Cluster Figure each BTE TRAPI result is most similar to.\n",
    "\n",
    "**View TRAPI Results Clustered by PFOCR Figure**\n",
    "\n",
    "With BTE TRAPI results clustered by PFOCR figure, we display the top figures, each with their most closely related BTE TRAPI results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TxW41muraAFL"
   },
   "source": [
    "## Get BTE TRAPI Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "N71NYHWeVCj9",
    "outputId": "d8c2c9ae-9750-455c-ff61-b357b4ff0b2d",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "200\n",
      "got response from cache\n"
     ]
    }
   ],
   "source": [
    "query = {\n",
    "    \"message\": {\n",
    "        \"query_graph\": {\n",
    "            \"edges\": {\n",
    "                \"e00\": {\n",
    "                    \"subject\": \"n00\",\n",
    "                    \"object\": \"n01\",\n",
    "                    \"predicates\": [\n",
    "                        \"biolink:genetic_association\"\n",
    "                    ]\n",
    "                }\n",
    "            },\n",
    "            \"nodes\": {\n",
    "                \"n00\": {\n",
    "                    \"categories\": [\n",
    "                        \"biolink:Gene\"\n",
    "                    ]\n",
    "                },\n",
    "                \"n01\": {\n",
    "                    \"categories\": [\n",
    "                        \"biolink:Disease\"\n",
    "                    ],\n",
    "                    \"ids\": [\n",
    "                        \"MONDO:0019565\"\n",
    "                    ]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "trapi_response = requests.post(\"http://localhost:3000/v1/query\", json=query)\n",
    "#trapi_response = requests.post(\"https://api.bte.ncats.io/v1/query\", json=query)\n",
    "#trapi_response = requests.post(\"https://arax.ncats.io/api/rtxkg2/v1.2/query\", json=query)\n",
    "#trapi_response = requests.post(\"https://arax.ncats.io/api/arax/v1.2/query\", json=query)\n",
    "print(trapi_response.status_code)\n",
    "if trapi_response.status_code != 200:\n",
    "    raise Exception(trapi_response.text)\n",
    "if trapi_response.from_cache:\n",
    "    print(f\"got response from cache\")\n",
    "\n",
    "trapi_message = trapi_response.json()[\"message\"]\n",
    "\n",
    "if len(trapi_message[\"results\"]) == 0:\n",
    "    raise Exception(f'No results returned')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "print(json.dumps(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyze this GraphML file in Cytoscape: ./bte_sleeve_query_template.graphml\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{('n00', 'n01'): Text(0.0, 0.0, 'biolink:genetic_association')}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcwAAAEuCAYAAAADVyIPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAx00lEQVR4nO3deZzNZf/H8deZGTPMYsxYG2SZLGMZ46bbHiVSUQ9kSYul6EeKG1lua6gQrbe7u1Q3oUSoUNabslOyZYYsaRhCY8nIGDPf3x+Xc2aGwQxz5izzfj4e8+g43+06p/Sez/W9vtdlsyzLQkRERG7Ix9UNEBER8QQKTBERkWxQYIqIiGSDAlNERCQbFJgiIiLZoMAUERHJBgWmiIjXWgcEATYv+AkC1ufu15NDNj2HKSLijX4C7gHOu7ohuSgYWAvEuOTqCkwREa+zH6gLnHV1Q5ygCLAVuCvPr6zAFBHxKpcxFVgskObapjiFD1AN2A745vmVRUTEa0wCfsU7wxLM5zqE+Zx5S4EpIuI1YoHxQJLTr7RqFVStCoGBcO+9cPhw+ra5c6FhQ7OtWTNnXD0JGIf5vHlHgSki4hUuAx2Ai06/0qlT0K4djBsHiYlQty506pS+PTwc+veHoUOd2YqLQEcg1ZkXyUSBKSLiFSZhuipzb1hK+fIweTJER0NoqAnFixdhwQKoXh06dICCBWHMGNixA+LizHH33w8dO0JERK41JQsWed01q8AUEfF4BzBdsRdy/cxz58LSpXDoEOzcCdOnw88/Q61a6fsEBUFkpHk/b9m7Zg/kydUUmCIiHm8ykOKUM7/4oqkUw8OhTRvYvh3OnzcVZ0ahofDnn05pwk2kAFPy5EoKTBERj5YEfIK5h5n7SpVKfx0YaMIyOBjOncu837lzEBLilCbcxGVgBnkx0EmBKSLi0T7FTB2Xd6pXN/cs7ZKS4MAB875r2IA5Tr+KAlNExGNZwETyorrKqG1b2L0b5s83g4DGjjUDg6pWNdtTU837ly9DWpp5neKcHuMrkoAJ5OaAp6woMEVEPNYm4HieX7V4cROWw4dDWBhs3gxzMhR4M2dCoULQuzesXWte9+zp7FYdAzY79QqaGk9ExGO1Bxbi7MrKM/gAbYEvnHYFBaaIiEdKAwKBZFc3xI0EYB6tcU7nqbpkRUQ80i+An6sb4Wb8MN+LcygwRUQ80g/of+FX8wF+dOrZRUTE42zEuxaHzg3nMd+LcygwRUQ80jo02OdqFuZ7cQ4FpoiIC5QvXx5/f39OnTqV6f3atWtjs9n49ddfAdiwYQP33XcfISEhhIaG0qZNG/bs2Q2Ymc7XrAGbDfr0yXz+xo3NvK9g/unra2boCQ6GChWge3fYty/zMcnJMGwY3HmneRSkUiV4/XW40dDQX381y3sFBprnMFeuzLz9zTfNbEGFC0OPHuYadiNHQs2a4OdnJnDPyLLglVdMWwoXhs6dM88u1K0b+Punf6bgYPP8p/le0rhw4QJ9+vShWLFihIaGcs899ziOHTNmDAUKFCA4ONjxc/Dgwet/yCsUmCIiLlKhQgU+++wzx5937drFhQvpE6hv3LiRli1b8uijj5KQkMChQ4eoVasWjRo14uBBX8d+QUHm2ccrGZulBg3MtHZnz5pQK1QI6tQxExDYdehg1rn85hszL+zMmfDBB9Cv3/XP+/jjULs2/PGHCbjHHoOTJ822ZctgwgRzzsOH4eBBGD06/di77oJJk+Dhh6897yefmOuvXw8JCfDXX/DCC5n3GTzYfCb7j68vgC/wC7169SIxMZHY2FgSExN58803Mx3bqVMnzp8/7/ipWLHi9T/kFQpMEREXeeqpp/jkk08cf54xYwZPP/2048+DBw/m6aefpl+/foSEhBAeHs748eOpX78KY8akT51TpIipuF5++ebX9PU1K4v8+9/QtGl6ZbdqFSxfbiYkqFHDVH3168OsWTB1Kuzff+259u2DbdvMdQsVgvbtTcU4f77988Azz5gp88LCTEVpr3oBunaFBx/Meg7aRYvMsWXLmupxyBD4/HO4cNMFWfyIi9vE119/zQcffEDx4sXx9fWlTp06N/9ybkKBKSLiIvXr1+fcuXPExsaSmprKnDlzePLJJwG4cOECGzZsoEOHDtcc17FjQ1asyDzZ+vDhJqj27s3+9du1MzPxAKxYAfXqmYDKqF49KFPGBCpA69amagSznFfFipkDr1at9GW+rl4GrFYt+P13U41mR8auYMsy3bm/ZHhq5N//Nquo1KmTHtIAW7bspFy5cowePZpixYpRs2ZN5mfcAVi0aBHh4eFUr16d9957L1vtUWCKiLiQvcpcsWIFUVFRlC5dGoDExETS0tK44447rjnmjjuCOXUq843FUqXg//4PRo3K/rUjIiAx0bw+dQqyuNSV65ntAIsXw9Ch5vXNlvm6erv9dXaWAWvVCj780HQznz0LEyea9+0V5osvmvA8cQLGjTMV9vr1ABZHjhxn9+7dhIaGkpCQwL/+9S+6du1KbGwsAB07diQ2NpaTJ08ybdo0xo4dm6lr/HoUmCIiLvTUU0/x6aefMn369EzdsWFhYfj4+HDs2LFrjjl27CTFil27QsmQIea+YcaVRG7k6FFToQEUKwZZXOrK9cz2q91sma+rt9tfZ2cZsB49zP3RZs1Ml+6995r3y5Qx//zb36BoUdN1/NBD8MQTsGABgEWhQj4UKFCAESNG4O/vT9OmTbn33ntZvnw5ANWqVSMiIgJfX18aNmxIv379+OKLm0+pp8AUEXGhcuXKUaFCBb755hvatWvneD8oKIgGDRowb968a46ZO3cLzZtfO8tP0aLQv7+5V5gdCxdCkybm9f33m0nU4+Mz72N/7777rj2+enUzkCdjxbhjR/oyX1cvA7ZjB5Qsadp5Mz4+5t7or7/CkSPmXKVLm5+s2Gz2Llwb0dGRWWy//hJoNpuNbM0Sa4mISJ4rV66ctWLFCsuyLGv//v3W1q1bLcuyrJSUFAuwDh06ZK1du9YKDAy03n77bevcuXNWYmKiNXz4cCs0NMjaty/Ysiys1auxSpfGsizzc/YsVtGiWOHhWP/9r3nvv//FatTIvL58GevgQay+fbGCgrB27kw/9qGHsO6+G2v3brPfxo1Yd92F1adP+j5X/9SrhzVwINZff2EtWIAVGop14oTZ9u23WCVLYv38M9bp01j33os1ZEj6sZcumeMefxxr+HDz+vJls+2PP7D278dKSzPHV6+O9f776cfOm4f1559YqalYy5ZhBQeb78KyQq1Ll76yIiMjrbFjx1opKSnWunXrrODgYCs2NtayLMv68ssvrcTERCstLc3avHmzFRERYU2fPv2m/84UmCIiLpAxMDPKGJiWZVlr1661mjZtagUFBVkhISHWQw89ZO3a9aFlWaFWVoFpWVgTJ2JB5sD08TEBGRiIdeedWE8/jbVnT+bj/voLa/BgrDJlsAoWxIqMxHrtNRNK9n1atcJ65ZX0Px86hNW0qdm/cmWsFSsyn3PKFKwSJbBCQrC6dcO6eDF9W9eupp0Zf+xt3rvXnK9QIdPeKVMyn7dxY6zChc15o6OxPvvMvi3Usqz/Wbt377bq169vBQYGWlFRUdaCBQsc33Hnzp2t8PBwKygoyKpSpYr19ttvZ+vfmVYrERHxOHuBumhqvKwEYeaTrZLrZ1Zgioh4HC3tdX0FgSScMURHg35ERDyOD1DV1Y1wU1XRepgiIpJBY+D6Iz/zJxvQxGlnV2CKiHikBkCwqxvhZoKB+k47uwJTRMSNlC9fnpVXL/kBrF27lipVMg5kqYu5l3mtbt1gxAj7cVAlm+Nf1qxJnxggtwQHm2c180YacPtzxl6PAlNExAM0adKEvZkmiq0EXL7e7hmOy9n8stm1Zo2ZXMC+tFaZMtCxI2zdmnm/8+fNfLN54zLme3EOBaaIiEfyAR7Glf8bj4gwgfjnn7Bpk1kPs0mT9Ina85YP0Bpnfh8KTBERN7N161aqVatGWFgY3bt35+LFi6xZs4YyGfpLY2NjadbsEEWKpFG9Onz9ddbnurqbtXx5mDwZoqPNZOidOsHFi1kf+847UK2amZruRmw2c42xY+HZZ82cthm32ZcG++Ybc76QEDPF3eTJ6fstXgwxMWapsoYNYefO9G0TJpglyUJCzPELF6Zv27/fLFMWGppGsWIr6dSpk2NbXFwcLVq0IDw8nCpVqjB37twbf5CbUGCKiLiZ2bNns2zZMg4cOMC+ffsYP358pu0pKSm0adOGli3bc+JEJO++ayYfz27X69y5sHQpHDpkginjGpV2Y8ea97/7Lj1wixSBdetufO527cwamUlJ12575hl4/31Tke7enT4/7U8/mcnW33/fLP313HPwyCNmOS8wYbl2rVm1ZPRoePLJ9IniR46Eli3h9Om7OHLkGC9cWWU6KSmJFi1a0KVLF06cOMGcOXPo06cPe/bsyd6XlAUFpoiIm+nbty9ly5YlPDyc4cOHX7P01KZNmzh//jxDhw7D338I990XTOvWkI0VqgCzNFZEhFmppE0b2L49fZtlwYABZjHp1auhePH0bWfOQOPGNz53RIQ5x5kz124rUAD27DGrloSFmRVHAD74wIRkvXpmgeuuXSEgwHTzAnToYM7r42Mq4kqVYMuW9HMePuxHQkIvChYsROMrDVy8eDHly5ene/fu+Pn5Ubt2bdq3b5/lZPbZpcAUEXEzZTOs4lyuXDkSEhIybU9ISKBs2bL4+PgAXYA0ypUzy3VlR6lS6a8DA819SLszZ0yADRt27VqX2XH0qOmGLVLk2m3z55tu2XLlTDfqxo3m/cOHYcoUc4z9Jz4e7B/7k0/Su2uLFDHVqX19zkmTwLIs/v73N6hevToff/zxlXMeZvPmzRQpUsTxM3v2bI4fP57zD3XFtevDiIiIS8VnWGPrt99+IyIiItP2iIgI4uPjSUtLw8cnCHia3357n8qVb3+m07AwmDXLjHhduBAaNcrZ8QsXmsoxKOjabXffDV99BSkp8K9/mWvEx0PZsjB8uPm52uHD0LOnGUjUoIGpQGNi7Et5QalSfkyb1hP4N+vWreP+++/nnnvuoWzZsjRt2pQVK1bk9Cu4LlWYIiJuZurUqRw5coTExEReeeWVTANZAOrVq0dgYCCTJk0iJSWFNWsas2iRRefOuXP9Zs1g9mxzP9Le9XkjlmUqy5dfhg8/hFdfvXafS5fMOc+eNd2ohQubLlYwgfif/5i1Ny3L3P9cssTc60xKMhWrvWv4v/81FabdvHk+HDnSBTCLbttsNnx8fGjdujX79u1j5syZpKSkkJKSwtatW4mNjb3l70WBKSLiZrp06ULLli2pWLEikZGRjLDPQnCFv78/ixYt4ttvv6VYsWI8++xonniiWK5OOtCiBXz8sbnHuW2beS842Ay+sUtISH8O8+67YdcuMyq3ZcuszzlzphmlW7iwCcjZs837devCtGnQt6+pcO+6K30gUrVqMHCgqS5LljTXSK96A9m6tR716nUiODiYRx55hLfffpuKFSsSEhLC8uXLmTNnDhEREZQqVYohQ4aQnHzrE9ZrtRIREQ+VmprKW2+9xYgRI7h06SLbt5uA8fV1dcvygg9QDdgO5M0HVmCKiHigvXv30qNHDzZs2OB4r0YNX7Zts1GgwM1nAPJ8hYBt5OWqLeqSFRHxIKmpqUyZMoWYmJhMYRkTE8OsWT9SoMDLmEWUvVkQMJK8XuJMFaaIiIfIqqr08/Nj5MiRDBs2jAIFCgCpQAywh+tNzu7Z8r4r1k6BKSLi5jLeq7yYYR67mJgYpk+fTq1ata46Yj9wN3AmD1uZV4oAW4G78vzK6pIVEXFje/fu5Z577mHQoEGOsPTz8+Pll19my5YtWYQlmDD5H963XmYwsBpXhCWowhQRcUs5ryqzsh5oCVxwVjPzUCCwHMjhTAq5SIEpIuJmsnevMrvWAQ8C52+2oxsLBpbiyrAEBaaIiNvInaoyK9uBe4FzeNZAIB+gMKYbNsa1TUGBKSLiFnK3qszKfqAtcAjIYu0ttxMEVAAW4qp7llfToB8RERe60XOVP/zwA6NGjcqFsAQTOtuB4ZiH/m25cE5nsGHaNxzTXvcIS1CFKSLiMs6vKq8nFuiI+1Wb9qpyLhDl4rZcSxWmiEgey7uq8nqiyFxtunqlRz8yV5XuF5agClNEJE+5rqq8ngPAFGAGpjs0LyvOIMACugIDgcg8vHbOKTBFRPKA80bA5pYk4DNgAnAc+AvnjKj1wVSTdwBDgc54yty3CkwRESdzv6ryRixgMzAZWIzpLvXBPMd5K3FhwzxHmQZcBloDg4B6uO/Ao6wpMEVEnMT9q8qbSQN+AX4ENmImQYjDTHpuv+9pXfmxkR6AlzGTwFcFGgMNgDpAJTx56IwCU0TECTyrqswJe4gmYLptLwLJQABQENPdGoGnh2NWFJgiIrnI86tKuR5XjyUWEfEa3ltVCigwRURum6rK/EGBKSJyG1RV5h8KTBGRW6CqMv9RYIqI5JCqyvxJgSkikk2qKvM3BaaISDaoqhQFpojIDaiqFDsFpojIdaiqlIwUmCIiV1FVKVlRYIqIZKCqUq5HgSkigqpKuTkFpojke6oqJTsUmCKSb6mqlJxQYIpIvqSqUnJKgSki+YqqSrlVCkwRyTdUVcrtUGCKiNdTVSm5QYEpIl5NVaXkFgWmiHglVZWS2xSYIuJ1VFWKMygwRcRrqKoUZ1JgiohXUFUpzqbAFBGPpqpS8ooCU0Q8lqpKyUsKTBHxOKoqxRUUmCLiUVRViqsoMEXEI6iqFFdTYIqI21NVKe5AgSkibktVpbgTBaaIuCVVleJuFJgi4lZUVYq7UmCKiNtQVSnuTIEpIi6nqlI8gQJTRFxKVaV4CgWmiLiEqkrxNApMEclzqirFEykwRSTPqKoUT6bAFJE8oapSPJ0CU0ScSlWleAubZVmWqxshIt4pLS2NdevW0bRpU8d7qirFU/m4ugEi4r18fHyoU6cObdu2BUxV+cMPPzBq1CiFpXgcVZgicltSU1Px9fW94T5nzpxh2rRp9O/fX0EpHkuBKSK3zLIsbDYbAO+++y6RkZHUrl2bO+64g7S0NHx8TCdWxtcinkqBKSK35cSJE4wcOZLY2Fiio6OJi4tj5cqVrm6WSK7Tr3wikm2WZZHxd+wff/yR9u3bExwczPfff8+//vUv/P39GTFiBGAqSxFvocAUkWxJTU3FZrNhs9k4efIkAHXq1KFq1aocPXqUM2fOADBlyhTmzZvHli1b8PHxQZ1Y4i0UmCJyQxcuXABwDOx55ZVXePTRR+nXrx/z5s1j4sSJHDt2jK1bt5KamkpUVBRPPvkkAwcOBHDc4xTxdLqHKSLXtWjRIi5cuEDHjh2x2Wx89NFHfPHFF7zzzjvs3LmTSZMm8e677/Lbb78xZ84cXn/9dSpUqADA6dOnCQsLc/EnEMk9qjBF5Br236MjIyPp1KkThw8fBmD37t106dKFSpUq0b59ewYOHMirr77KY489xsWLF5kzZw6XL18GICwsjNTUVJd9BpHcpsAUkUwOHTrEoUOHAChXrhyfffYZr732GidOnCA6Opp3333XsW/lypUpWrQoAFOnTuW5557Dzy99xs2bPZ8p4kk0l6yIOCQnJ7Nt2zb27dvH8ePHueOOO2jatCmFCxfms88+47nnnmPhwoX069ePt99+m6+++oo///yTlJQU7rzzTmw2m565FK+le5gikinkvvnmGx5//HGqVKnCwoULKV26NAsWLGD16tU8+uijREdH06tXL8cjIx988AGlSpVyZfNF8oQqTJF8zP5cZcaKsH79+jz//POkpqYSHx9P6dKlefjhhzl06BDLli2jYsWKfPnll5w8eZLixYsDmslH8gf9Fy6Sj9lsNnx8fNixYwcPPfQQEydOZOPGjbz66qsUKlSIJUuWcOTIEQICAmjevDnnz59nx44dAI6wTE1NVVhKvqAuWZF85ujRo8yZM4fnnnuO4OBg5s+fz7hx4xg/fjzx8fF88sknTJ06ldDQUN58802qVq3K2bNniY6Oplq1akRGRrr6I4i4hAJTJJ/55ZdfKFSoEEWLFqVQoUJs2rSJypUr8+OPPzJgwADq1q3L77//zldffcXmzZuZPn06hw8f5tNPP3VUlRknXRfJLxSYIvmAZVmkpaU5HvM4evQow4YN4/HHH+fBBx9k3759PP/880yZMoUSJUrwwAMP0KpVKyZOnMjFixcpWLCg4zwKSsmvdONBJB+w2Wz4+vqya9cuFi1aRLFixahVqxZLly7lxIkTHDlyhPDwcKKjozl8+DC1a9fGsiwuXbqEv78/kD6XrEh+pcAUySfGjRvHU089lWkQT4ECBZg3bx4NGzbk4MGDPPHEE7Rr145OnToxadIk/P39HQN6NAmB5Hd6rETEC9kf87B3ocbHx7N9+3ZWr17tmN81JiaGw4cPs3jxYnbv3s2aNWvYuHEjr7/+OhEREZnOIyKqMEW8TsaQS05OBuC3335j586djrC8dOkSAI0bN+bOO+9k5syZ+Pr6cv/99xMREeGYA1ZhKZJOg35EvNC2bdt47bXXCAwMpH379tSoUYOpU6fyt7/9jSeeeAKAjRs3UqFCBXx9fSlYsCAhISEubrWIe1OXrIiHs//Oax+Qc/jwYXr37s2gQYNITExk+fLlbNiwgYYNG/LWW2+RlpZGQkICM2fO5M0336RFixaAul9FbkaBKeLBUlNTHYNxTp8+TZEiRTh48CBBQUF06NABgO+//54FCxYQFRXF0KFD2bRpE7/++ivLly933KsEdb+K3Iz+hoh4MHtYjho1iq5du3L06FGioqLw9/dn7dq1AFSvXp3du3eTlpbGww8/zLhx45g5c2ame5UicnMKTBEPYw85+3OSnTt35vDhw8yYMYMyZcoQGBjIQw89xH/+8x/Onj3rWFXE/k+7jBMZiMjNqUtWxINkDLmTJ09SokQJKlSoQM2aNYmLiyM2NpaAgADuvvtujh8/zuOPP87+/fsZPHgw0dHRmc6lLliRnFFgingQ+8oivXv3pnnz5gwdOpTIyEjeeecdGjRoQFJSEpcuXSI6OppXX32Vo0ePEhgY6HicRFPbidw6PVYi4sauDrh9+/bx2GOPMWrUKB577DHH++fOnaNw4cIA9OvXj2rVqvHcc885tmsErMjt098gETeV1dytCQkJtGjRgqCgIObMmUP//v1ZsmQJycnJbNu2jXvvvZdjx47Rrl27TMcpLEVunypMETc3adIkfHx8aNq0KTVr1uSll14iPj6e++67j7i4OE6cOMH48ePZtWsXycnJPPnkk4CqSpHcpsAUcVNHjhzhySefJCoqioYNGzJjxgxGjx5NkyZNHPscP36cTp068eGHH1KpUiXH+wpLkdynwBRxU+vXr+fChQvcf//9dOzYkWPHjhEQEMDs2bMJCwvjlVde4YsvvmDQoEH06NHD1c0V8Xr6FVTETd199900aNCARx55hBo1arBs2TIKFizIpEmTCAgIoHjx4qxatcoRlvrdV8S59FiJiItcPQfs1fz9/YmPjyc4OJjRo0cDEBoayk8//cTRo0d54YUXgPTp8fS4iIhzqcIUcYG0tDRsNhs2m42jR49ed78KFSrw888/M2zYMO6++27Kly/PokWLKF26tGMfzdYjkjd0D1PEhcaPH8/KlSt59913qVmzZqZt9spx7969bNmyhUuXLvHMM88AGtQj4goKTJE88tprr9G9e3dKlSqFZVm8/vrrbNy4kalTp2ZaNeRmFJYirqG/dSJOZv+dtGbNmpQqVcrx3saNG+nTpw8RERFcunQpy2MzriZif62wFHEN/c0TcaKDBw86BuO0bt2avn37MmHCBHx8fKhRowY//fQTqamp+Pv7A2YmHzDhaFmW4/7kG2+8wYoVK1zzIUQEUGCKOM1ff/3F+PHj+e6775g4cSLz58+nc+fOTJ8+nfj4eGrXrs3vv//OtGnTAJg4cSIDBw7k7NmzjlGvmzdvplmzZgQEBPDAAw+4+BOJ5G96rEQkl9lHwBYqVIioqChatGjBI488Qvfu3SlRogQPPPAAQ4YM4aOPPqJgwYKMHTuWL7/8Eh8fHz7++GNCQ0MBGDNmDN999x2zZs2iTJkyLv5UIqJBPyK56OoBOV9//bVjBOwbb7zheL9OnTr07t2bZ599lsTERE6fPk1kZKRj+65duzhw4ACPPvqonq8UcRMKTJFctn37dsaOHUuzZs0cFWblypV54403aN26NQDff/89rVu3Zt++fY6BQACXL1/Gz08dPyLuSPcwRXLRhg0b6NGjB7169SI5OZmhQ4dy8uRJRowYwahRozh79iwTJ06kevXqrF27NlNYAgpLETemClPkFtknFsho1qxZBAYGEhERQa9evfi///s/+vTpA8CAAQOIjY0lOTmZefPmUbRoUeDaRaJFxD3p11mRW5DxkY8lS5ZQtmxZoqOjCQgIoGvXrjRq1IiFCxcSGRnJH3/8QWJiIpMnT+bo0aOULVs207kUliKeQYEpcgtsNhu//vorzz//PBcuXCA+Pp6PPvqIqKgoOnXqRM2aNYmMjGTnzp307duXnj17UqlSJUdYZlWdioh7U5esSDZc3W26detWx8CegQMH8s4777B9+3b69u1LcnIyzzzzDHXr1uWHH36gf//+9OrVy4WtF5HcoApT5CYyPipy8eJFChYsSEBAAJcuXWLfvn0AvPjii/To0YNvv/2WwYMHs27dOn7//XfCw8MpWbLkNecREc+jv70iN2EPuREjRvDMM8/wz3/+kypVqtCnTx9sNhurVq0C4KWXXmLJkiUsXbqUsLAwoqKiKFmypOaAFfESqjBFsmCvBlNTU0lNTaVfv374+/szY8YMSpcujb+/P927d2f//v0sXryY6OhooqKiGDFiBA0bNszUfat7lSLeQb/yimRgWRaWZTmqQR8fH/z9/YmMjKRz58784x//oFatWnTv3p1y5crRvHlzzp49y6effgrAQw89RJEiRdDQABHvo0E/IldkHNhz7Ngx3njjDYoXL07//v1p27YtBw8e5LnnnqN///4ArFq1iubNm7Nx40aqVavmmANWRLyTAlPkKv/85z/ZsmUL/v7+JCUlMWPGDL7//ntmzpzJrFmzKFmyJBMmTGDu3LksWLCA8uXLA5qAQMTb6R6mSAZxcXFs27aNZcuWsW/fPj7//HPGjBnD9OnT2bNnDy+++CLHjx8nLCyMxYsXExER4ThWYSni3RSYki9drxrcv38/x48fx9fXl6ioKDp37ky/fv2YOXMmEyZM4Ny5c+zbt4+6desCmoBAJD/RoB/Jl+xhefHiRQC+++47EhMTefDBBylfvjyffPIJAFWrViUkJITPP/+cXbt2UbhwYYWlSD6lClPyje3bt7N37146deoEwIcffohlWfTs2ZMVK1ZQsmRJQkND6dmzJ4MGDaJ06dJs3boVX19fSpQowf79+6lZs6bjfApLkfxFFabkG4mJifTt25dDhw4BsHnzZipXrgzA+PHjqVq1Kr6+vjz88MMMHjyYhQsXEhcX53hkJCgoyGVtFxHX0yhZ8WoZ//O22WwMGDCAgwcPMn/+fFq1asXHH39M2bJlSUlJoUCBAgBcuHCBwMBAAE6ePMmQIUNISEhg+vTp16xfKSL5hwJTvFbGuVtPnz5NWFgYALVq1eLvf/87vr6+dO7cmWbNmjmOefPNN/H39+f5558nJSWF999/3zHTj4jkbwpM8UqXL1/Gz8/coh8/fjzLli2jbdu2DBgwgB07dnDffffRsGFDwsLCOH/+PEWKFKFly5Z07Ngx05yvly5dwt/f31UfQ0TciO5hilew/943YcIEAPz8/Dh27Bh9+vQhMTGRMWPGsHLlSqZOnUqtWrX4xz/+4ZgbdvLkyTRp0oQWLVo4wjItLQ1AYSkiDholK14hLS0NX19fwsLCHM9Yrlq1ilmzZrF69Wrq1KnD+fPnWbx4MatWrWLEiBGUK1eO//znP/Tu3ZuKFSsC6c9namUREbmaumTF43366ad89dVXfP7556SlpdGmTRuGDx9Ow4YN6dq1KwULFuT9998H4LXXXiMhIYERI0Zw9uxZwsLCKF68uIs/gYh4Av0aLR6vTZs2bNu2jQULFuDj40P9+vV59tlnARg8eDCnTp3is88+A3A8g3nq1CkqV65M8eLFHd2vIiI3ogpTPJJ9lh37SNi5c+cyZswY1q9fT1hYGI0aNaJBgwZMnjyZ2bNnM2PGDKZNm0a5cuVITk4mICDA1R9BRDyMKkzxOPb7lQA7duwgJSWFjh070qBBA1544QUA5s6dy8cff8yPP/7IE088QUxMDAkJCQAEBARovUoRyTFVmOIxMk4uEBsbS48ePQgJCSE0NJRevXrRtGlTateuzciRI+ncuTMTJ05k0qRJ/PHHH5r3VURumypMcXvJycn07duXgwcPAqY7dtq0afTs2ZPly5fTrFkz5s2bx549e3jzzTcZN24cf/zxB0OGDGH69OmA5n0VkdunwBS3lpKSQkBAAKNGjaJkyZKOpbc2btxIaGgoAB06dCAyMpI1a9bQsmVLKlSo4AjKNm3auLD1IuJN9BymuKXff/+dkiVLOrpgS5QoQbdu3ShatCivvvoq3bp1Y8OGDbRs2ZISJUpQoEABjhw5AsCXX37pmOVHRCS3qMIUt7N3714aNWpEUlIS8fHxtGrVivHjx1OrVi12797Nhg0bqFGjBpcvX6Z379788ssvfPvtt47JB+xrXaamprryY4iIl9GgH3FLAwcOZNeuXZQoUYLmzZtz8uRJfv/9d5YvX06TJk34xz/+QWhoKGPHjuXEiRM0btyYF1980dXNFhEvpsAUt5SYmMijjz5KiRIlmD9/PklJSezatYupU6dy9OhRGjVqxOjRo/Hz88s0ejbjCiUiIrlJ/2cRtxQeHs5LL73E2rVrOXnyJEFBQRQvXpyQkBAGDBjApk2bHM9VFihQwPFcpcJSRJxFFaa4jQsXLlCgQAFHtQjQv39/4uLiWLp0KbGxsXTp0oX//e9/BAUFaSUREclTCkxxudTUVN566y3ee+89tm/fTnBwsGPbmTNnqFWrFqVKlaJ06dK0atWKXr16Aekri4iI5AUFprjU3r176d69Oxs3bgSgR48evP3225lCc8mSJSxZsoTx48cTHh7uqqaKSD6nwBSXsFeVI0aM4OLFi473Y2JiWLt2babAzEiDekTEVfR/Hslze/fupUmTJgwaNMgRln5+frz88sts2bJFYSkibkkVpuSZG1WV06dPp1atWi5snYjIjWn+MMkTV9+rBFNVjhw5kmHDhmUaGSsi4o4UmOJUqipFxFsoMMVpVFWKiDdRYEquU1UpIt5IgSm5SlWliHgrBabkClWVIuLtFJhy21RVikh+oMCUW6aqUkTyEwWm3BJVlSKS3ygwJUdUVYpIfqXAlGxTVSki+ZkCU25KVaWIiAJTbkJVpYiIocCULKmqFBHJTIEp11BVKSJyLQWmOKiqFBG5PgWmAKoqRURuRoGZz6mqFBHJHgVmPqaqUkQk+xSY+ZCqShGRnFNg5jOqKkVEbo0CM59QVSkicnsUmPmAqkoRkdunwPRiqipFRHKPAtNLqaoUEcldCkwvo6pSRMQ5FJheRFWliIjzKDC9gKpKERHnU2B6OFWVIiJ5Q4HpoVRViojkLQWmB1JVKSKS9xSYHkRVpYiI6ygwPYSqShER11JgujlVlSIi7kGB6cZUVYqIuA8FphtSVSki4n4UmG5GVaWIiHtSYLoJVZUiIu5NgekGVFWKiLg/BaYLqaoUEfEcHhCYacAvwFHgL+AicAnwBwoChYDSQCXAx0VtzDlVlSIinsXNAtMejj8AG4F1QBxQAPC9so915cd25QcgFUgBqgKNgQZAXdwxRFVVioh4JptlWZZrm2ABm4DJwBJMhvsA569syykbEIwJ38vAw8AgoD7pAesaqipFRDyXCwMzCZgNTAKOY7pb05xwHR9Mt+0dwBDgcSDICde5PlWVIiKezwWBeQBTTX5CeiWZV+yV59OYqjPS6VdUVSki4h3yMDBTgYnAeMz9xst5c9ks+WHui47AVJ2+N979FqiqFBHxLnkUmLFAR+AQpivWXQQBFYC5QFSunVVVpYiI93HyENJU4FWgDvAz7hWWYNrzM6Z9r2Lae+tSU1OZMmUKMTExmcIyJiaGH374gVGjRiksRUQ8lBMrzP1AW9yvqrwee7W5ELgrx0erqhQR8W5OCsyfgPuAczhn5Kuz+ACFgdVATLaO0L1KEZH8wQmBuQ54kLwd/ZrbgoGlQKMb7qWqUkQk/8jlwFwHPABcyL1TukwgsJysQlNVpYhI/pOLgfkTcA+eXVleLRhYS8buWVWVIiL5Uy4F5n7M3K1nb/9UbqcIsJXU1AqqKkVE8rFcCMzLmAosFs8a4JNdPiQnR9K8eVHWr9/keFdVpYhI/pILq5VMAn7FO8MSII2UlF9o3PgX1q8376iqFBHJf26zwozFPPT/V261x21duAD16vnSocMoVZUiIvnQbQSmvSt2D7e2DJdnSU2FS5ciKVRoL86Ye1ZERNzbbUyNNwkzi4/3hyWAry8UKnQc87lFRCS/ucUK8wBQk/zQFXutQsAu8mJpMBERcR+3WGFOxizRlR+lAFNc3QgREcljt1BhJgEl8I7ZfG5VIHACM2G7iIjkB7dQYX4K2HK9IZ7FBsxxdSNERCQP5TAwLWAirliua9UqqFoVAgPh3nvh8OH0bcnJ0KMHFC4MpUrBG284uzVJwATyy4AnERHJcWBuAo47pSE3cuoUtGsH48ZBYiLUrQudOqVvHzMGfvnFhOjq1TBpEixd6uxWHQM2O/siIiLiJnIYmJNx5r3L8uVh8mSIjobQUBOKFy/CggVQvTp06AAFC5qA3LED4uLMcTNmwMiREBYGUVHQsydMn+60Zl7xF+b7EBGR/CAHgZkGLMHZ3ZBz55rq8NAh2LnTBN/PP0PGWeiCgiAy0rx/+jQcO5Z5e61aZptzpQGL8d4pAUVEJKMcBOYv5MrUszfx4osQEQHh4dCmDWzfDufPm4ozo9BQ+PNPs83+56u3OZ8f5nsRERFvl4PA/CFnu9+iUqXSXwcGmkAMDoZz5zLvd+4chISYbfY/X73N+XyAH/PiQiIi4mI5SMCNuGpx6OrVzT1Lu6QkOHDAvB8WBnfckXn7jh1mm/Odx3wvIiLi7XIQmOtw1WMUbdvC7t0wf74ZBDR2rBkYVLWq2f700zB+vLmfGRcH06ZBt2550TIL872IiIi3y2ZgpgFxTm3IjRQvbsJy+HBTUW7eDHMyzBvw8stmEFC5ctC0Kbz0ErRqlVeti0MDf0REvF82p8bbC9TFVV2y7i0Icx+ziqsbIiIiTpTNCvMoWgPyevyABFc3QkREnCybgZkfl/HKCX0/IiLeLpuBeRHNm3o9Fub7ERERb5bNwLyEAvN6LCDZ1Y0QEREnu2lgli9fnpUr47h6Sa+1a6FKNse5dOsGI0bk/Lg1a6BMmezt6zo2IMDVjRARESfLZoXpz9WB2aQJ7N2b8wve6nHZ8eefMGCAmcQ9KAjuvBMee8w8huI8NqCgMy8gIiJuIAeB6d6Sk+G++2DXLli82EyPFxsLnTvDt986++qFnH0BERFxsWwF5tatx6hW7RxhYdC9u5lt5+ru0thYaNYMihQx09J9/XXW57r6uOst6ZWVd96BatXgyJFrt82cad7/8kuoUQN8fU2V+dhjZjkwu7g4aNHCTO5epYpZHcWuWzd4/nl4+GEzF229emYKvhsfexmIuM43JyIi3iJbgTl79nKWLSvAgQOwb5+Zhi6jlBSzskjLlnDiBLz7LjzxRPa7XrNa0utqY8ea97/7Lj1wixSBdVdmplu5Eh54wITk9SQlmcDr0sW0c84c6NMH9uxJ32fOHBg92kyzd9ddZnahGx+bAlTK3gcVERGPla3A7Nu3L2XLRhEebgLks88yb9+0yawqMnQo+PubrtHWra/d73qyWtLLzrLMfcnly2H1ajNNnt2ZM9C4sXl96lTmlU62bzeBWrhw+iCjxYtNRdu9O/j5Qe3a0L49zJuXflzbtvD3v5vtTzyR3pbrHxtOXqziIiIirpWtBS7Lli0LNAZ2Uq6cRcJVE9skJEDZsuCTITfKlYOjR7PXiKuX9Mp4/jNn4IMP4PPPr10TM6OiRc1C0nYxMebYlSvh2WfNe4cPmwFARYqk73f5Mjz11PXbYl9v8/rHFsvWZxQREc+WrdIoPj4eaAAE89tvphrMKCIC4uMhLcMc5L/9BqVL334Dw8JMdde9O6xff/39mjc3VWhS0vX3KVvWTM5+5kz6z/nz8N57N29H1seG8N57w3LycURExENlKzCnTp3KkSNlSExM5ZVXzMCcjOrVM9XYpEnmfuaaNbBokRmhmhuaNYPZs6FdO9iyJet9nn7arItpXwosNdUMHvrhh/R9Wrc292BnzjTtTEmBrVvNgKWbyfrYFGJji+TGRxQRETeXrcDs0qULLVv2pmLFC0RGpk9CYOfvbwLy22+hWDEzGOaTT9LXq8wNLVrAxx+be5zbtpn3goPNRAgABQuae5zVqplRrvZ7l1u3po+EDQkxVeicOaYqLlUKhgwxj6TcTNbHJpOcXOrmB4uIiMfL5vJedu2BL9H6j2B+12gLfOHqhoiISB7I4fDOQeghfbtCmO9DRETygxxWmBZQGdjvrPZ4kEqYhbVtN9tRRES8QA4rTBswGAh2Rls8SDAwBIWliEj+kcMKEyAJKAFccEZ7PEQgcPLKP0VEJD+4hSlqgoCnyeacB17ID+iKwlJEJH+5hQoT4ABQE/grt9vjAQoBu4BIVzdERETy0C1OghoJjMBUm/lJIDAShaWISP5zixUmQCoQA/yMGT3r7XyAasB2wNe1TRERkTx3G8ts+AJzgYK51RY3FwDMQ2EpIpI/3ea6VFHkj67ZIExXbC7O9SciIh7lNrpk7exds3vwzinz1BUrIiK5svKxL7AQKHz7p3JLhTGfT2EpIpKf5UJgAtwF/A/vmwEoGFiN+XwiIpKf5VJgAtQGluI9D/QHYj5PjIvbISIi7iAX7mGKiIh4v1ysMEVERLyXAlNERCQbFJgiIiLZoMAUERHJBgWmiIhINvw/+D3U05RuxNQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "G = nx.Graph()\n",
    "query_node_labels = {}\n",
    "for k, v in query['message']['query_graph']['nodes'].items():\n",
    "    node_ids = v.get('ids', [])\n",
    "    node_categories = v.get('categories', [])\n",
    "    G.add_node(k, categories=', '.join(node_categories))\n",
    "    query_node_labels[k] = '\\n'.join([k] + node_ids + node_categories)\n",
    "    #G.add_node(k, color='red')\n",
    "\n",
    "query_edge_labels = {}\n",
    "q_node_id_keys = set(['object', 'subject'])\n",
    "for q_edge_id,edge_v in query['message']['query_graph']['edges'].items():\n",
    "    for predicate in edge_v['predicates']:\n",
    "        G.add_edge(edge_v['subject'], edge_v['object'], predicate=predicate)\n",
    "    #G.add_edge(edge_v['subject'], edge_v['object'], predicates=', '.join(edge_v['predicates']))\n",
    "    #G.add_edge(edge_v['subject'], edge_v['object'], predicates=[1,2,3])\n",
    "    #G.add_edge(edge_v['subject'], edge_v['object'])\n",
    "    query_edge_labels[(edge_v['subject'], edge_v['object'])] = '\\n'.join(edge_v['predicates'])\n",
    "    \n",
    "query_template_f = './bte_sleeve_query_template.graphml'\n",
    "print(f'Analyze this GraphML file in Cytoscape: {query_template_f}')\n",
    "nx.write_graphml(G, query_template_f)\n",
    "\n",
    "options = {\n",
    "    'node_color': 'yellow',\n",
    "    'node_size': 5000,\n",
    "    'edge_color': 'black',\n",
    "    'width': 3,\n",
    "}\n",
    "query_template_layout = nx.spring_layout(G)\n",
    "nx.draw(G, pos=query_template_layout, **options)\n",
    "nx.draw_networkx_labels(G, query_template_layout, labels=query_node_labels, font_size=12,\n",
    "                        font_color='k', font_family='sans-serif',\n",
    "                        font_weight='normal', alpha=None, bbox=None, horizontalalignment='center',\n",
    "                        verticalalignment='center', ax=None, clip_on=False)\n",
    "nx.draw_networkx_edge_labels(G, pos=query_template_layout,\n",
    "                             edge_labels=query_edge_labels,\n",
    "                             label_pos=0.5, \n",
    "                             font_size=10, font_color='k', font_family='sans-serif',\n",
    "                             font_weight='normal', alpha=None, bbox=None, horizontalalignment='center',\n",
    "                             verticalalignment='center', ax=None, rotate=True, clip_on=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: we're including categories specified in the query template\n",
    "# in the next cell, and in a cell further down, we're also including\n",
    "# any categories from the TRAPI results.\n",
    "#\n",
    "# TODO: Some categories are supersets of others. Should we handle\n",
    "# this systematically?\n",
    "\n",
    "curie_categories = set()\n",
    "for v in query[\"message\"][\"query_graph\"][\"nodes\"].values():\n",
    "    if \"categories\" in v:\n",
    "        for category in v[\"categories\"]:\n",
    "            curie_categories.add(category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XRfyhW_Is73C",
    "outputId": "76ae9c51-0587-4f1e-91ac-f8d728ce6878"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all CURIE prefixes found:\n",
      "{'UMLS', 'SNOMEDCT', 'ICD9', 'UniProtKB', 'HGNC', 'PR', 'DOID', 'MEDDRA', 'OMIM', 'ENSEMBL', 'MESH', 'ICD10', 'ORPHANET', 'NCBIGene', 'MONDO'}\n",
      "\n",
      "unified CURIE prefixes found:\n",
      "{'NCBIGene', 'MESH'}\n",
      "\n",
      "failed to unify 21 CURIEs\n"
     ]
    }
   ],
   "source": [
    "# for genes/gene products, chemicals and diseases\n",
    "preferred_prefixes = set([\"NCBIGene\", \"MESH\"])\n",
    "trapi_results_unified_curies = set()\n",
    "unification_failed_curies = set()\n",
    "unified_prefixes = set()\n",
    "all_prefixes = set()\n",
    "curie_to_name = dict()\n",
    "curie_to_categories = dict()\n",
    "curie_to_unified_curie = dict()\n",
    "for k, v in trapi_message[\"knowledge_graph\"][\"nodes\"].items():\n",
    "    name = v[\"name\"]\n",
    "    categories = v[\"categories\"]\n",
    "    curie_categories |= set(categories)\n",
    "\n",
    "    for a in v[\"attributes\"]:\n",
    "        if a[\"attribute_type_id\"] == \"biolink:xref\":\n",
    "            curies = a[\"value\"]\n",
    "\n",
    "            # k should always be one of the curies\n",
    "            if not k in curies:\n",
    "                raise Exception(f\"key {k} not in {curies}\")\n",
    "\n",
    "            unified_curie = None\n",
    "            intersecting_trapi_results_unified_curies = trapi_results_unified_curies.intersection(set(curies))\n",
    "            if len(intersecting_trapi_results_unified_curies) > 1:\n",
    "                raise Exception(f\"matching multiple: {k} to {list(intersecting_trapi_results_unified_curies)}\")\n",
    "            elif len(intersecting_trapi_results_unified_curies) == 1:\n",
    "                unified_curie = list(intersecting_trapi_results_unified_curies)[0]\n",
    "            else:\n",
    "                # get curie for preferred prefix. usually this is k, but not always.\n",
    "                for curie in curies:\n",
    "                    [prefix, identifier] = curie.split(\":\")\n",
    "                    if prefix in preferred_prefixes:\n",
    "                        unified_curie = curie\n",
    "                        trapi_results_unified_curies.add(unified_curie)\n",
    "                        break\n",
    "\n",
    "            if not unified_curie:\n",
    "                if k in curie_to_unified_curie:\n",
    "                    unified_curie = curie_to_unified_curie[k]\n",
    "                else:\n",
    "                    unification_failed_curies.add(k)\n",
    "                    break\n",
    "\n",
    "                #raise Exception(f\"failed to find a unified curie for {k} in {curies}\")\n",
    "\n",
    "            [unified_prefix, unified_identifier] = unified_curie.split(\":\")\n",
    "            unified_prefixes.add(unified_prefix)\n",
    "\n",
    "            for curie in curies:\n",
    "                [prefix, identifier] = curie.split(\":\")\n",
    "                all_prefixes.add(prefix)\n",
    "                if not curie in curie_to_unified_curie:\n",
    "                    curie_to_unified_curie[curie] = unified_curie\n",
    "                if not curie in curie_to_name:\n",
    "                    curie_to_name[curie] = name\n",
    "                    curie_to_categories[curie] = categories\n",
    "                elif curie_to_name[curie] != name:\n",
    "                    print(f\"curie {curie} has multiple primary names: {curie_to_name[curie]} and {name}\")\n",
    "                    #raise Exception(f\"curie{curie} has multiple names: {curie_to_name[curie]} and {name}\")\n",
    "\n",
    "\n",
    "#print(f'curie_categories: {curie_categories}')\n",
    "#print(\"\")\n",
    "#print(f\"curie_to_name key count: {len(curie_to_name.keys())}\")\n",
    "#print(f\"curie_to_unified_curie key count: {len(curie_to_unified_curie.keys())}\")\n",
    "#print(\"\")\n",
    "print(\"all CURIE prefixes found:\")\n",
    "print(all_prefixes)\n",
    "print(\"\")\n",
    "print(\"unified CURIE prefixes found:\")\n",
    "print(unified_prefixes)\n",
    "print(\"\")\n",
    "print(f\"failed to unify {len(unification_failed_curies)} CURIEs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = []\n",
    "q_node_id_keys = set([\"object\", \"subject\"])\n",
    "q_node_ids = []\n",
    "q_edge_ids = []\n",
    "for q_edge_id,edge_v in query[\"message\"][\"query_graph\"][\"edges\"].items():\n",
    "    q_edge_ids.append(q_edge_id)\n",
    "    q_node_id_found = False\n",
    "    for k,v in edge_v.items():\n",
    "        if (k in q_node_id_keys) and (type(v) is str):\n",
    "            if v not in columns:\n",
    "                q_node_ids.append(v)\n",
    "                columns.append(v)\n",
    "            if not q_node_id_found:\n",
    "                q_node_id_found = True\n",
    "                columns.append(q_edge_id)\n",
    "\n",
    "query_nodes_with_ids = set()\n",
    "for k, v in query['message']['query_graph']['nodes'].items():\n",
    "    node_ids = v.get('ids', [])\n",
    "    if node_ids:\n",
    "        query_nodes_with_ids.add(k)\n",
    "\n",
    "for q_node_id in q_node_ids:\n",
    "    columns.append(f\"{q_node_id}_original_curie\")\n",
    "    columns.append(f\"{q_node_id}_unified_curie\")\n",
    "\n",
    "columns.append(\"unified_curie_set\")\n",
    "\n",
    "trapi_result_columns = []\n",
    "for i in range(len(q_node_ids)):\n",
    "    trapi_result_columns.append(q_node_ids[i])\n",
    "    if i < len(q_edge_ids):\n",
    "        trapi_result_columns.append(q_edge_ids[i])\n",
    "\n",
    "unified_curie_columns = []\n",
    "for q_node_id in q_node_ids:\n",
    "    unified_curie_columns.append(f\"{q_node_id}_unified_curie\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "Wjqcf5E4Yk_h"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "warning: predicate direction(s) may be switched\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n00</th>\n",
       "      <th>e00</th>\n",
       "      <th>n01</th>\n",
       "      <th>n00_original_curie</th>\n",
       "      <th>n00_unified_curie</th>\n",
       "      <th>n01_original_curie</th>\n",
       "      <th>n01_unified_curie</th>\n",
       "      <th>unified_curie_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F8</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:2157</td>\n",
       "      <td>NCBIGene:2157</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:2157}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAMTS13</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:11093</td>\n",
       "      <td>NCBIGene:11093</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{NCBIGene:11093, MESH:C531844}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SELP</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:6403</td>\n",
       "      <td>NCBIGene:6403</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:6403}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PF4</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:5196</td>\n",
       "      <td>NCBIGene:5196</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:5196}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MUC5B</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:727897</td>\n",
       "      <td>NCBIGene:727897</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:727897}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NCBIGene:100128998</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:100128998</td>\n",
       "      <td>NCBIGene:100128998</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:100128998}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NCBIGene:100294715</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:100294715</td>\n",
       "      <td>NCBIGene:100294715</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:100294715}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>LRPAP1</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:4043</td>\n",
       "      <td>NCBIGene:4043</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:4043}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   n00                             e00  \\\n",
       "0                  VWF  gene_associated_with_condition   \n",
       "1                   F8  gene_associated_with_condition   \n",
       "2             ADAMTS13  gene_associated_with_condition   \n",
       "3                 SELP  gene_associated_with_condition   \n",
       "4                  PF4  gene_associated_with_condition   \n",
       "..                 ...                             ...   \n",
       "65               MUC5B  gene_associated_with_condition   \n",
       "66  NCBIGene:100128998  gene_associated_with_condition   \n",
       "67  NCBIGene:100294715  gene_associated_with_condition   \n",
       "68              LRPAP1  gene_associated_with_condition   \n",
       "69                ANO2  gene_associated_with_condition   \n",
       "\n",
       "                                  n01  n00_original_curie   n00_unified_curie  \\\n",
       "0   hereditary von Willebrand disease       NCBIGene:7450       NCBIGene:7450   \n",
       "1   hereditary von Willebrand disease       NCBIGene:2157       NCBIGene:2157   \n",
       "2   hereditary von Willebrand disease      NCBIGene:11093      NCBIGene:11093   \n",
       "3   hereditary von Willebrand disease       NCBIGene:6403       NCBIGene:6403   \n",
       "4   hereditary von Willebrand disease       NCBIGene:5196       NCBIGene:5196   \n",
       "..                                ...                 ...                 ...   \n",
       "65  hereditary von Willebrand disease     NCBIGene:727897     NCBIGene:727897   \n",
       "66  hereditary von Willebrand disease  NCBIGene:100128998  NCBIGene:100128998   \n",
       "67  hereditary von Willebrand disease  NCBIGene:100294715  NCBIGene:100294715   \n",
       "68  hereditary von Willebrand disease       NCBIGene:4043       NCBIGene:4043   \n",
       "69  hereditary von Willebrand disease      NCBIGene:57101      NCBIGene:57101   \n",
       "\n",
       "   n01_original_curie n01_unified_curie                   unified_curie_set  \n",
       "0       MONDO:0019565      MESH:C531844       {MESH:C531844, NCBIGene:7450}  \n",
       "1       MONDO:0019565      MESH:C531844       {MESH:C531844, NCBIGene:2157}  \n",
       "2       MONDO:0019565      MESH:C531844      {NCBIGene:11093, MESH:C531844}  \n",
       "3       MONDO:0019565      MESH:C531844       {MESH:C531844, NCBIGene:6403}  \n",
       "4       MONDO:0019565      MESH:C531844       {MESH:C531844, NCBIGene:5196}  \n",
       "..                ...               ...                                 ...  \n",
       "65      MONDO:0019565      MESH:C531844     {MESH:C531844, NCBIGene:727897}  \n",
       "66      MONDO:0019565      MESH:C531844  {MESH:C531844, NCBIGene:100128998}  \n",
       "67      MONDO:0019565      MESH:C531844  {MESH:C531844, NCBIGene:100294715}  \n",
       "68      MONDO:0019565      MESH:C531844       {MESH:C531844, NCBIGene:4043}  \n",
       "69      MONDO:0019565      MESH:C531844      {MESH:C531844, NCBIGene:57101}  \n",
       "\n",
       "[70 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trapi_results = trapi_message[\"results\"]\n",
    "\n",
    "result_row_data = []\n",
    "for trapi_result in trapi_results:\n",
    "    curie_to_qnode_ids = dict()\n",
    "    for qnode_id, entries in trapi_result[\"node_bindings\"].items():\n",
    "        for entry in entries:\n",
    "            curie = entry[\"id\"]\n",
    "            if curie not in curie_to_qnode_ids:\n",
    "                curie_to_qnode_ids[curie] = []\n",
    "            curie_to_qnode_ids[curie].append(qnode_id)\n",
    "\n",
    "    row_data_template = dict()\n",
    "    q_edge_id_to_predicates = dict()\n",
    "    trapi_result_curie_set = set()\n",
    "    for qedge_id, entries in trapi_result[\"edge_bindings\"].items():\n",
    "        for entry in entries:\n",
    "            curie = entry[\"id\"]\n",
    "            kg_entry = trapi_message[\"knowledge_graph\"][\"edges\"][curie]\n",
    "            subject_curie = kg_entry[\"subject\"]\n",
    "            object_curie = kg_entry[\"object\"]\n",
    "            predicate_curie = kg_entry[\"predicate\"]\n",
    "            [predicate_prefix, predicate_identifier] = predicate_curie.split(\":\")\n",
    "\n",
    "            if qedge_id not in q_edge_id_to_predicates:\n",
    "                q_edge_id_to_predicates[qedge_id] = set()\n",
    "            q_edge_id_to_predicates[qedge_id].add(predicate_identifier)\n",
    "\n",
    "            for curie in [subject_curie, object_curie]:\n",
    "                for qnode_id in curie_to_qnode_ids[curie]:\n",
    "                    if curie in curie_to_unified_curie:\n",
    "                        unified_curie = curie_to_unified_curie[curie]\n",
    "                    else:\n",
    "                        break\n",
    "\n",
    "                    name = curie_to_name[curie]\n",
    "                    row_data_template[qnode_id] = name\n",
    "\n",
    "                    trapi_result_curie_set.add(unified_curie)\n",
    "                    row_data_template[qnode_id + \"_original_curie\"] = curie\n",
    "                    row_data_template[qnode_id + \"_unified_curie\"] = unified_curie\n",
    "\n",
    "    if len(trapi_result_curie_set) != len(q_node_ids):\n",
    "        #print(f'skipping {list(curie_to_qnode_ids.keys())}')\n",
    "        continue\n",
    "\n",
    "    row_data_template[\"unified_curie_set\"] = trapi_result_curie_set\n",
    "    q_edge_ids_processed = set()\n",
    "    row_datas = [row_data_template]\n",
    "    for q_edge_id,predicates in q_edge_id_to_predicates.items():\n",
    "        next_row_datas = []\n",
    "        for row_data in row_datas:\n",
    "            for predicate in predicates:\n",
    "                next_row_data = deepcopy(row_data)\n",
    "                next_row_data[q_edge_id] = predicate\n",
    "                next_row_datas.append(\n",
    "                    next_row_data \n",
    "                )\n",
    "        row_datas = next_row_datas\n",
    "    result_row_data += row_datas\n",
    "\n",
    "print(\"warning: predicate direction(s) may be switched\")\n",
    "trapi_results_df = pd.DataFrame.from_records(result_row_data, columns=columns)\n",
    "trapi_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MESH:C531844'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "user_specified_ids = set()\n",
    "for query_node in query_nodes_with_ids:\n",
    "    user_specified_ids.update(trapi_results_df[f'{query_node}_unified_curie'].tolist())\n",
    "user_specified_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unique CURIE count per query node:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n00: 70\n",
      "n01: 1\n"
     ]
    }
   ],
   "source": [
    "for q_node_id in q_node_ids:\n",
    "    print(f'{q_node_id}: {len(set(trapi_results_df[q_node_id]))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total results count (excluding any that couldn't be unified):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(trapi_results_df[\"unified_curie_set\"].map(sorted).map(tuple)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's too too high for a researcher to efficiently go through manually. Let's try filtering to make that more manageable."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "uCW1x4PxnjYJ",
    "tags": []
   },
   "source": [
    "## Get PFOCR Data\n",
    "\n",
    "We could get it from the API, but for now, we'll just go ahead and download the entire JSON file we gave to BTE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "OyJJUHbepKp2",
    "outputId": "7237b5f0-8740-47cb-9e3d-b03ea02bdcab"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "status_code: 200\n"
     ]
    }
   ],
   "source": [
    "pfocr_url = \"https://www.dropbox.com/s/1f14t5zaseocyg6/bte_chemicals_diseases_genes.ndjson?dl=1\"\n",
    "pfocr_request = requests.get(pfocr_url)\n",
    "print(f\"status_code: {pfocr_request.status_code}\")\n",
    "if pfocr_request.status_code != 200:\n",
    "    print(pfocr_request.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "8fUaenkS0gg1"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_id</th>\n",
       "      <th>chemical_curie_set</th>\n",
       "      <th>disease_curie_set</th>\n",
       "      <th>gene_curie_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:2921, NCBIGene:6426, NCBIGene:2919, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC5793760__cshperspect-TGF-022210_F4.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:654, NCBIGene:27302, NCBIGene:4087, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC5793761__cshperspect-TGF-031989_F1.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:654, NCBIGene:27302, NCBIGene:5594, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC5830892__cshperspect-CEL-027961_F2.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:71, NCBIGene:58, NCBIGene:59, NCBIGe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC5830900__cshperspect-TGF-031997_F1.jpg</td>\n",
       "      <td>{MESH:D011374}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:654, NCBIGene:27302, NCBIGene:3625, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77714</th>\n",
       "      <td>PMC2804790__253_2009_2262_Fig1_HTML.jpg</td>\n",
       "      <td>{MESH:C031105, MESH:D000440, MESH:C043801, MES...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:130, NCBIGene:124, NCBIGene:131, NCB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77715</th>\n",
       "      <td>PMC6332787__thnov09p0126g006.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:8295, NCBIGene:6387, NCBIGene:5133, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77716</th>\n",
       "      <td>PMC5807036__IJO-52-03-0787-g01.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:3643, NCBIGene:285, NCBIGene:3091, N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77717</th>\n",
       "      <td>PMC6770832__cancers-11-01236-g005.jpg</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:4092, NCBIGene:7157, NCBIGene:4088, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77718</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>{MESH:C029269, MESH:D014443}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{NCBIGene:2066, NCBIGene:3845, NCBIGene:306, N...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77719 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       figure_id  \\\n",
       "0      PMC5732092__cshperspect-CYT-028522_F2.jpg   \n",
       "1      PMC5793760__cshperspect-TGF-022210_F4.jpg   \n",
       "2      PMC5793761__cshperspect-TGF-031989_F1.jpg   \n",
       "3      PMC5830892__cshperspect-CEL-027961_F2.jpg   \n",
       "4      PMC5830900__cshperspect-TGF-031997_F1.jpg   \n",
       "...                                          ...   \n",
       "77714    PMC2804790__253_2009_2262_Fig1_HTML.jpg   \n",
       "77715           PMC6332787__thnov09p0126g006.jpg   \n",
       "77716         PMC5807036__IJO-52-03-0787-g01.jpg   \n",
       "77717      PMC6770832__cancers-11-01236-g005.jpg   \n",
       "77718  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   \n",
       "\n",
       "                                      chemical_curie_set disease_curie_set  \\\n",
       "0                                                     {}                {}   \n",
       "1                                                     {}                {}   \n",
       "2                                                     {}                {}   \n",
       "3                                                     {}                {}   \n",
       "4                                         {MESH:D011374}                {}   \n",
       "...                                                  ...               ...   \n",
       "77714  {MESH:C031105, MESH:D000440, MESH:C043801, MES...                {}   \n",
       "77715                                                 {}                {}   \n",
       "77716                                                 {}                {}   \n",
       "77717                                                 {}                {}   \n",
       "77718                       {MESH:C029269, MESH:D014443}                {}   \n",
       "\n",
       "                                          gene_curie_set  \n",
       "0      {NCBIGene:2921, NCBIGene:6426, NCBIGene:2919, ...  \n",
       "1      {NCBIGene:654, NCBIGene:27302, NCBIGene:4087, ...  \n",
       "2      {NCBIGene:654, NCBIGene:27302, NCBIGene:5594, ...  \n",
       "3      {NCBIGene:71, NCBIGene:58, NCBIGene:59, NCBIGe...  \n",
       "4      {NCBIGene:654, NCBIGene:27302, NCBIGene:3625, ...  \n",
       "...                                                  ...  \n",
       "77714  {NCBIGene:130, NCBIGene:124, NCBIGene:131, NCB...  \n",
       "77715  {NCBIGene:8295, NCBIGene:6387, NCBIGene:5133, ...  \n",
       "77716  {NCBIGene:3643, NCBIGene:285, NCBIGene:3091, N...  \n",
       "77717  {NCBIGene:4092, NCBIGene:7157, NCBIGene:4088, ...  \n",
       "77718  {NCBIGene:2066, NCBIGene:3845, NCBIGene:306, N...  \n",
       "\n",
       "[77719 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "figures_df_data = []\n",
    "curies_to_figure_ids = {}\n",
    "figure_id_to_curies = {}\n",
    "figure_id_to_pfocr_result = {}\n",
    "for line in pfocr_request.text.splitlines():\n",
    "    pfocr_result = json.loads(line)\n",
    "    figure_id = pfocr_result[\"_id\"]\n",
    "\n",
    "    chemical_curies = set()\n",
    "    disease_curies = set()\n",
    "    gene_curies = set()\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"chemicals\"][\"mesh\"]:\n",
    "        curie = \"MESH:\" + identifier\n",
    "        chemical_curies.add(curie)\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"diseases\"][\"mesh\"]:\n",
    "        curie = \"MESH:\" + identifier\n",
    "        disease_curies.add(curie)\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"genes\"][\"ncbigene\"]:\n",
    "        curie = \"NCBIGene:\" + identifier\n",
    "        gene_curies.add(curie)\n",
    "\n",
    "    figure_id_to_pfocr_result[figure_id] = pfocr_result\n",
    "    figure_id_to_curies[figure_id] = curies\n",
    "\n",
    "    figures_df_data.append({\n",
    "        \"figure_id\": figure_id,\n",
    "        \"chemical_curie_set\": chemical_curies,\n",
    "        \"disease_curie_set\": disease_curies,\n",
    "        \"gene_curie_set\": gene_curies,\n",
    "    })\n",
    "\n",
    "figures_df = pd.DataFrame.from_records(figures_df_data)\n",
    "figures_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "biolink_category_to_pfocr_category = {\n",
    "    \"biolink:Drug\": \"chemical\",\n",
    "    \"biolink:ChemicalEntity\": \"chemical\",\n",
    "    \"biolink:SmallMolecule\": \"chemical\",\n",
    "    \"biolink:Disease\": \"disease\",\n",
    "    \"biolink:DiseaseOrPhenotypicFeature\": \"disease\",\n",
    "    \"biolink:Gene\": \"gene\",\n",
    "    \"biolink:GeneProduct\": \"gene\",\n",
    "    \"biolink:Protein\": \"gene\",\n",
    "}\n",
    "\n",
    "# limit to categories returned from query\n",
    "relevant_biolink_categories = set(biolink_category_to_pfocr_category.keys()).intersection(curie_categories)\n",
    "relevant_pfocr_categories = set([\n",
    "    biolink_category_to_pfocr_category[\n",
    "        biolink_category\n",
    "    ] for biolink_category in relevant_biolink_categories\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_url</th>\n",
       "      <th>figure_title</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figure_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC5732092__cshperspect-CYT-028522_F2.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Interleukin (IL)-17RA/RC signaling pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5793760__cshperspect-TGF-022210_F4.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>TGFB signaling pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5793761__cshperspect-TGF-031989_F1.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Bone morphogenetic protein (BMP) signaling pat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5830892__cshperspect-CEL-027961_F2.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>FERM-binding partners of Crumbs3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5830900__cshperspect-TGF-031997_F1.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Role of the TGFB family in mammary gland devel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC2804790__253_2009_2262_Fig1_HTML.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "      <td>Ehrlich degradation pathway from amino acid to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6332787__thnov09p0126g006.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Illustration of the mechanisms underlying LPD ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC5807036__IJO-52-03-0787-g01.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "      <td>Tumorigenic proteins significantly upregulated...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6770832__cancers-11-01236-g005.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Simplified TGFB pathway leading to p21 expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</th>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "      <td>EGFR and ANXA3 to fluoropyrimidine resistance ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77719 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                  figure_url  \\\n",
       "figure_id                                                                                      \n",
       "PMC5732092__cshperspect-CYT-028522_F2.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5793760__cshperspect-TGF-022210_F4.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5793761__cshperspect-TGF-031989_F1.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5830892__cshperspect-CEL-027961_F2.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC5830900__cshperspect-TGF-031997_F1.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "...                                                                                      ...   \n",
       "PMC2804790__253_2009_2262_Fig1_HTML.jpg    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...   \n",
       "PMC6332787__thnov09p0126g006.jpg           https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "PMC5807036__IJO-52-03-0787-g01.jpg         https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...   \n",
       "PMC6770832__cancers-11-01236-g005.jpg      https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "PMC4609065__12885_2015_1721_Fig5_HTML.jpg  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...   \n",
       "\n",
       "                                                                                figure_title  \n",
       "figure_id                                                                                     \n",
       "PMC5732092__cshperspect-CYT-028522_F2.jpg        Interleukin (IL)-17RA/RC signaling pathways  \n",
       "PMC5793760__cshperspect-TGF-022210_F4.jpg                            TGFB signaling pathways  \n",
       "PMC5793761__cshperspect-TGF-031989_F1.jpg  Bone morphogenetic protein (BMP) signaling pat...  \n",
       "PMC5830892__cshperspect-CEL-027961_F2.jpg                   FERM-binding partners of Crumbs3  \n",
       "PMC5830900__cshperspect-TGF-031997_F1.jpg  Role of the TGFB family in mammary gland devel...  \n",
       "...                                                                                      ...  \n",
       "PMC2804790__253_2009_2262_Fig1_HTML.jpg    Ehrlich degradation pathway from amino acid to...  \n",
       "PMC6332787__thnov09p0126g006.jpg           Illustration of the mechanisms underlying LPD ...  \n",
       "PMC5807036__IJO-52-03-0787-g01.jpg         Tumorigenic proteins significantly upregulated...  \n",
       "PMC6770832__cancers-11-01236-g005.jpg      Simplified TGFB pathway leading to p21 expression  \n",
       "PMC4609065__12885_2015_1721_Fig5_HTML.jpg  EGFR and ANXA3 to fluoropyrimidine resistance ...  \n",
       "\n",
       "[77719 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "figure_metadata_records = []\n",
    "figure_curie_df_data = []\n",
    "for line in pfocr_request.text.splitlines():\n",
    "    pfocr_result = json.loads(line)\n",
    "    figure_id = pfocr_result[\"_id\"]\n",
    "\n",
    "    chemical_curies = set()\n",
    "    disease_curies = set()\n",
    "    gene_curies = set()\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"chemicals\"][\"mesh\"]:\n",
    "        curie = \"MESH:\" + identifier\n",
    "        chemical_curies.add(curie)\n",
    "        figure_curie_df_data.append({\n",
    "            \"figure_id\": figure_id,\n",
    "            \"curie\": curie,\n",
    "            \"category\": \"chemical\",\n",
    "        })\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"diseases\"][\"mesh\"]:\n",
    "        curie = \"MESH:\" + identifier\n",
    "        disease_curies.add(curie)\n",
    "        figure_curie_df_data.append({\n",
    "            \"figure_id\": figure_id,\n",
    "            \"curie\": curie,\n",
    "            \"category\": \"disease\",\n",
    "        })\n",
    "    for identifier in pfocr_result[\"associatedWith\"][\"mentions\"][\"genes\"][\"ncbigene\"]:\n",
    "        curie = \"NCBIGene:\" + identifier\n",
    "        gene_curies.add(curie)\n",
    "        figure_curie_df_data.append({\n",
    "            \"figure_id\": figure_id,\n",
    "            \"curie\": curie,\n",
    "            \"category\": \"gene\",\n",
    "        })\n",
    "\n",
    "    figure_id_to_pfocr_result[figure_id] = pfocr_result\n",
    "    figure_id_to_curies[figure_id] = curies\n",
    "\n",
    "    figure_metadata_records.append({\n",
    "        \"figure_id\": figure_id,\n",
    "        \"figure_url\": pfocr_result[\"associatedWith\"][\"figureUrl\"],\n",
    "        \"figure_title\": pfocr_result[\"associatedWith\"][\"title\"],\n",
    "    })\n",
    "\n",
    "figure_metadata_df = pd.DataFrame.from_records(figure_metadata_records, index='figure_id')\n",
    "figure_metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figure_id</th>\n",
       "      <th>curie</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:5601</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:5595</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:1051</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:10131</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC5732092__cshperspect-CYT-028522_F2.jpg</td>\n",
       "      <td>NCBIGene:23765</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665596</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:1956</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665597</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:4893</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665598</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:2064</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665599</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:3845</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665600</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>NCBIGene:306</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1665601 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         figure_id           curie category\n",
       "0        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:5601     gene\n",
       "1        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:5595     gene\n",
       "2        PMC5732092__cshperspect-CYT-028522_F2.jpg   NCBIGene:1051     gene\n",
       "3        PMC5732092__cshperspect-CYT-028522_F2.jpg  NCBIGene:10131     gene\n",
       "4        PMC5732092__cshperspect-CYT-028522_F2.jpg  NCBIGene:23765     gene\n",
       "...                                            ...             ...      ...\n",
       "1665596  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:1956     gene\n",
       "1665597  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:4893     gene\n",
       "1665598  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:2064     gene\n",
       "1665599  PMC4609065__12885_2015_1721_Fig5_HTML.jpg   NCBIGene:3845     gene\n",
       "1665600  PMC4609065__12885_2015_1721_Fig5_HTML.jpg    NCBIGene:306     gene\n",
       "\n",
       "[1665601 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "figure_curie_df = pd.DataFrame.from_records(figure_curie_df_data)\n",
    "figure_curie_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['figure_id', 'curie', 'category'], dtype='object')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "figure_curie_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Identify Cluster Figures\n",
    "\n",
    "See details in write-up at the top."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all_relevant_curie_count: 15683\n",
      "i: 0\n",
      "i: 1\n",
      "i: 2\n",
      "i: 3\n",
      "i: 4\n",
      "i: 5\n",
      "i: 6\n",
      "i: 7\n",
      "i: 8\n",
      "i: 9\n",
      "i: 10\n",
      "i: 11\n",
      "i: 12\n",
      "i: 13\n",
      "i: 14\n",
      "i: 15\n",
      "i: 16\n",
      "i: 17\n",
      "i: 18\n",
      "i: 19\n",
      "i: 20\n",
      "cluster figure count: 61\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "      <th>figure_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>next_best_figure_ids</th>\n",
       "      <th>next_best_p_value</th>\n",
       "      <th>yes_trapi_results_yes_figure</th>\n",
       "      <th>yes_trapi_results_no_figure</th>\n",
       "      <th>no_trapi_results_yes_figure</th>\n",
       "      <th>no_trapi_results_no_figure</th>\n",
       "      <th>figure_curies_in_trapi_results</th>\n",
       "      <th>curie_set</th>\n",
       "      <th>curies_to_exclude</th>\n",
       "      <th>len_curies_to_exclude</th>\n",
       "      <th>relevant_figures_unique_curie_count</th>\n",
       "      <th>relevant_figure_count</th>\n",
       "      <th>remaining_trapi_results_unified_curies</th>\n",
       "      <th>cluster_figure_id_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>PMC6797170__pone.0223840.g007.jpg</td>\n",
       "      <td>2.423696e-08</td>\n",
       "      <td>[PMC5016740__JCHIMP-6-32461-g003.jpg]</td>\n",
       "      <td>2.662649e-08</td>\n",
       "      <td>9</td>\n",
       "      <td>62</td>\n",
       "      <td>11</td>\n",
       "      <td>15601</td>\n",
       "      <td>{NCBIGene:2158, NCBIGene:2152, NCBIGene:7450, ...</td>\n",
       "      <td>{NCBIGene:2152, NCBIGene:5329, NCBIGene:7056, ...</td>\n",
       "      <td>{NCBIGene:2158, NCBIGene:2152, NCBIGene:7450, ...</td>\n",
       "      <td>9</td>\n",
       "      <td>15683</td>\n",
       "      <td>74144</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>PMC2583569__zii0120877060006.jpg</td>\n",
       "      <td>3.888715e-07</td>\n",
       "      <td>[PMC6046593__RTH2-2-439-g001.jpg]</td>\n",
       "      <td>4.253464e-07</td>\n",
       "      <td>4</td>\n",
       "      <td>58</td>\n",
       "      <td>12</td>\n",
       "      <td>15609</td>\n",
       "      <td>{NCBIGene:2811, NCBIGene:3674, NCBIGene:64805,...</td>\n",
       "      <td>{NCBIGene:5335, NCBIGene:2811, NCBIGene:3674, ...</td>\n",
       "      <td>{NCBIGene:2811, NCBIGene:3674, NCBIGene:64805,...</td>\n",
       "      <td>4</td>\n",
       "      <td>15683</td>\n",
       "      <td>74143</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>PMC5016740__JCHIMP-6-32461-g003.jpg</td>\n",
       "      <td>2.653114e-06</td>\n",
       "      <td>[PMC3242600__bjc2011442f1.jpg, PMC4933178__zbc...</td>\n",
       "      <td>4.023077e-05</td>\n",
       "      <td>3</td>\n",
       "      <td>55</td>\n",
       "      <td>5</td>\n",
       "      <td>15620</td>\n",
       "      <td>{NCBIGene:1351, NCBIGene:2147, NCBIGene:2160}</td>\n",
       "      <td>{NCBIGene:2243, NCBIGene:2158, NCBIGene:695, N...</td>\n",
       "      <td>{NCBIGene:1351, NCBIGene:2147, NCBIGene:2160}</td>\n",
       "      <td>3</td>\n",
       "      <td>15683</td>\n",
       "      <td>74142</td>\n",
       "      <td>58</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>PMC3242600__bjc2011442f1.jpg</td>\n",
       "      <td>3.614660e-05</td>\n",
       "      <td>[PMC3328452__pone.0035544.g008.jpg]</td>\n",
       "      <td>3.800687e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>15627</td>\n",
       "      <td>{NCBIGene:5045, NCBIGene:4524}</td>\n",
       "      <td>{NCBIGene:5045, NCBIGene:4524, NCBIGene:1806}</td>\n",
       "      <td>{NCBIGene:5045, NCBIGene:4524, NCBIGene:6418}</td>\n",
       "      <td>3</td>\n",
       "      <td>15683</td>\n",
       "      <td>74141</td>\n",
       "      <td>55</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>PMC4933178__zbc0291646780009.jpg</td>\n",
       "      <td>3.614660e-05</td>\n",
       "      <td>[PMC3328452__pone.0035544.g008.jpg]</td>\n",
       "      <td>3.800687e-05</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>15627</td>\n",
       "      <td>{NCBIGene:6418, NCBIGene:5045}</td>\n",
       "      <td>{NCBIGene:6418, NCBIGene:5045, NCBIGene:10743}</td>\n",
       "      <td>{NCBIGene:5045, NCBIGene:4524, NCBIGene:6418}</td>\n",
       "      <td>3</td>\n",
       "      <td>15683</td>\n",
       "      <td>74141</td>\n",
       "      <td>55</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>17</td>\n",
       "      <td>PMC2533280__nihms64392f1.jpg</td>\n",
       "      <td>5.725904e-03</td>\n",
       "      <td>[PMC5839936__cir-137-0999-g007.jpg, PMC6321635...</td>\n",
       "      <td>7.627729e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>15663</td>\n",
       "      <td>{NCBIGene:8646}</td>\n",
       "      <td>{NCBIGene:55897, NCBIGene:3674, NCBIGene:8646,...</td>\n",
       "      <td>{NCBIGene:8646}</td>\n",
       "      <td>1</td>\n",
       "      <td>15681</td>\n",
       "      <td>74088</td>\n",
       "      <td>15</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>18</td>\n",
       "      <td>PMC5839936__cir-137-0999-g007.jpg</td>\n",
       "      <td>7.120802e-03</td>\n",
       "      <td>[PMC5348403__oncotarget-07-81423-g004.jpg]</td>\n",
       "      <td>1.154736e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>15662</td>\n",
       "      <td>{NCBIGene:285}</td>\n",
       "      <td>{NCBIGene:654, NCBIGene:285, NCBIGene:27122, N...</td>\n",
       "      <td>{NCBIGene:10332, NCBIGene:285}</td>\n",
       "      <td>2</td>\n",
       "      <td>15681</td>\n",
       "      <td>74087</td>\n",
       "      <td>14</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>18</td>\n",
       "      <td>PMC6321635__ijms-19-03856-g003.jpg</td>\n",
       "      <td>7.120802e-03</td>\n",
       "      <td>[PMC5348403__oncotarget-07-81423-g004.jpg]</td>\n",
       "      <td>1.154736e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>15662</td>\n",
       "      <td>{NCBIGene:285}</td>\n",
       "      <td>{NCBIGene:2277, NCBIGene:285, NCBIGene:5228, N...</td>\n",
       "      <td>{NCBIGene:10332, NCBIGene:285}</td>\n",
       "      <td>2</td>\n",
       "      <td>15681</td>\n",
       "      <td>74087</td>\n",
       "      <td>14</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>18</td>\n",
       "      <td>PMC7599810__antioxidants-09-00936-g002.jpg</td>\n",
       "      <td>7.120802e-03</td>\n",
       "      <td>[PMC5348403__oncotarget-07-81423-g004.jpg]</td>\n",
       "      <td>1.154736e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>15662</td>\n",
       "      <td>{NCBIGene:10332}</td>\n",
       "      <td>{MESH:C000657245, NCBIGene:7124, NCBIGene:3606...</td>\n",
       "      <td>{NCBIGene:10332, NCBIGene:285}</td>\n",
       "      <td>2</td>\n",
       "      <td>15681</td>\n",
       "      <td>74087</td>\n",
       "      <td>14</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>19</td>\n",
       "      <td>PMC5348403__oncotarget-07-81423-g004.jpg</td>\n",
       "      <td>9.905310e-03</td>\n",
       "      <td>[PMC3188957__nihms-318135-f0004.jpg]</td>\n",
       "      <td>1.066352e-02</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>15659</td>\n",
       "      <td>{NCBIGene:3673}</td>\n",
       "      <td>{NCBIGene:3672, NCBIGene:7422, NCBIGene:4318, ...</td>\n",
       "      <td>{NCBIGene:3673}</td>\n",
       "      <td>1</td>\n",
       "      <td>15681</td>\n",
       "      <td>74084</td>\n",
       "      <td>12</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     i                                   figure_id       p_value  \\\n",
       "0    0           PMC6797170__pone.0223840.g007.jpg  2.423696e-08   \n",
       "1    1            PMC2583569__zii0120877060006.jpg  3.888715e-07   \n",
       "2    2         PMC5016740__JCHIMP-6-32461-g003.jpg  2.653114e-06   \n",
       "3    3                PMC3242600__bjc2011442f1.jpg  3.614660e-05   \n",
       "4    3            PMC4933178__zbc0291646780009.jpg  3.614660e-05   \n",
       "..  ..                                         ...           ...   \n",
       "56  17                PMC2533280__nihms64392f1.jpg  5.725904e-03   \n",
       "57  18           PMC5839936__cir-137-0999-g007.jpg  7.120802e-03   \n",
       "58  18          PMC6321635__ijms-19-03856-g003.jpg  7.120802e-03   \n",
       "59  18  PMC7599810__antioxidants-09-00936-g002.jpg  7.120802e-03   \n",
       "60  19    PMC5348403__oncotarget-07-81423-g004.jpg  9.905310e-03   \n",
       "\n",
       "                                 next_best_figure_ids  next_best_p_value  \\\n",
       "0               [PMC5016740__JCHIMP-6-32461-g003.jpg]       2.662649e-08   \n",
       "1                   [PMC6046593__RTH2-2-439-g001.jpg]       4.253464e-07   \n",
       "2   [PMC3242600__bjc2011442f1.jpg, PMC4933178__zbc...       4.023077e-05   \n",
       "3                 [PMC3328452__pone.0035544.g008.jpg]       3.800687e-05   \n",
       "4                 [PMC3328452__pone.0035544.g008.jpg]       3.800687e-05   \n",
       "..                                                ...                ...   \n",
       "56  [PMC5839936__cir-137-0999-g007.jpg, PMC6321635...       7.627729e-03   \n",
       "57         [PMC5348403__oncotarget-07-81423-g004.jpg]       1.154736e-02   \n",
       "58         [PMC5348403__oncotarget-07-81423-g004.jpg]       1.154736e-02   \n",
       "59         [PMC5348403__oncotarget-07-81423-g004.jpg]       1.154736e-02   \n",
       "60               [PMC3188957__nihms-318135-f0004.jpg]       1.066352e-02   \n",
       "\n",
       "    yes_trapi_results_yes_figure  yes_trapi_results_no_figure  \\\n",
       "0                              9                           62   \n",
       "1                              4                           58   \n",
       "2                              3                           55   \n",
       "3                              2                           53   \n",
       "4                              2                           53   \n",
       "..                           ...                          ...   \n",
       "56                             1                           14   \n",
       "57                             1                           13   \n",
       "58                             1                           13   \n",
       "59                             1                           13   \n",
       "60                             1                           11   \n",
       "\n",
       "    no_trapi_results_yes_figure  no_trapi_results_no_figure  \\\n",
       "0                            11                       15601   \n",
       "1                            12                       15609   \n",
       "2                             5                       15620   \n",
       "3                             1                       15627   \n",
       "4                             1                       15627   \n",
       "..                          ...                         ...   \n",
       "56                            5                       15663   \n",
       "57                            7                       15662   \n",
       "58                            7                       15662   \n",
       "59                            7                       15662   \n",
       "60                           12                       15659   \n",
       "\n",
       "                       figure_curies_in_trapi_results  \\\n",
       "0   {NCBIGene:2158, NCBIGene:2152, NCBIGene:7450, ...   \n",
       "1   {NCBIGene:2811, NCBIGene:3674, NCBIGene:64805,...   \n",
       "2       {NCBIGene:1351, NCBIGene:2147, NCBIGene:2160}   \n",
       "3                      {NCBIGene:5045, NCBIGene:4524}   \n",
       "4                      {NCBIGene:6418, NCBIGene:5045}   \n",
       "..                                                ...   \n",
       "56                                    {NCBIGene:8646}   \n",
       "57                                     {NCBIGene:285}   \n",
       "58                                     {NCBIGene:285}   \n",
       "59                                   {NCBIGene:10332}   \n",
       "60                                    {NCBIGene:3673}   \n",
       "\n",
       "                                            curie_set  \\\n",
       "0   {NCBIGene:2152, NCBIGene:5329, NCBIGene:7056, ...   \n",
       "1   {NCBIGene:5335, NCBIGene:2811, NCBIGene:3674, ...   \n",
       "2   {NCBIGene:2243, NCBIGene:2158, NCBIGene:695, N...   \n",
       "3       {NCBIGene:5045, NCBIGene:4524, NCBIGene:1806}   \n",
       "4      {NCBIGene:6418, NCBIGene:5045, NCBIGene:10743}   \n",
       "..                                                ...   \n",
       "56  {NCBIGene:55897, NCBIGene:3674, NCBIGene:8646,...   \n",
       "57  {NCBIGene:654, NCBIGene:285, NCBIGene:27122, N...   \n",
       "58  {NCBIGene:2277, NCBIGene:285, NCBIGene:5228, N...   \n",
       "59  {MESH:C000657245, NCBIGene:7124, NCBIGene:3606...   \n",
       "60  {NCBIGene:3672, NCBIGene:7422, NCBIGene:4318, ...   \n",
       "\n",
       "                                    curies_to_exclude  len_curies_to_exclude  \\\n",
       "0   {NCBIGene:2158, NCBIGene:2152, NCBIGene:7450, ...                      9   \n",
       "1   {NCBIGene:2811, NCBIGene:3674, NCBIGene:64805,...                      4   \n",
       "2       {NCBIGene:1351, NCBIGene:2147, NCBIGene:2160}                      3   \n",
       "3       {NCBIGene:5045, NCBIGene:4524, NCBIGene:6418}                      3   \n",
       "4       {NCBIGene:5045, NCBIGene:4524, NCBIGene:6418}                      3   \n",
       "..                                                ...                    ...   \n",
       "56                                    {NCBIGene:8646}                      1   \n",
       "57                     {NCBIGene:10332, NCBIGene:285}                      2   \n",
       "58                     {NCBIGene:10332, NCBIGene:285}                      2   \n",
       "59                     {NCBIGene:10332, NCBIGene:285}                      2   \n",
       "60                                    {NCBIGene:3673}                      1   \n",
       "\n",
       "    relevant_figures_unique_curie_count  relevant_figure_count  \\\n",
       "0                                 15683                  74144   \n",
       "1                                 15683                  74143   \n",
       "2                                 15683                  74142   \n",
       "3                                 15683                  74141   \n",
       "4                                 15683                  74141   \n",
       "..                                  ...                    ...   \n",
       "56                                15681                  74088   \n",
       "57                                15681                  74087   \n",
       "58                                15681                  74087   \n",
       "59                                15681                  74087   \n",
       "60                                15681                  74084   \n",
       "\n",
       "    remaining_trapi_results_unified_curies  cluster_figure_id_count  \n",
       "0                                       71                        1  \n",
       "1                                       62                        2  \n",
       "2                                       58                        3  \n",
       "3                                       55                        5  \n",
       "4                                       55                        5  \n",
       "..                                     ...                      ...  \n",
       "56                                      15                       57  \n",
       "57                                      14                       60  \n",
       "58                                      14                       60  \n",
       "59                                      14                       60  \n",
       "60                                      12                       61  \n",
       "\n",
       "[61 rows x 17 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Importing reduce for \n",
    "# rolling computations\n",
    "from functools import reduce\n",
    "  \n",
    "def set_union(series):\n",
    "    return reduce(lambda x, y: x | y, series)\n",
    "\n",
    "def create_set(series):\n",
    "    return set(series.tolist())\n",
    "\n",
    "\n",
    "# TODO: remove cluster_figure_data and its associated DF.\n",
    "# It's just a temporary thing to show what's happening inside.\n",
    "cluster_figure_data = []\n",
    "def get_next_cluster_figures(\n",
    "    all_relevant_curie_count,\n",
    "    remaining_relevant_figure_curie_df,\n",
    "    remaining_trapi_results_unified_curies,\n",
    "    iteration_limit=20,\n",
    "    i=0,\n",
    "    cluster_figure_ids=[],\n",
    "):\n",
    "    print(f'i: {i}')\n",
    "    \n",
    "    relevant_figures_unique_curie_count = len(set(\n",
    "        remaining_relevant_figure_curie_df[\"curie\"].drop_duplicates()\n",
    "    ))\n",
    "    \n",
    "    relevant_figure_count = len(set(\n",
    "        remaining_relevant_figure_curie_df[\"figure_id\"].drop_duplicates()\n",
    "    ))\n",
    "\n",
    "    if (\n",
    "        i >= iteration_limit\n",
    "    ) or (\n",
    "        len(remaining_trapi_results_unified_curies) == 0\n",
    "    ) or (\n",
    "        relevant_figure_count == 0\n",
    "    ):\n",
    "        return cluster_figure_ids\n",
    "    \n",
    "    remaining_figures_df = remaining_relevant_figure_curie_df[[\"figure_id\", \"curie\"]].groupby(\"figure_id\").agg(\n",
    "        create_set\n",
    "    ).rename(columns={\n",
    "        \"curie\": \"curie_set\"\n",
    "    })\n",
    "\n",
    "    remaining_figures_df[\"figure_curie_count\"] = remaining_figures_df[\"curie_set\"].map(len)\n",
    "\n",
    "    remaining_figures_df[\"figure_curies_in_trapi_results\"] = remaining_figures_df[\"curie_set\"].map(\n",
    "        lambda curie_set: remaining_trapi_results_unified_curies.intersection(\n",
    "            set(curie_set)\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # number of CURIEs in both TRAPI results and figure\n",
    "    remaining_figures_df[\"yes_trapi_results_yes_figure\"] = remaining_figures_df[\n",
    "        \"figure_curies_in_trapi_results\"\n",
    "    ].map(len)\n",
    "\n",
    "    # number of CURIEs in TRAPI results but not in figure\n",
    "    remaining_figures_df[\"yes_trapi_results_no_figure\"] = (\n",
    "        len(remaining_trapi_results_unified_curies) -\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # number of CURIEs not in TRAPI results but in figure\n",
    "    remaining_figures_df[\"no_trapi_results_yes_figure\"] = (\n",
    "        remaining_figures_df[\"figure_curie_count\"] -\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # number of CURIEs not in TRAPI results and not in figure,\n",
    "    # ie., number of unique CURIEs only in other figures\n",
    "    remaining_figures_df[\"no_trapi_results_no_figure\"] = (\n",
    "        all_relevant_curie_count -\n",
    "        len(remaining_trapi_results_unified_curies) -\n",
    "        remaining_figures_df[\"no_trapi_results_yes_figure\"]\n",
    "    )\n",
    "\n",
    "    # see https://stackoverflow.com/a/58661068/5354298\n",
    "    _, _, twosided = pvalue_npy(\n",
    "        remaining_figures_df[\"yes_trapi_results_yes_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"yes_trapi_results_no_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"no_trapi_results_yes_figure\"].to_numpy('uint'),\n",
    "        remaining_figures_df[\"no_trapi_results_no_figure\"].to_numpy('uint'),\n",
    "    )\n",
    "\n",
    "    remaining_figures_df[\"p_value\"] = pd.Series(twosided, index=remaining_figures_df.index)\n",
    "    remaining_figures_df[\"p_value\"].sort_values()\n",
    "    \n",
    "    min_df = remaining_figures_df[remaining_figures_df[\"p_value\"] == remaining_figures_df[\"p_value\"].min()]\n",
    "        \n",
    "    cluster_figure_ids.extend(min_df.index.tolist())\n",
    "    \n",
    "    # TODO: should we always keep the user-specified CURIE ids?\n",
    "    curies_to_exclude = set(min_df[\"figure_curies_in_trapi_results\"].agg(set_union)) - user_specified_ids\n",
    "    \n",
    "    next_min_df_score = remaining_figures_df[\n",
    "        remaining_figures_df[\"p_value\"] > remaining_figures_df[\"p_value\"].min()\n",
    "    ][\"p_value\"].min()\n",
    "    next_min_df = remaining_figures_df[remaining_figures_df[\"p_value\"] == next_min_df_score]\n",
    "    \n",
    "    for cluster_figure_id, min_row in min_df.iterrows():\n",
    "        cluster_figure_data.append({\n",
    "            \"i\": i,\n",
    "            \"figure_id\": cluster_figure_id,\n",
    "            \"p_value\": min_row[\"p_value\"],\n",
    "            \"next_best_figure_ids\": next_min_df.index.tolist(),\n",
    "            \"next_best_p_value\": next_min_df[\"p_value\"].min(),\n",
    "            \"yes_trapi_results_yes_figure\": min_row[\"yes_trapi_results_yes_figure\"],\n",
    "            \"yes_trapi_results_no_figure\": min_row[\"yes_trapi_results_no_figure\"],\n",
    "            \"no_trapi_results_yes_figure\": min_row[\"no_trapi_results_yes_figure\"],\n",
    "            \"no_trapi_results_no_figure\": min_row[\"no_trapi_results_no_figure\"],\n",
    "            \"figure_curies_in_trapi_results\": min_row[\"figure_curies_in_trapi_results\"],\n",
    "            \"curie_set\": min_row[\"curie_set\"],\n",
    "            \"curies_to_exclude\": curies_to_exclude,\n",
    "            \"len_curies_to_exclude\": len(curies_to_exclude),\n",
    "            \"relevant_figures_unique_curie_count\": relevant_figures_unique_curie_count,\n",
    "            \"relevant_figure_count\": relevant_figure_count,\n",
    "            \"remaining_trapi_results_unified_curies\": len(remaining_trapi_results_unified_curies),\n",
    "            \"cluster_figure_id_count\": len(cluster_figure_ids),\n",
    "        })\n",
    "\n",
    "    if len(curies_to_exclude) == 0:\n",
    "        print(\"Only matching user-specified CURIE(s). Cluster figure search complete.\")\n",
    "        return cluster_figure_ids\n",
    "\n",
    "    next_remaining_relevant_figure_curie_df = remaining_relevant_figure_curie_df[(~remaining_relevant_figure_curie_df[\"figure_id\"].isin(\n",
    "        cluster_figure_ids\n",
    "    ))]\n",
    "    \n",
    "    return get_next_cluster_figures(\n",
    "        all_relevant_curie_count,\n",
    "        next_remaining_relevant_figure_curie_df,\n",
    "        remaining_trapi_results_unified_curies - curies_to_exclude,\n",
    "        iteration_limit,\n",
    "        i + 1,\n",
    "        cluster_figure_ids,\n",
    "    )\n",
    "\n",
    "\n",
    "# don't worry whether cluster figure has user-specified IDs\n",
    "relevant_figure_curie_df = figure_curie_df[figure_curie_df[\"category\"].isin(\n",
    "    list(relevant_pfocr_categories)\n",
    ")]\n",
    "# only use cluster figures that have the user-specified ids\n",
    "#figures_with_user_specified_ids = set(figure_curie_df[figure_curie_df[\"curie\"].isin(user_specified_ids)][\"figure_id\"])\n",
    "#print(f'number of figures w/ user-specified CURIEs: {len(figures_with_user_specified_ids)}')\n",
    "#relevant_figure_curie_df = figure_curie_df[figure_curie_df[\"category\"].isin(\n",
    "#    list(relevant_pfocr_categories)\n",
    "#) & figure_curie_df[\"figure_id\"].isin(\n",
    "#    figures_with_user_specified_ids\n",
    "#)]\n",
    "\n",
    "all_relevant_curie_count = len(set(\n",
    "    relevant_figure_curie_df[\"curie\"].drop_duplicates()\n",
    "))\n",
    "print(f'all_relevant_curie_count: {all_relevant_curie_count}')\n",
    "\n",
    "cluster_figure_ids_out = get_next_cluster_figures(\n",
    "    all_relevant_curie_count,\n",
    "    relevant_figure_curie_df,\n",
    "    trapi_results_unified_curies,\n",
    ")\n",
    "print(f'cluster figure count: {len(cluster_figure_ids_out)}')\n",
    "cluster_figure_df = pd.DataFrame.from_records(cluster_figure_data)\n",
    "cluster_figure_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cluster BTE TRAPI Results by Selected PFOCR Figures (Cluster Figures)\n",
    "\n",
    "See details in write-up at the top."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n00</th>\n",
       "      <th>e00</th>\n",
       "      <th>n01</th>\n",
       "      <th>n00_original_curie</th>\n",
       "      <th>n00_unified_curie</th>\n",
       "      <th>n01_original_curie</th>\n",
       "      <th>n01_unified_curie</th>\n",
       "      <th>trapi_result_curie_set</th>\n",
       "      <th>crossjoin_key</th>\n",
       "      <th>figure_id</th>\n",
       "      <th>...</th>\n",
       "      <th>figure_url</th>\n",
       "      <th>figure_title</th>\n",
       "      <th>figure_curie_count</th>\n",
       "      <th>common_curie_set</th>\n",
       "      <th>yes_trapi_result_yes_figure</th>\n",
       "      <th>yes_trapi_result_no_figure</th>\n",
       "      <th>no_trapi_result_yes_figure</th>\n",
       "      <th>no_trapi_result_no_figure</th>\n",
       "      <th>p_value</th>\n",
       "      <th>trapi_result_curies_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1262766__1743-7075-2-22-1.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>The kynuric pathway of melatonin metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:7450)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1489360__zam0070669260006.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Postulated biosynthetic pathway of ABA in C</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:7450)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1572915__134-0704215f2.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Biosynthesis of lyso-PAF and PAF via the remod...</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:7450)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC181284__pp0838025001.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>A, Metabolic pathway for the synthesis of MAs,...</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:7450)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VWF</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>NCBIGene:7450</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:7450}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1932822__zam0140779890003.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Six distinct ring-cleaving pathways present in...</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:7450)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4265</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6797170__pone.0223840.g007.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Identification of protein changes in the blood...</td>\n",
       "      <td>20</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>15661</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:57101)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4266</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC7238249__plants-09-00422-g002.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "      <td>Proposed model of the regulatory functions of ...</td>\n",
       "      <td>4</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>15677</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:57101)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4267</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC7242693__zbc9992023450012.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "      <td>Synthesis of bacteriochlorophylls c, d, e, and...</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:57101)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4268</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC7599810__antioxidants-09-00936-g002.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "      <td>NETosis activated by SARS-CoV-2</td>\n",
       "      <td>8</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>15673</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:57101)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4269</th>\n",
       "      <td>ANO2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>NCBIGene:57101</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:57101}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC7608132__41401_2020_389_Fig5_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...</td>\n",
       "      <td>Proposed metabolic pathway of alflutinib in HLMs</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>15680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(MESH:C531844, NCBIGene:57101)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4270 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       n00                             e00                                n01  \\\n",
       "0      VWF  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "1      VWF  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "2      VWF  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "3      VWF  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "4      VWF  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "...    ...                             ...                                ...   \n",
       "4265  ANO2  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "4266  ANO2  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "4267  ANO2  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "4268  ANO2  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "4269  ANO2  gene_associated_with_condition  hereditary von Willebrand disease   \n",
       "\n",
       "     n00_original_curie n00_unified_curie n01_original_curie  \\\n",
       "0         NCBIGene:7450     NCBIGene:7450      MONDO:0019565   \n",
       "1         NCBIGene:7450     NCBIGene:7450      MONDO:0019565   \n",
       "2         NCBIGene:7450     NCBIGene:7450      MONDO:0019565   \n",
       "3         NCBIGene:7450     NCBIGene:7450      MONDO:0019565   \n",
       "4         NCBIGene:7450     NCBIGene:7450      MONDO:0019565   \n",
       "...                 ...               ...                ...   \n",
       "4265     NCBIGene:57101    NCBIGene:57101      MONDO:0019565   \n",
       "4266     NCBIGene:57101    NCBIGene:57101      MONDO:0019565   \n",
       "4267     NCBIGene:57101    NCBIGene:57101      MONDO:0019565   \n",
       "4268     NCBIGene:57101    NCBIGene:57101      MONDO:0019565   \n",
       "4269     NCBIGene:57101    NCBIGene:57101      MONDO:0019565   \n",
       "\n",
       "     n01_unified_curie          trapi_result_curie_set  crossjoin_key  \\\n",
       "0         MESH:C531844   {MESH:C531844, NCBIGene:7450}              0   \n",
       "1         MESH:C531844   {MESH:C531844, NCBIGene:7450}              0   \n",
       "2         MESH:C531844   {MESH:C531844, NCBIGene:7450}              0   \n",
       "3         MESH:C531844   {MESH:C531844, NCBIGene:7450}              0   \n",
       "4         MESH:C531844   {MESH:C531844, NCBIGene:7450}              0   \n",
       "...                ...                             ...            ...   \n",
       "4265      MESH:C531844  {MESH:C531844, NCBIGene:57101}              0   \n",
       "4266      MESH:C531844  {MESH:C531844, NCBIGene:57101}              0   \n",
       "4267      MESH:C531844  {MESH:C531844, NCBIGene:57101}              0   \n",
       "4268      MESH:C531844  {MESH:C531844, NCBIGene:57101}              0   \n",
       "4269      MESH:C531844  {MESH:C531844, NCBIGene:57101}              0   \n",
       "\n",
       "                                       figure_id  ...  \\\n",
       "0               PMC1262766__1743-7075-2-22-1.jpg  ...   \n",
       "1               PMC1489360__zam0070669260006.jpg  ...   \n",
       "2                  PMC1572915__134-0704215f2.jpg  ...   \n",
       "3                    PMC181284__pp0838025001.jpg  ...   \n",
       "4               PMC1932822__zam0140779890003.jpg  ...   \n",
       "...                                          ...  ...   \n",
       "4265           PMC6797170__pone.0223840.g007.jpg  ...   \n",
       "4266        PMC7238249__plants-09-00422-g002.jpg  ...   \n",
       "4267            PMC7242693__zbc9992023450012.jpg  ...   \n",
       "4268  PMC7599810__antioxidants-09-00936-g002.jpg  ...   \n",
       "4269    PMC7608132__41401_2020_389_Fig5_HTML.jpg  ...   \n",
       "\n",
       "                                             figure_url  \\\n",
       "0     https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "1     https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "2     https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "3     https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "4     https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "...                                                 ...   \n",
       "4265  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "4266  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...   \n",
       "4267  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...   \n",
       "4268  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...   \n",
       "4269  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7...   \n",
       "\n",
       "                                           figure_title figure_curie_count  \\\n",
       "0           The kynuric pathway of melatonin metabolism                  1   \n",
       "1           Postulated biosynthetic pathway of ABA in C                  1   \n",
       "2     Biosynthesis of lyso-PAF and PAF via the remod...                  1   \n",
       "3     A, Metabolic pathway for the synthesis of MAs,...                  1   \n",
       "4     Six distinct ring-cleaving pathways present in...                  1   \n",
       "...                                                 ...                ...   \n",
       "4265  Identification of protein changes in the blood...                 20   \n",
       "4266  Proposed model of the regulatory functions of ...                  4   \n",
       "4267  Synthesis of bacteriochlorophylls c, d, e, and...                  1   \n",
       "4268                    NETosis activated by SARS-CoV-2                  8   \n",
       "4269   Proposed metabolic pathway of alflutinib in HLMs                  1   \n",
       "\n",
       "      common_curie_set yes_trapi_result_yes_figure  \\\n",
       "0                   {}                           0   \n",
       "1                   {}                           0   \n",
       "2                   {}                           0   \n",
       "3                   {}                           0   \n",
       "4                   {}                           0   \n",
       "...                ...                         ...   \n",
       "4265                {}                           0   \n",
       "4266                {}                           0   \n",
       "4267                {}                           0   \n",
       "4268                {}                           0   \n",
       "4269                {}                           0   \n",
       "\n",
       "      yes_trapi_result_no_figure  no_trapi_result_yes_figure  \\\n",
       "0                              2                           1   \n",
       "1                              2                           1   \n",
       "2                              2                           1   \n",
       "3                              2                           1   \n",
       "4                              2                           1   \n",
       "...                          ...                         ...   \n",
       "4265                           2                          20   \n",
       "4266                           2                           4   \n",
       "4267                           2                           1   \n",
       "4268                           2                           8   \n",
       "4269                           2                           1   \n",
       "\n",
       "      no_trapi_result_no_figure  p_value         trapi_result_curies_key  \n",
       "0                         15680      1.0   (MESH:C531844, NCBIGene:7450)  \n",
       "1                         15680      1.0   (MESH:C531844, NCBIGene:7450)  \n",
       "2                         15680      1.0   (MESH:C531844, NCBIGene:7450)  \n",
       "3                         15680      1.0   (MESH:C531844, NCBIGene:7450)  \n",
       "4                         15680      1.0   (MESH:C531844, NCBIGene:7450)  \n",
       "...                         ...      ...                             ...  \n",
       "4265                      15661      1.0  (MESH:C531844, NCBIGene:57101)  \n",
       "4266                      15677      1.0  (MESH:C531844, NCBIGene:57101)  \n",
       "4267                      15680      1.0  (MESH:C531844, NCBIGene:57101)  \n",
       "4268                      15673      1.0  (MESH:C531844, NCBIGene:57101)  \n",
       "4269                      15680      1.0  (MESH:C531844, NCBIGene:57101)  \n",
       "\n",
       "[4270 rows x 21 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "figures_df = relevant_figure_curie_df[\n",
    "    relevant_figure_curie_df[\"figure_id\"].isin(set(cluster_figure_df[\"figure_id\"]))\n",
    "][[\"figure_id\", \"curie\"]].groupby(\"figure_id\").agg(\n",
    "    create_set\n",
    ").rename(columns={\n",
    "    \"curie\": \"figure_curie_set\"\n",
    "}).join(\n",
    "    figure_metadata_df\n",
    ").reset_index()\n",
    "\n",
    "figures_df[\"figure_curie_count\"] = figures_df[\"figure_curie_set\"].map(len)\n",
    "\n",
    "figures_df['crossjoin_key'] = 0\n",
    "trapi_results_df['crossjoin_key'] = 0\n",
    "\n",
    "trapi_results_with_figures_df = trapi_results_df.merge(\n",
    "    figures_df, on='crossjoin_key', how='outer'\n",
    ").rename(columns={\n",
    "    \"unified_curie_set\": \"trapi_result_curie_set\"\n",
    "})\n",
    "\n",
    "trapi_results_with_figures_df[\"common_curie_set\"] = trapi_results_with_figures_df.apply(\n",
    "    lambda r: r[\"trapi_result_curie_set\"] & r[\"figure_curie_set\"], axis=1\n",
    ")\n",
    "trapi_results_with_figures_df\n",
    "# number of CURIEs in both TRAPI result and figure\n",
    "# trapi_curies_in_figure\n",
    "# x\n",
    "# Selected & Having the property\n",
    "trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"] = (\n",
    "    trapi_results_with_figures_df[\"common_curie_set\"].map(len)\n",
    ")\n",
    "\n",
    "# number of CURIEs in TRAPI result but not in figure\n",
    "# q_node_id_count - trapi_curies_in_figure\n",
    "# n - x\n",
    "# Selected & Not Having the property\n",
    "trapi_results_with_figures_df[\"yes_trapi_result_no_figure\"] = (\n",
    "    len(q_node_ids) - trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# number of CURIEs not in TRAPI result but in figure\n",
    "# curies_in_figure - trapi_curies_in_figure\n",
    "# N - x\n",
    "# Not Selected & Having the property\n",
    "trapi_results_with_figures_df[\"no_trapi_result_yes_figure\"] = (\n",
    "    trapi_results_with_figures_df[\"figure_curie_set\"].map(len) - \n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# number of CURIEs not in TRAPI result and not in figure,\n",
    "# ie., number of unique CURIEs only in other figures\n",
    "# 28735 - curies_in_figure - q_node_id_count + trapi_curies_in_figure\n",
    "# M - (n + N) + x\n",
    "# Not Selected & Not Having the property\n",
    "trapi_results_with_figures_df[\"no_trapi_result_no_figure\"] = (\n",
    "    all_relevant_curie_count -\n",
    "    trapi_results_with_figures_df[\"figure_curie_set\"].map(len) -\n",
    "    len(q_node_ids) +\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"]\n",
    ")\n",
    "\n",
    "# see https://stackoverflow.com/a/58661068/5354298\n",
    "_, _, twosided = pvalue_npy(\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_yes_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"yes_trapi_result_no_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"no_trapi_result_yes_figure\"].to_numpy('uint'),\n",
    "    trapi_results_with_figures_df[\"no_trapi_result_no_figure\"].to_numpy('uint'),\n",
    ")\n",
    "\n",
    "trapi_results_with_figures_df[\"p_value\"] = pd.Series(twosided, index=trapi_results_with_figures_df.index)\n",
    "trapi_results_with_figures_df[\"p_value\"].sort_values()\n",
    "\n",
    "# Needed in order to use the TRAPI result CURIEs as a key. A Python Set is not hashable.\n",
    "trapi_results_with_figures_df[\"trapi_result_curies_key\"] = trapi_results_with_figures_df[\n",
    "    \"trapi_result_curie_set\"\n",
    "].map(\n",
    "    lambda x: tuple(sorted(x))\n",
    ")\n",
    "\n",
    "trapi_results_with_figures_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([  99.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n",
       "        4171.]),\n",
       " array([1.27526621e-04, 1.00114774e-01, 2.00102021e-01, 3.00089269e-01,\n",
       "        4.00076516e-01, 5.00063763e-01, 6.00051011e-01, 7.00038258e-01,\n",
       "        8.00025505e-01, 9.00012753e-01, 1.00000000e+00]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUAklEQVR4nO3df4xl5X3f8ffHyw+ntWPATBDd3XZpsla6dhWMpkDkqnWggWVdeYnqWKAmbBDqpilUTmulgfQPHDtIRq1Na8kmWZetFysxpk5SVoSUbgELuSo/hoAxC6FMAIfdYnbiBRILmRby7R/3Wesaz+zc2blzh8nzfklXc873POec52GWzz1zzrn3pKqQJPXhLavdAUnS5Bj6ktQRQ1+SOmLoS1JHDH1J6shxq92Bozn11FNr06ZNq90NSVpTHnrooT+rqqn5lr2pQ3/Tpk3MzMysdjckaU1J8s2Flnl6R5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOvKm/kSuJK22TVf/wars99lPfmBFtuuRviR1xNCXpI4Y+pLUEUNfkjoycugnWZfk4SS3t/kzktyfZDbJl5Oc0OontvnZtnzT0DauafUnk1w49tFIko5qKUf6HwGeGJq/Hrihqn4MeBG4otWvAF5s9RtaO5JsAS4B3g1sBT6XZN3yui9JWoqRQj/JBuADwH9q8wHOA77SmuwBLm7T29s8bfn5rf124JaqerWqngFmgbPHMAZJ0ohGPdL/D8C/Af6yzb8TeKmqXmvzB4D1bXo98BxAW/5ya/+9+jzrfE+SnUlmkszMzc2NPhJJ0qIWDf0k/xg4VFUPTaA/VNWuqpququmpqXkf8ShJOkajfCL3fcAHk2wD3gr8MPAfgZOSHNeO5jcAB1v7g8BG4ECS44B3AN8eqh8xvI4kaQIWPdKvqmuqakNVbWJwIfbuqvqnwD3Ah1qzHcBtbXpvm6ctv7uqqtUvaXf3nAFsBh4Y20gkSYtaznfv/CpwS5LfAB4Gbmr1m4AvJpkFDjN4o6Cq9ie5FXgceA24sqpeX8b+JUlLtKTQr6qvAl9t008zz903VfVd4GcXWP864LqldlKSNB5+IleSOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1JFRHoz+1iQPJPl6kv1Jfr3Vv5DkmSSPtNeZrZ4kn0kym+TRJGcNbWtHkqfaa8cCu5QkrZBRnpz1KnBeVX0nyfHA15L8YVv2K1X1lTe0v4jB8283A+cANwLnJDkFuBaYBgp4KMneqnpxHAORJC1ulAejV1V9p80e3151lFW2Aze39e4DTkpyOnAhsK+qDreg3wdsXV73JUlLMdI5/STrkjwCHGIQ3Pe3Rde1Uzg3JDmx1dYDzw2tfqDVFqq/cV87k8wkmZmbm1vaaCRJRzVS6FfV61V1JrABODvJe4BrgB8H/h5wCvCr4+hQVe2qqumqmp6amhrHJiVJzZLu3qmql4B7gK1V9Xw7hfMq8J+Bs1uzg8DGodU2tNpCdUnShIxy985UkpPa9A8BPw38cTtPT5IAFwOPtVX2Ape1u3jOBV6uqueBO4ELkpyc5GTgglaTJE3IKHfvnA7sSbKOwZvErVV1e5K7k0wBAR4B/nlrfwewDZgFXgEuB6iqw0k+ATzY2n28qg6PbSSSpEUtGvpV9Sjw3nnq5y3QvoArF1i2G9i9xD5KksbET+RKUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjoyyuMS35rkgSRfT7I/ya+3+hlJ7k8ym+TLSU5o9RPb/GxbvmloW9e0+pNJLlyxUUmS5jXKkf6rwHlV9RPAmcDW9uzb64EbqurHgBeBK1r7K4AXW/2G1o4kW4BLgHcDW4HPtUcwSpImZNHQr4HvtNnj26uA84CvtPoeBg9HB9je5mnLz28PT98O3FJVr1bVMwyeoXv2OAYhSRrNSOf0k6xL8ghwCNgH/AnwUlW91pocANa36fXAcwBt+cvAO4fr86wzvK+dSWaSzMzNzS15QJKkhY0U+lX1elWdCWxgcHT+4yvVoaraVVXTVTU9NTW1UruRpC4t6e6dqnoJuAf4SeCkJMe1RRuAg236ILARoC1/B/Dt4fo860iSJmCUu3emkpzUpn8I+GngCQbh/6HWbAdwW5ve2+Zpy++uqmr1S9rdPWcAm4EHxjQOSdIIjlu8CacDe9qdNm8Bbq2q25M8DtyS5DeAh4GbWvubgC8mmQUOM7hjh6ran+RW4HHgNeDKqnp9vMORJB3NoqFfVY8C752n/jTz3H1TVd8FfnaBbV0HXLf0bkqSxsFP5EpSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOjLK4xI3JrknyeNJ9if5SKt/LMnBJI+017ahda5JMpvkySQXDtW3ttpskqtXZkiSpIWM8rjE14CPVtUfJXk78FCSfW3ZDVX174cbJ9nC4BGJ7wb+BvA/kryrLf4sg2fsHgAeTLK3qh4fx0AkSYsb5XGJzwPPt+m/SPIEsP4oq2wHbqmqV4Fn2rNyjzxWcbY9ZpEkt7S2hr4kTciSzukn2cTgebn3t9JVSR5NsjvJya22HnhuaLUDrbZQ/Y372JlkJsnM3NzcUronSVrEyKGf5G3A7wK/XFV/DtwI/ChwJoO/BD41jg5V1a6qmq6q6ampqXFsUpLUjHJOnyTHMwj8366q3wOoqheGln8euL3NHgQ2Dq2+odU4Sl2SNAGj3L0T4Cbgiar69FD99KFmPwM81qb3ApckOTHJGcBm4AHgQWBzkjOSnMDgYu/e8QxDkjSKUY703wf8PPCNJI+02q8BlyY5EyjgWeAXAapqf5JbGVygfQ24sqpeB0hyFXAnsA7YXVX7xzYSSdKiRrl752tA5ll0x1HWuQ64bp76HUdbT5K0svxEriR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0Z5clZG5Pck+TxJPuTfKTVT0myL8lT7efJrZ4kn0ky2x6aftbQtna09k8l2bFyw5IkzWeUI/3XgI9W1RbgXODKJFuAq4G7qmozcFebB7iIwSMSNwM7GTxAnSSnANcC5wBnA9ceeaOQJE3GoqFfVc9X1R+16b8AngDWA9uBPa3ZHuDiNr0duLkG7gNOas/TvRDYV1WHq+pFYB+wdZyDkSQd3ZLO6SfZBLwXuB84raqeb4u+BZzWptcDzw2tdqDVFqpLkiZk5NBP8jbgd4Ffrqo/H15WVcXgAenLlmRnkpkkM3Nzc+PYpCSpGSn0kxzPIPB/u6p+r5VfaKdtaD8PtfpBYOPQ6htabaH696mqXVU1XVXTU1NTSxmLJGkRo9y9E+Am4Imq+vTQor3AkTtwdgC3DdUva3fxnAu83E4D3QlckOTkdgH3glaTJE3IcSO0eR/w88A3kjzSar8GfBK4NckVwDeBD7dldwDbgFngFeBygKo6nOQTwIOt3cer6vA4BiFJGs2ioV9VXwOywOLz52lfwJULbGs3sHspHZQkjY+fyJWkjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdWSUxyXuTnIoyWNDtY8lOZjkkfbaNrTsmiSzSZ5McuFQfWurzSa5evxDkSQtZpQj/S8AW+ep31BVZ7bXHQBJtgCXAO9u63wuybok64DPAhcBW4BLW1tJ0gSN8rjEe5NsGnF724FbqupV4Jkks8DZbdlsVT0NkOSW1vbxpXdZknSslnNO/6okj7bTPye32nrguaE2B1ptofoPSLIzyUySmbm5uWV0T5L0Rsca+jcCPwqcCTwPfGpcHaqqXVU1XVXTU1NT49qsJIkRTu/Mp6peODKd5PPA7W32ILBxqOmGVuModUnShBzTkX6S04dmfwY4cmfPXuCSJCcmOQPYDDwAPAhsTnJGkhMYXOzde+zdliQdi0WP9JN8CXg/cGqSA8C1wPuTnAkU8CzwiwBVtT/JrQwu0L4GXFlVr7ftXAXcCawDdlfV/nEPRpJ0dKPcvXPpPOWbjtL+OuC6eep3AHcsqXeSpLHyE7mS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4sGvpJdic5lOSxodopSfYlear9PLnVk+QzSWaTPJrkrKF1drT2TyXZsTLDkSQdzShH+l8Atr6hdjVwV1VtBu5q8wAXMXgu7mZgJ3AjDN4kGDxm8RzgbODaI28UkqTJWTT0q+pe4PAbytuBPW16D3DxUP3mGrgPOKk9RP1CYF9VHa6qF4F9/OAbiSRphR3rOf3Tqur5Nv0t4LQ2vR54bqjdgVZbqP4DkuxMMpNkZm5u7hi7J0maz7Iv5FZVATWGvhzZ3q6qmq6q6ampqXFtVpLEsYf+C+20De3noVY/CGwcareh1RaqS5Im6FhDfy9w5A6cHcBtQ/XL2l085wIvt9NAdwIXJDm5XcC9oNUkSRN03GINknwJeD9wapIDDO7C+SRwa5IrgG8CH27N7wC2AbPAK8DlAFV1OMkngAdbu49X1RsvDkuSVtiioV9Vly6w6Px52hZw5QLb2Q3sXlLvJElj5SdyJakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdWVboJ3k2yTeSPJJkptVOSbIvyVPt58mtniSfSTKb5NEkZ41jAJKk0Y3jSP+nqurMqppu81cDd1XVZuCuNg9wEbC5vXYCN45h35KkJViJ0zvbgT1teg9w8VD95hq4DzgpyekrsH9J0gKWG/oF/PckDyXZ2WqnVdXzbfpbwGltej3w3NC6B1rt+yTZmWQmyczc3NwyuydJGrbog9EX8fer6mCSHwH2Jfnj4YVVVUlqKRusql3ALoDp6eklrStJOrplHelX1cH28xDw+8DZwAtHTtu0n4da84PAxqHVN7SaJGlCjjn0k/z1JG8/Mg1cADwG7AV2tGY7gNva9F7gsnYXz7nAy0OngSRJE7Cc0zunAb+f5Mh2fqeq/luSB4Fbk1wBfBP4cGt/B7ANmAVeAS5fxr4lScfgmEO/qp4GfmKe+reB8+epF3Dlse5PkrR8fiJXkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHVkuY9LfFPbdPUfrMp+n/3kB1Zlv5K0GI/0JakjEw/9JFuTPJlkNsnVk96/JPVsoqGfZB3wWeAiYAtwaZItk+yDJPVs0kf6ZwOzVfV0Vf1f4BZg+4T7IEndmvSF3PXAc0PzB4Bzhhsk2QnsbLPfSfLkMvZ3KvBny1j/mOT6Se/x+6zKmFdRb+MFx9yFXL+sMf+thRa86e7eqapdwK5xbCvJTFVNj2Nba0VvY+5tvOCYe7FSY5706Z2DwMah+Q2tJkmagEmH/oPA5iRnJDkBuATYO+E+SFK3Jnp6p6peS3IVcCewDthdVftXcJdjOU20xvQ25t7GC465Fysy5lTVSmxXkvQm5CdyJakjhr4kdWTNh/5iX+uQ5MQkX27L70+yaRW6OVYjjPlfJ3k8yaNJ7kqy4D27a8WoX9+R5J8kqSRr/va+Ucac5MPtd70/ye9Muo/jNsK/7b+Z5J4kD7d/39tWo5/jkmR3kkNJHltgeZJ8pv33eDTJWcveaVWt2ReDi8F/Avxt4ATg68CWN7T5F8BvtulLgC+vdr8nMOafAv5am/6lHsbc2r0duBe4D5he7X5P4Pe8GXgYOLnN/8hq93sCY94F/FKb3gI8u9r9XuaY/wFwFvDYAsu3AX8IBDgXuH+5+1zrR/qjfK3DdmBPm/4KcH6STLCP47bomKvqnqp6pc3ex+DzEGvZqF/f8QngeuC7k+zcChllzP8M+GxVvQhQVYcm3MdxG2XMBfxwm34H8H8m2L+xq6p7gcNHabIduLkG7gNOSnL6cva51kN/vq91WL9Qm6p6DXgZeOdEercyRhnzsCsYHCmsZYuOuf3Zu7GqVuchCuM3yu/5XcC7kvzPJPcl2Tqx3q2MUcb8MeDnkhwA7gD+5WS6tmqW+v/7ot50X8Og8Unyc8A08A9Xuy8rKclbgE8Dv7DKXZm04xic4nk/g7/m7k3yd6vqpdXs1Aq7FPhCVX0qyU8CX0zynqr6y9Xu2Fqx1o/0R/lah++1SXIcgz8Jvz2R3q2Mkb7KIsk/Av4t8MGqenVCfVspi4357cB7gK8meZbBuc+9a/xi7ii/5wPA3qr6f1X1DPC/GbwJrFWjjPkK4FaAqvpfwFsZfBnbX1Vj/+qatR76o3ytw15gR5v+EHB3tSska9SiY07yXuC3GAT+Wj/PC4uMuaperqpTq2pTVW1icB3jg1U1szrdHYtR/m3/VwZH+SQ5lcHpnqcn2MdxG2XMfwqcD5Dk7zAI/bmJ9nKy9gKXtbt4zgVerqrnl7PBNX16pxb4WockHwdmqmovcBODPwFnGVwwuWT1erx8I4753wFvA/5Lu2b9p1X1wVXr9DKNOOa/UkYc853ABUkeB14HfqWq1uxfsSOO+aPA55P8KwYXdX9hLR/EJfkSgzfuU9t1imuB4wGq6jcZXLfYBswCrwCXL3ufa/i/lyRpidb66R1J0hIY+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakj/x95gempbxkjpQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(trapi_results_with_figures_df[\"p_value\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View TRAPI Results Clustered by PFOCR Figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n00</th>\n",
       "      <th>e00</th>\n",
       "      <th>n01</th>\n",
       "      <th>n00_original_curie</th>\n",
       "      <th>n00_unified_curie</th>\n",
       "      <th>n01_original_curie</th>\n",
       "      <th>n01_unified_curie</th>\n",
       "      <th>trapi_result_curie_set</th>\n",
       "      <th>crossjoin_key</th>\n",
       "      <th>figure_id</th>\n",
       "      <th>...</th>\n",
       "      <th>figure_url</th>\n",
       "      <th>figure_title</th>\n",
       "      <th>figure_curie_count</th>\n",
       "      <th>common_curie_set</th>\n",
       "      <th>yes_trapi_result_yes_figure</th>\n",
       "      <th>yes_trapi_result_no_figure</th>\n",
       "      <th>no_trapi_result_yes_figure</th>\n",
       "      <th>no_trapi_result_no_figure</th>\n",
       "      <th>p_value</th>\n",
       "      <th>trapi_result_curies_key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1586</th>\n",
       "      <td>COCH</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:1690}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1262766__1743-7075-2-22-1.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>The kynuric pathway of melatonin metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>{NCBIGene:1690}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15681</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>(MESH:C531844, NCBIGene:1690)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>COCH</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:1690}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1489360__zam0070669260006.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Postulated biosynthetic pathway of ABA in C</td>\n",
       "      <td>1</td>\n",
       "      <td>{NCBIGene:1690}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15681</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>(MESH:C531844, NCBIGene:1690)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>COCH</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:1690}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1572915__134-0704215f2.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Biosynthesis of lyso-PAF and PAF via the remod...</td>\n",
       "      <td>1</td>\n",
       "      <td>{NCBIGene:1690}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15681</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>(MESH:C531844, NCBIGene:1690)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1589</th>\n",
       "      <td>COCH</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:1690}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC181284__pp0838025001.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>A, Metabolic pathway for the synthesis of MAs,...</td>\n",
       "      <td>1</td>\n",
       "      <td>{NCBIGene:1690}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15681</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>(MESH:C531844, NCBIGene:1690)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>COCH</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>NCBIGene:1690</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:1690}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC1932822__zam0140779890003.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Six distinct ring-cleaving pathways present in...</td>\n",
       "      <td>1</td>\n",
       "      <td>{NCBIGene:1690}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15681</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>(MESH:C531844, NCBIGene:1690)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>F2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:2147</td>\n",
       "      <td>NCBIGene:2147</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:2147}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6637639__12864_2019_5836_Fig1_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Interaction network of proteins underexpressed...</td>\n",
       "      <td>81</td>\n",
       "      <td>{NCBIGene:2147}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>15601</td>\n",
       "      <td>0.010303</td>\n",
       "      <td>(MESH:C531844, NCBIGene:2147)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2249</th>\n",
       "      <td>ITIH1</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:3697</td>\n",
       "      <td>NCBIGene:3697</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:3697}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6637639__12864_2019_5836_Fig1_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Interaction network of proteins underexpressed...</td>\n",
       "      <td>81</td>\n",
       "      <td>{NCBIGene:3697}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>15601</td>\n",
       "      <td>0.010303</td>\n",
       "      <td>(MESH:C531844, NCBIGene:3697)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>SERPINC1</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:462</td>\n",
       "      <td>NCBIGene:462</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:462}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6637639__12864_2019_5836_Fig1_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Interaction network of proteins underexpressed...</td>\n",
       "      <td>81</td>\n",
       "      <td>{NCBIGene:462}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>15601</td>\n",
       "      <td>0.010303</td>\n",
       "      <td>(MESH:C531844, NCBIGene:462)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>PLG</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:5340</td>\n",
       "      <td>NCBIGene:5340</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:5340}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6637639__12864_2019_5836_Fig1_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Interaction network of proteins underexpressed...</td>\n",
       "      <td>81</td>\n",
       "      <td>{NCBIGene:5340}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>15601</td>\n",
       "      <td>0.010303</td>\n",
       "      <td>(MESH:C531844, NCBIGene:5340)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>CRISPLD2</td>\n",
       "      <td>gene_associated_with_condition</td>\n",
       "      <td>hereditary von Willebrand disease</td>\n",
       "      <td>NCBIGene:83716</td>\n",
       "      <td>NCBIGene:83716</td>\n",
       "      <td>MONDO:0019565</td>\n",
       "      <td>MESH:C531844</td>\n",
       "      <td>{MESH:C531844, NCBIGene:83716}</td>\n",
       "      <td>0</td>\n",
       "      <td>PMC6637639__12864_2019_5836_Fig1_HTML.jpg</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>Interaction network of proteins underexpressed...</td>\n",
       "      <td>81</td>\n",
       "      <td>{NCBIGene:83716}</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>15601</td>\n",
       "      <td>0.010303</td>\n",
       "      <td>(MESH:C531844, NCBIGene:83716)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           n00                             e00  \\\n",
       "1586      COCH  gene_associated_with_condition   \n",
       "1587      COCH  gene_associated_with_condition   \n",
       "1588      COCH  gene_associated_with_condition   \n",
       "1589      COCH  gene_associated_with_condition   \n",
       "1590      COCH  gene_associated_with_condition   \n",
       "...        ...                             ...   \n",
       "541         F2  gene_associated_with_condition   \n",
       "2249     ITIH1  gene_associated_with_condition   \n",
       "724   SERPINC1  gene_associated_with_condition   \n",
       "602        PLG  gene_associated_with_condition   \n",
       "3652  CRISPLD2  gene_associated_with_condition   \n",
       "\n",
       "                                    n01 n00_original_curie n00_unified_curie  \\\n",
       "1586  hereditary von Willebrand disease      NCBIGene:1690     NCBIGene:1690   \n",
       "1587  hereditary von Willebrand disease      NCBIGene:1690     NCBIGene:1690   \n",
       "1588  hereditary von Willebrand disease      NCBIGene:1690     NCBIGene:1690   \n",
       "1589  hereditary von Willebrand disease      NCBIGene:1690     NCBIGene:1690   \n",
       "1590  hereditary von Willebrand disease      NCBIGene:1690     NCBIGene:1690   \n",
       "...                                 ...                ...               ...   \n",
       "541   hereditary von Willebrand disease      NCBIGene:2147     NCBIGene:2147   \n",
       "2249  hereditary von Willebrand disease      NCBIGene:3697     NCBIGene:3697   \n",
       "724   hereditary von Willebrand disease       NCBIGene:462      NCBIGene:462   \n",
       "602   hereditary von Willebrand disease      NCBIGene:5340     NCBIGene:5340   \n",
       "3652  hereditary von Willebrand disease     NCBIGene:83716    NCBIGene:83716   \n",
       "\n",
       "     n01_original_curie n01_unified_curie          trapi_result_curie_set  \\\n",
       "1586      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:1690}   \n",
       "1587      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:1690}   \n",
       "1588      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:1690}   \n",
       "1589      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:1690}   \n",
       "1590      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:1690}   \n",
       "...                 ...               ...                             ...   \n",
       "541       MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:2147}   \n",
       "2249      MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:3697}   \n",
       "724       MONDO:0019565      MESH:C531844    {MESH:C531844, NCBIGene:462}   \n",
       "602       MONDO:0019565      MESH:C531844   {MESH:C531844, NCBIGene:5340}   \n",
       "3652      MONDO:0019565      MESH:C531844  {MESH:C531844, NCBIGene:83716}   \n",
       "\n",
       "      crossjoin_key                                  figure_id  ...  \\\n",
       "1586              0           PMC1262766__1743-7075-2-22-1.jpg  ...   \n",
       "1587              0           PMC1489360__zam0070669260006.jpg  ...   \n",
       "1588              0              PMC1572915__134-0704215f2.jpg  ...   \n",
       "1589              0                PMC181284__pp0838025001.jpg  ...   \n",
       "1590              0           PMC1932822__zam0140779890003.jpg  ...   \n",
       "...             ...                                        ...  ...   \n",
       "541               0  PMC6637639__12864_2019_5836_Fig1_HTML.jpg  ...   \n",
       "2249              0  PMC6637639__12864_2019_5836_Fig1_HTML.jpg  ...   \n",
       "724               0  PMC6637639__12864_2019_5836_Fig1_HTML.jpg  ...   \n",
       "602               0  PMC6637639__12864_2019_5836_Fig1_HTML.jpg  ...   \n",
       "3652              0  PMC6637639__12864_2019_5836_Fig1_HTML.jpg  ...   \n",
       "\n",
       "                                             figure_url  \\\n",
       "1586  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "1587  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "1588  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "1589  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "1590  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...   \n",
       "...                                                 ...   \n",
       "541   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "2249  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "724   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "602   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "3652  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   \n",
       "\n",
       "                                           figure_title figure_curie_count  \\\n",
       "1586        The kynuric pathway of melatonin metabolism                  1   \n",
       "1587        Postulated biosynthetic pathway of ABA in C                  1   \n",
       "1588  Biosynthesis of lyso-PAF and PAF via the remod...                  1   \n",
       "1589  A, Metabolic pathway for the synthesis of MAs,...                  1   \n",
       "1590  Six distinct ring-cleaving pathways present in...                  1   \n",
       "...                                                 ...                ...   \n",
       "541   Interaction network of proteins underexpressed...                 81   \n",
       "2249  Interaction network of proteins underexpressed...                 81   \n",
       "724   Interaction network of proteins underexpressed...                 81   \n",
       "602   Interaction network of proteins underexpressed...                 81   \n",
       "3652  Interaction network of proteins underexpressed...                 81   \n",
       "\n",
       "      common_curie_set yes_trapi_result_yes_figure  \\\n",
       "1586   {NCBIGene:1690}                           1   \n",
       "1587   {NCBIGene:1690}                           1   \n",
       "1588   {NCBIGene:1690}                           1   \n",
       "1589   {NCBIGene:1690}                           1   \n",
       "1590   {NCBIGene:1690}                           1   \n",
       "...                ...                         ...   \n",
       "541    {NCBIGene:2147}                           1   \n",
       "2249   {NCBIGene:3697}                           1   \n",
       "724     {NCBIGene:462}                           1   \n",
       "602    {NCBIGene:5340}                           1   \n",
       "3652  {NCBIGene:83716}                           1   \n",
       "\n",
       "      yes_trapi_result_no_figure  no_trapi_result_yes_figure  \\\n",
       "1586                           1                           0   \n",
       "1587                           1                           0   \n",
       "1588                           1                           0   \n",
       "1589                           1                           0   \n",
       "1590                           1                           0   \n",
       "...                          ...                         ...   \n",
       "541                            1                          80   \n",
       "2249                           1                          80   \n",
       "724                            1                          80   \n",
       "602                            1                          80   \n",
       "3652                           1                          80   \n",
       "\n",
       "      no_trapi_result_no_figure   p_value         trapi_result_curies_key  \n",
       "1586                      15681  0.000128   (MESH:C531844, NCBIGene:1690)  \n",
       "1587                      15681  0.000128   (MESH:C531844, NCBIGene:1690)  \n",
       "1588                      15681  0.000128   (MESH:C531844, NCBIGene:1690)  \n",
       "1589                      15681  0.000128   (MESH:C531844, NCBIGene:1690)  \n",
       "1590                      15681  0.000128   (MESH:C531844, NCBIGene:1690)  \n",
       "...                         ...       ...                             ...  \n",
       "541                       15601  0.010303   (MESH:C531844, NCBIGene:2147)  \n",
       "2249                      15601  0.010303   (MESH:C531844, NCBIGene:3697)  \n",
       "724                       15601  0.010303    (MESH:C531844, NCBIGene:462)  \n",
       "602                       15601  0.010303   (MESH:C531844, NCBIGene:5340)  \n",
       "3652                      15601  0.010303  (MESH:C531844, NCBIGene:83716)  \n",
       "\n",
       "[99 rows x 21 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all\n",
    "cooccurrence_df = trapi_results_with_figures_df[\n",
    "    trapi_results_with_figures_df[\"p_value\"] < 0.5\n",
    "].sort_values(\n",
    "    [\"p_value\", \"figure_id\", \"trapi_result_curies_key\"]\n",
    ")\n",
    "\n",
    "### first and any other query node\n",
    "##q_node_id_combination = (q_node_ids[0],)\n",
    "\n",
    "## first and second query nodes\n",
    "##q_node_id_combination = (q_node_ids[0], q_node_ids[1])\n",
    "\n",
    "### first and last query nodes\n",
    "##q_node_id_combination = (q_node_ids[0], q_node_ids[-1])\n",
    "\n",
    "### second and last query nodes\n",
    "##q_node_id_combination = (q_node_ids[1], q_node_ids[-1])\n",
    "\n",
    "### every query node with 'ids' specified\n",
    "##q_node_id_combination = query_nodes_with_ids\n",
    "\n",
    "#print(f'{\" & \".join(q_node_id_combination)}')\n",
    "#s = pd.Series([True] * len(trapi_results_with_figures_df), index=trapi_results_with_figures_df.index)\n",
    "#for q_node_id in q_node_id_combination:\n",
    "#    s = s & trapi_results_with_figures_df.apply(\n",
    "#        lambda r: r[f'{q_node_id}_unified_curie'] in r[\"common_curie_set\"], axis=1\n",
    "#    )\n",
    "#cooccurrence_df = trapi_results_with_figures_df[s == True].sort_values(\n",
    "#    \"p_value\"\n",
    "#)\n",
    "\n",
    "cooccurrence_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1262766/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1262766/bin/1743-7075-2-22-1.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            The kynuric pathway of melatonin metabolism\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1489360/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1489360/bin/zam0070669260006.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            Postulated biosynthetic pathway of ABA in C\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1572915/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1572915/bin/134-0704215f2.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            Biosynthesis of lyso-PAF and PAF via the remodelling pathway\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC181284/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC181284/bin/pp0838025001.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            A, Metabolic pathway for the synthesis of MAs, based on existing data\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1932822/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1932822/bin/zam0140779890003.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            Six distinct ring-cleaving pathways present in three Roseobacter genomes\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2078203/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2078203/bin/nihms33135f4.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            The Ferric-Hydroperoxo Complex of Chloroperoxidase\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:55576, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">STAB2</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2576650/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2576650/bin/zjb0220882380006.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            Model illustrating the RstA-mediated regulatory pathway that controls Fur activity in Salmonella enterica\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:55576, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">STAB2</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2635004/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2635004/bin/psb0103_0089_fig001.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            Tryptophan-dependent biosynthetic pathway of IAA in plants and of melatonin in mammals\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:1690, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">COCH</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2788834/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2788834/bin/zbc0460991910010.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            A putative model for the mode of action of 5201F\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:821, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">CANX</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
       "    <figure style=\"margin: 5px !important;\">\n",
       "        <a target=\"_blank\" href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2812603/\">\n",
       "          <img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2812603/bin/nihms135155f1.jpg\" style=\"max-height: 500px; max-width: 100%;\">\n",
       "        </a>\n",
       "        <br>\n",
       "        <caption>\n",
       "            The Mycobacterium tuberculosis Cytochrome P450 System\n",
       "        </caption>\n",
       "    </figure>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<div>\n",
       "    <span style=\"font-weight: bold; font-size: small;\">\n",
       "        TRAPI result NCBIGene:55576, MESH:C531844\n",
       "    </span> \n",
       "    (p-value: 1.3e-04)\n",
       "    <br>\n",
       "    \n",
       "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">STAB2</span>\n",
       " - gene_associated_with_condition - hereditary von Willebrand disease</span>\n",
       "</div>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "...\n",
      "\n",
      "\n",
      "plus 51 more figures, each with TRAPI results\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "from IPython.core.display import HTML \n",
    "\n",
    "\n",
    "outer_limit = 10\n",
    "inner_limit = 10\n",
    "\n",
    "count = 0\n",
    "for figure_id, one_figure_id_df in cooccurrence_df.groupby(\"figure_id\", sort=False):\n",
    "    [pmc,filename] = figure_id.split(\"__\")\n",
    "    paper_url = f\"https://www.ncbi.nlm.nih.gov/pmc/articles/{pmc}/\"\n",
    "    figure_url = f\"https://www.ncbi.nlm.nih.gov/pmc/articles/{pmc}/bin/{filename}\"\n",
    "    figure_title = list(set(one_figure_id_df[\"figure_title\"]))[0]\n",
    "\n",
    "    display(HTML(data=f'''\n",
    "<div style=\"border: 1px solid slategray; padding: 6px;\">\n",
    "    <figure style=\"margin: 5px !important;\">\n",
    "        <a target=\"_blank\" href=\"{paper_url}\">\n",
    "          <img src=\"{figure_url}\" style=\"max-height: 500px; max-width: 100%;\">\n",
    "        </a>\n",
    "        <br>\n",
    "        <caption>\n",
    "            {figure_title}\n",
    "        </caption>\n",
    "    </figure>\n",
    "</div>\n",
    "'''))\n",
    "\n",
    "    trapi_results_row_sets = []\n",
    "    \n",
    "    for trapi_result_curie_combo, one_trapi_result_curie_combo_df in one_figure_id_df[(\n",
    "        trapi_result_columns +\n",
    "        unified_curie_columns +\n",
    "        [\"trapi_result_curies_key\", \"common_curie_set\", \"p_value\"]\n",
    "    )].groupby(\"trapi_result_curies_key\", sort=False):\n",
    "        overlap_names = []\n",
    "        for curie in one_trapi_result_curie_combo_df[\"common_curie_set\"].tolist()[0]:\n",
    "            name = curie_to_name[curie]\n",
    "            overlap_names.append(name)\n",
    "\n",
    "        trapi_results_rows = []\n",
    "        for i,row in one_trapi_result_curie_combo_df.iterrows():\n",
    "            trapi_result_chunks = []\n",
    "            for trapi_result_column in trapi_result_columns:\n",
    "                row_value = row[trapi_result_column]\n",
    "                if row_value in overlap_names:\n",
    "                    trapi_result_chunk = f'''\n",
    "<span style=\"border: 1px orange solid; padding: 1px; border-radius: 4px;\">{row_value}</span>\n",
    "'''\n",
    "                else:\n",
    "                    trapi_result_chunk = f'{row_value}'\n",
    "                trapi_result_chunks.append(trapi_result_chunk)\n",
    "                \n",
    "            trapi_results_rows.append(\n",
    "                f'{\" - \".join(trapi_result_chunks)}</span>'\n",
    "            )\n",
    "        \n",
    "        ordered_trapi_result_curie_combo = []\n",
    "        for unified_curie_column in unified_curie_columns:\n",
    "            ordered_trapi_result_curie_combo.append(list(set(\n",
    "                one_trapi_result_curie_combo_df[unified_curie_column]\n",
    "            ))[0])\n",
    "        \n",
    "        p_value = list(set(one_trapi_result_curie_combo_df[\"p_value\"]))[0]\n",
    "        trapi_results_row_sets.append(f'''\n",
    "<div>\n",
    "    <span style=\"font-weight: bold; font-size: small;\">\n",
    "        TRAPI result {\", \".join(ordered_trapi_result_curie_combo)}\n",
    "    </span> \n",
    "    (p-value: {p_value:.1e})\n",
    "    <br>\n",
    "    {\"<br>\".join(trapi_results_rows)}\n",
    "</div>\n",
    "''')\n",
    "\n",
    "    trapi_results_row_set_count = len(trapi_results_row_sets)\n",
    "    display(HTML(data=\"<br>\".join(trapi_results_row_sets[:inner_limit])))\n",
    "    if trapi_results_row_set_count > inner_limit:\n",
    "        print(f\"...\")\n",
    "        print(\"\")\n",
    "        print(f\"plus {trapi_results_row_set_count - inner_limit} more TRAPI results\")\n",
    "    print(\"\")\n",
    "\n",
    "    count += 1\n",
    "    if count >= outer_limit:\n",
    "        print(\"...\")\n",
    "        print(\"\")\n",
    "        print(f'''\n",
    "plus {len(set(cooccurrence_df[\"figure_id\"])) - outer_limit} more figures, each with TRAPI results\n",
    "''')\n",
    "        print(\"\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WIP\n",
    "\n",
    "Is there a good way to merge figures when we get a tie for best p-value?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'p_value'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [23]\u001b[0m, in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcollections\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m defaultdict\n\u001b[0;32m----> 4\u001b[0m tied_figures_df \u001b[38;5;241m=\u001b[39m \u001b[43mfigures_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort_values\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mp_value\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mhead()\n\u001b[1;32m      6\u001b[0m all_cluster_figure_curies \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, row \u001b[38;5;129;01min\u001b[39;00m tied_figures_df\u001b[38;5;241m.\u001b[39miterrows():\n",
      "File \u001b[0;32m~/Library/Caches/pypoetry/virtualenvs/pathway-figure-ocr-python-Grd1lLXD-py3.9/lib/python3.10/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Library/Caches/pypoetry/virtualenvs/pathway-figure-ocr-python-Grd1lLXD-py3.9/lib/python3.10/site-packages/pandas/core/frame.py:6313\u001b[0m, in \u001b[0;36mDataFrame.sort_values\u001b[0;34m(self, by, axis, ascending, inplace, kind, na_position, ignore_index, key)\u001b[0m\n\u001b[1;32m   6309\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(by):\n\u001b[1;32m   6310\u001b[0m     \u001b[38;5;66;03m# len(by) == 1\u001b[39;00m\n\u001b[1;32m   6312\u001b[0m     by \u001b[38;5;241m=\u001b[39m by[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m-> 6313\u001b[0m     k \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label_or_level_values\u001b[49m\u001b[43m(\u001b[49m\u001b[43mby\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6315\u001b[0m     \u001b[38;5;66;03m# need to rewrap column in Series to apply key function\u001b[39;00m\n\u001b[1;32m   6316\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   6317\u001b[0m         \u001b[38;5;66;03m# error: Incompatible types in assignment (expression has type\u001b[39;00m\n\u001b[1;32m   6318\u001b[0m         \u001b[38;5;66;03m# \"Series\", variable has type \"ndarray\")\u001b[39;00m\n",
      "File \u001b[0;32m~/Library/Caches/pypoetry/virtualenvs/pathway-figure-ocr-python-Grd1lLXD-py3.9/lib/python3.10/site-packages/pandas/core/generic.py:1840\u001b[0m, in \u001b[0;36mNDFrame._get_label_or_level_values\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1838\u001b[0m     values \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes[axis]\u001b[38;5;241m.\u001b[39mget_level_values(key)\u001b[38;5;241m.\u001b[39m_values\n\u001b[1;32m   1839\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1840\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key)\n\u001b[1;32m   1842\u001b[0m \u001b[38;5;66;03m# Check for duplicates\u001b[39;00m\n\u001b[1;32m   1843\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m values\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "\u001b[0;31mKeyError\u001b[0m: 'p_value'"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "tied_figures_df = figures_df.sort_values(\"p_value\").head()\n",
    "\n",
    "all_cluster_figure_curies = set()\n",
    "for i, row in tied_figures_df.iterrows():\n",
    "    all_cluster_figure_curies |= row[\"curie_set\"]\n",
    "all_cluster_figure_curies_keys = sorted(all_cluster_figure_curies)\n",
    "\n",
    "cluster_figure_curie_to_figure_id = defaultdict(set)\n",
    "cluster_figure_curie_records = []\n",
    "for i, row in tied_figures_df.iterrows():\n",
    "    row_data = dict()\n",
    "    figure_id = row[\"figure_id\"]\n",
    "    row_data[\"figure_id\"] = figure_id\n",
    "    for curie in all_cluster_figure_curies_keys:\n",
    "        if curie in row[\"curie_set\"]:\n",
    "            row_data[curie] = 1\n",
    "        else:\n",
    "            row_data[curie] = 0\n",
    "    cluster_figure_curie_records.append(row_data)\n",
    "cluster_figure_curie_df = pd.DataFrame(cluster_figure_curie_records)\n",
    "cluster_figure_curie_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "linked = linkage(cluster_figure_curie_df[all_cluster_figure_curies_keys].to_numpy(), 'single')\n",
    "print(linked)\n",
    "\n",
    "labelList = cluster_figure_curie_df[\"figure_id\"].to_list()\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "linked = linkage(cluster_figure_curie_df[all_cluster_figure_curies_keys].to_numpy(), 'single')\n",
    "print(linked)\n",
    "\n",
    "labelList = cluster_figure_curie_df[\"figure_id\"].to_list()\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = np.array([[5,3],\n",
    "    [10,15],\n",
    "    [15,12],\n",
    "    [24,10],\n",
    "    [30,30],\n",
    "    [85,70],\n",
    "    [71,80],\n",
    "    [60,78],\n",
    "    [70,55],\n",
    "    [80,91],])\n",
    "\n",
    "linked = linkage(X, 'single')\n",
    "print(linked)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_figure_curie_df[all_cluster_figure_curies_keys].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_figure_curie_df[\"figure_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import jaccard\n",
    "\n",
    "\n",
    "\n",
    "print(jaccard([0, 0], [0, 1]))\n",
    "print(jaccard([1, 0], [0, 1]))\n",
    "print(jaccard([1, 1], [0, 1]))\n",
    "print(jaccard([1, 1, 1], [0, 1, 1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(pdist([[0, 0], [0, 1], [1, 1]], 'jaccard'))\n",
    "print(pdist([[1, 1], [1, 2], [2, 2]], 'jaccard'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(pdist([[0, 0, 0], [0, 0, 1]], 'jaccard'))\n",
    "print(pdist([[1, 1, 1], [1, 1, 0]], 'jaccard'))\n",
    "print(pdist([[2, 2, 2], [2, 2, 0]], 'jaccard'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(pdist([[2, 1, 1], [1, 1, 1]], 'jaccard'))\n",
    "print(pdist([[0, 1, 0], [1, 1, 1]], 'jaccard'))\n",
    "print(pdist([[1, 1, 1], [1, 1, 10]], 'jaccard'))\n",
    "print(pdist([[True, True, True], [True, True, False]], 'jaccard'))\n",
    "print(pdist([[False, True, True], [False, False, True]], 'jaccard'))\n",
    "print(pdist([[True, True, True], [True, True, False]], 'jaccard'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "X = np.array([[1,1,1,1],\n",
    "    [0,1,1,1],\n",
    "    [1,1,1,1],\n",
    "    [0,0,0,0],])\n",
    "linked = linkage(pdist(X, 'jaccard'))\n",
    "print(linked)\n",
    "\n",
    "labelList = []\n",
    "for row in X:\n",
    "    row_str = [str(x) for x in row]\n",
    "    labelList.append(\",\".join(row_str))\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "X = np.array([[1,1,1,1],\n",
    "    [1,1,1,1],\n",
    "    [0,1,1,1],\n",
    "    [0,0,0,0],])\n",
    "linked = linkage(pdist(X, 'jaccard'))\n",
    "print(linked)\n",
    "\n",
    "labelList = []\n",
    "for row in X:\n",
    "    row_str = [str(x) for x in row]\n",
    "    labelList.append(\",\".join(row_str))\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from scipy.spatial.distance import squareform\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "X = np.array([[1,1,1,1],\n",
    "    [0,0,0,0],\n",
    "    [0,1,1,1],])\n",
    "cdm = pdist(X, 'jaccard')\n",
    "print(cdm)\n",
    "print(\"\")\n",
    "print(squareform(cdm))\n",
    "print(\"\")\n",
    "linked = linkage(cdm)\n",
    "print(linked)\n",
    "\n",
    "labelList = []\n",
    "for row in X:\n",
    "    row_str = [str(x) for x in row]\n",
    "    labelList.append(\",\".join(row_str))\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_tree(centers):\n",
    "    clusters = {}\n",
    "    to_merge = linkage(pdist(X, 'jaccard'))\n",
    "    for i, merge in enumerate(to_merge):\n",
    "        if merge[0] <= len(to_merge):\n",
    "            # if it is an original point read it from the centers array\n",
    "            a = centers[int(merge[0]) - 1]\n",
    "        else:\n",
    "            # other wise read the cluster that has been created\n",
    "            a = clusters[int(merge[0])]\n",
    "\n",
    "        if merge[1] <= len(to_merge):\n",
    "            b = centers[int(merge[1]) - 1]\n",
    "        else:\n",
    "            b = clusters[int(merge[1])]\n",
    "        # the clusters are 1-indexed by scipy\n",
    "        clusters[1 + i + len(to_merge)] = {\n",
    "            'children' : [a, b]\n",
    "        }\n",
    "        # ^ you could optionally store other info here (e.g distances)\n",
    "    return clusters\n",
    "\n",
    "X = np.array([[1,1,1,1],\n",
    "    [0,0,0,0],\n",
    "    [0,1,1,1],])\n",
    "for k, v in create_tree(X).items():\n",
    "    print(k)\n",
    "    print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "linked = linkage(cluster_figure_curie_df[all_cluster_figure_curies_keys].to_numpy(), 'single')\n",
    "print(linked)\n",
    "\n",
    "labelList = cluster_figure_curie_df[\"figure_id\"].to_list()\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "X = np.array([[5,3],\n",
    "    [10,15],\n",
    "    [15,12],\n",
    "    [24,10],\n",
    "    [30,30],\n",
    "    [85,70],\n",
    "    [71,80],\n",
    "    [60,78],\n",
    "    [70,55],\n",
    "    [80,91],])\n",
    "X = cluster_figure_curie_df[all_cluster_figure_curies_keys]\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "labels = range(1, 11)\n",
    "labels = cluster_figure_curie_df[\"figure_id\"]\n",
    "plt.figure(figsize=(10, 7))\n",
    "plt.subplots_adjust(bottom=0.1)\n",
    "plt.scatter(X[:,0],X[:,1], label='True Position')\n",
    "\n",
    "for label, x, y in zip(labels, X[:, 0], X[:, 1]):\n",
    "    plt.annotate(\n",
    "        label,\n",
    "        xy=(x, y), xytext=(-3, 3),\n",
    "        textcoords='offset points', ha='right', va='bottom')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "X = np.array([[5,3],\n",
    "    [10,15],\n",
    "    [15,12],\n",
    "    [24,10],\n",
    "    [30,30],\n",
    "    [85,70],\n",
    "    [71,80],\n",
    "    [60,78],\n",
    "    [70,55],\n",
    "    [80,91],])\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "labels = range(1, 11)\n",
    "plt.figure(figsize=(10, 7))\n",
    "plt.subplots_adjust(bottom=0.1)\n",
    "plt.scatter(X[:,0],X[:,1], label='True Position')\n",
    "\n",
    "for label, x, y in zip(labels, X[:, 0], X[:, 1]):\n",
    "    plt.annotate(\n",
    "        label,\n",
    "        xy=(x, y), xytext=(-3, 3),\n",
    "        textcoords='offset points', ha='right', va='bottom')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "linked = linkage(X, 'single')\n",
    "print(linked)\n",
    "\n",
    "labelList = range(1, 11)\n",
    "\n",
    "plt.figure(figsize=(10, 7))\n",
    "dendrogram(linked,\n",
    "            orientation='top',\n",
    "            labels=labelList,\n",
    "            distance_sort='descending',\n",
    "            show_leaf_counts=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I chose one of the three figures tied for the lowest p-value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_figure_row = figures_df[figures_df[\"figure_id\"] == \"PMC186586__45793-8f2_F1TT.jpg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_figure_curies = set(best_figure_row[\"curie_set\"].tolist()[0])\n",
    "best_figure_curies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "bte_clustering.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
