{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The lab_black extension is already loaded. To reload it, use:\n",
      "  %reload_ext lab_black\n"
     ]
    }
   ],
   "source": [
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export 2020-02-24 Results as GMT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "import json\n",
    "import os\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import requests\n",
    "import requests_cache\n",
    "\n",
    "import rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "requests_cache.install_cache(\"pfocr_cache\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdf2pandas import rds2pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>number</th>\n",
       "      <th>reftext</th>\n",
       "      <th>year</th>\n",
       "      <th>pathway_score</th>\n",
       "      <th>pmc_ranked_result_index</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>papertitle</th>\n",
       "      <th>caption</th>\n",
       "      <th>figlink</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>filename</th>\n",
       "      <th>source_f</th>\n",
       "      <th>type.man</th>\n",
       "      <th>automl_index</th>\n",
       "      <th>figurl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC5653847__41598_2017_14124_Fig8_HTML.jpg</td>\n",
       "      <td>Figure 8</td>\n",
       "      <td>Céline Barthelemy, et al. Sci Rep. 2017;7:13816.</td>\n",
       "      <td>2017</td>\n",
       "      <td>0.968270</td>\n",
       "      <td>133303</td>\n",
       "      <td>Model of FTY720-induced transporter endocytosi...</td>\n",
       "      <td>FTY720-induced endocytosis of yeast and human ...</td>\n",
       "      <td>Model of FTY720-induced transporter endocytosi...</td>\n",
       "      <td>/pmc/articles/PMC5653847/figure/Fig8/</td>\n",
       "      <td>PMC5653847</td>\n",
       "      <td>41598_2017_14124_Fig8_HTML.jpg</td>\n",
       "      <td>../data/images/PMC5653847__41598_2017_14124_Fi...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3012</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC4187043__zh20191474070013.jpg</td>\n",
       "      <td>Fig. 13</td>\n",
       "      <td>Yuan Wei, et al. Am J Physiol Renal Physiol. 2...</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.965793</td>\n",
       "      <td>79929</td>\n",
       "      <td>Proposed signaling pathway by which the stimul...</td>\n",
       "      <td>Angiotensin II type 2 receptor regulates ROMK-...</td>\n",
       "      <td>Proposed signaling pathway by which the stimul...</td>\n",
       "      <td>/pmc/articles/PMC4187043/figure/F13/</td>\n",
       "      <td>PMC4187043</td>\n",
       "      <td>zh20191474070013.jpg</td>\n",
       "      <td>../data/images/PMC4187043__zh20191474070013.jpg</td>\n",
       "      <td>NA</td>\n",
       "      <td>4323</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC5746550__rsob-7-170228-g1.jpg</td>\n",
       "      <td>Figure 1</td>\n",
       "      <td>Georgia R. Frost, et al. Open Biol. 2017 Dec;7...</td>\n",
       "      <td>2017</td>\n",
       "      <td>0.962470</td>\n",
       "      <td>98034</td>\n",
       "      <td>AŒ≤ production</td>\n",
       "      <td>The role of astrocytes in amyloid production a...</td>\n",
       "      <td>AŒ≤ production. In the amyloidogenic pathway (...</td>\n",
       "      <td>/pmc/articles/PMC5746550/figure/RSOB170228F1/</td>\n",
       "      <td>PMC5746550</td>\n",
       "      <td>rsob-7-170228-g1.jpg</td>\n",
       "      <td>../data/images/PMC5746550__rsob-7-170228-g1.jpg</td>\n",
       "      <td>NA</td>\n",
       "      <td>6334</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC4211692__pone.0110875.g008.jpg</td>\n",
       "      <td>Figure 8</td>\n",
       "      <td>Enida Gjoni, et al. PLoS One. 2014;9(10):e110875.</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.966721</td>\n",
       "      <td>142401</td>\n",
       "      <td></td>\n",
       "      <td>Glucolipotoxicity Impairs Ceramide Flow from t...</td>\n",
       "      <td>Glucolipotoxicity impairs CERT- and vesicular-...</td>\n",
       "      <td>/pmc/articles/PMC4211692/figure/pone-0110875-g...</td>\n",
       "      <td>PMC4211692</td>\n",
       "      <td>pone.0110875.g008.jpg</td>\n",
       "      <td>../data/images/PMC4211692__pone.0110875.g008.jpg</td>\n",
       "      <td>NA</td>\n",
       "      <td>3808</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC2588433__nihms78212f8.jpg</td>\n",
       "      <td>Figure 8</td>\n",
       "      <td>Amanda L. Lewis, et al. J Biol Chem. ;282(38):...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.966758</td>\n",
       "      <td>67398</td>\n",
       "      <td></td>\n",
       "      <td>NeuA sialic acid O-acetylesterase activity mod...</td>\n",
       "      <td>Bacterial Sia biosynthesis can be divided into...</td>\n",
       "      <td>/pmc/articles/PMC2588433/figure/F8/</td>\n",
       "      <td>PMC2588433</td>\n",
       "      <td>nihms78212f8.jpg</td>\n",
       "      <td>../data/images/PMC2588433__nihms78212f8.jpg</td>\n",
       "      <td>NA</td>\n",
       "      <td>3790</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64638</th>\n",
       "      <td>PMC4216988__zh20221474360006.jpg</td>\n",
       "      <td>Fig. 6</td>\n",
       "      <td>Marcelo D. Carattino, et al. Am J Physiol Rena...</td>\n",
       "      <td>2014</td>\n",
       "      <td>0.143076</td>\n",
       "      <td>108774</td>\n",
       "      <td>Hypothetical mechanism of activation of ENaC b...</td>\n",
       "      <td>Prostasin interacts with the epithelial Na+ ch...</td>\n",
       "      <td>Hypothetical mechanism of activation of ENaC b...</td>\n",
       "      <td>/pmc/articles/PMC4216988/figure/F6/</td>\n",
       "      <td>PMC4216988</td>\n",
       "      <td>zh20221474360006.jpg</td>\n",
       "      <td>../data/images/PMC4216988__zh20221474360006.jpg</td>\n",
       "      <td>pathway</td>\n",
       "      <td>77324</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64639</th>\n",
       "      <td>PMC2873070__nihms128887f5.jpg</td>\n",
       "      <td>Scheme 1</td>\n",
       "      <td>Hua Cheng, et al. Neurobiol Aging. ;31(7):1188...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.127176</td>\n",
       "      <td>143547</td>\n",
       "      <td>A schematic diagram of a proposed working mode...</td>\n",
       "      <td>Apolipoprotein E mediates sulfatide depletion ...</td>\n",
       "      <td>A schematic diagram of a proposed working mode...</td>\n",
       "      <td>/pmc/articles/PMC2873070/figure/F5/</td>\n",
       "      <td>PMC2873070</td>\n",
       "      <td>nihms128887f5.jpg</td>\n",
       "      <td>../data/images/PMC2873070__nihms128887f5.jpg</td>\n",
       "      <td>pathway</td>\n",
       "      <td>78813</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64640</th>\n",
       "      <td>PMC3651446__pnas.1220523110fig06.jpg</td>\n",
       "      <td>Fig. 6</td>\n",
       "      <td>Jiun-Ming Wu, et al. Proc Natl Acad Sci U S A....</td>\n",
       "      <td>2013</td>\n",
       "      <td>0.055546</td>\n",
       "      <td>159643</td>\n",
       "      <td>Models for nucleation of centrosomal and kinet...</td>\n",
       "      <td>Aurora kinase inhibitors reveal mechanisms of ...</td>\n",
       "      <td>Models for nucleation of centrosomal and kinet...</td>\n",
       "      <td>/pmc/articles/PMC3651446/figure/fig06/</td>\n",
       "      <td>PMC3651446</td>\n",
       "      <td>pnas.1220523110fig06.jpg</td>\n",
       "      <td>../data/images/PMC3651446__pnas.1220523110fig0...</td>\n",
       "      <td>pathway</td>\n",
       "      <td>114977</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64641</th>\n",
       "      <td>PMC6770832__cancers-11-01236-g005.jpg</td>\n",
       "      <td>Figure 5</td>\n",
       "      <td>Carmel Mothersill, et al. Cancers (Basel). 201...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.140041</td>\n",
       "      <td>618</td>\n",
       "      <td>A simplified TGFŒ≤ pathway leading to p21 expr...</td>\n",
       "      <td>Relevance of Non-Targeted Effects for Radiothe...</td>\n",
       "      <td>A simplified TGFŒ≤ pathway leading to p21 expr...</td>\n",
       "      <td>/pmc/articles/PMC6770832/figure/cancers-11-012...</td>\n",
       "      <td>PMC6770832</td>\n",
       "      <td>cancers-11-01236-g005.jpg</td>\n",
       "      <td>../data/images/PMC6770832__cancers-11-01236-g0...</td>\n",
       "      <td>pathway</td>\n",
       "      <td>77606</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64642</th>\n",
       "      <td>PMC4609065__12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>Fig. 5</td>\n",
       "      <td>Hiro Takahashi, et al. BMC Cancer. 2015;15:718.</td>\n",
       "      <td>2015</td>\n",
       "      <td>0.090798</td>\n",
       "      <td>20865</td>\n",
       "      <td>Hypothetical model of EGFR and ANXA3 to fluoro...</td>\n",
       "      <td>Construction of possible integrated predictive...</td>\n",
       "      <td>Hypothetical model of EGFR and ANXA3 to fluoro...</td>\n",
       "      <td>/pmc/articles/PMC4609065/figure/Fig5/</td>\n",
       "      <td>PMC4609065</td>\n",
       "      <td>12885_2015_1721_Fig5_HTML.jpg</td>\n",
       "      <td>../data/images/PMC4609065__12885_2015_1721_Fig...</td>\n",
       "      <td>pathway</td>\n",
       "      <td>84038</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64643 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            figid    number  \\\n",
       "0      PMC5653847__41598_2017_14124_Fig8_HTML.jpg  Figure 8   \n",
       "1                PMC4187043__zh20191474070013.jpg   Fig. 13   \n",
       "2                PMC5746550__rsob-7-170228-g1.jpg  Figure 1   \n",
       "3               PMC4211692__pone.0110875.g008.jpg  Figure 8   \n",
       "4                    PMC2588433__nihms78212f8.jpg  Figure 8   \n",
       "...                                           ...       ...   \n",
       "64638            PMC4216988__zh20221474360006.jpg    Fig. 6   \n",
       "64639               PMC2873070__nihms128887f5.jpg  Scheme 1   \n",
       "64640        PMC3651446__pnas.1220523110fig06.jpg    Fig. 6   \n",
       "64641       PMC6770832__cancers-11-01236-g005.jpg  Figure 5   \n",
       "64642   PMC4609065__12885_2015_1721_Fig5_HTML.jpg    Fig. 5   \n",
       "\n",
       "                                                 reftext  year  pathway_score  \\\n",
       "0       Céline Barthelemy, et al. Sci Rep. 2017;7:13816.  2017       0.968270   \n",
       "1      Yuan Wei, et al. Am J Physiol Renal Physiol. 2...  2014       0.965793   \n",
       "2      Georgia R. Frost, et al. Open Biol. 2017 Dec;7...  2017       0.962470   \n",
       "3      Enida Gjoni, et al. PLoS One. 2014;9(10):e110875.  2014       0.966721   \n",
       "4      Amanda L. Lewis, et al. J Biol Chem. ;282(38):...    NA       0.966758   \n",
       "...                                                  ...   ...            ...   \n",
       "64638  Marcelo D. Carattino, et al. Am J Physiol Rena...  2014       0.143076   \n",
       "64639  Hua Cheng, et al. Neurobiol Aging. ;31(7):1188...    NA       0.127176   \n",
       "64640  Jiun-Ming Wu, et al. Proc Natl Acad Sci U S A....  2013       0.055546   \n",
       "64641  Carmel Mothersill, et al. Cancers (Basel). 201...  2019       0.140041   \n",
       "64642    Hiro Takahashi, et al. BMC Cancer. 2015;15:718.  2015       0.090798   \n",
       "\n",
       "       pmc_ranked_result_index  \\\n",
       "0                       133303   \n",
       "1                        79929   \n",
       "2                        98034   \n",
       "3                       142401   \n",
       "4                        67398   \n",
       "...                        ...   \n",
       "64638                   108774   \n",
       "64639                   143547   \n",
       "64640                   159643   \n",
       "64641                      618   \n",
       "64642                    20865   \n",
       "\n",
       "                                                figtitle  \\\n",
       "0      Model of FTY720-induced transporter endocytosi...   \n",
       "1      Proposed signaling pathway by which the stimul...   \n",
       "2                                         AŒ≤ production   \n",
       "3                                                          \n",
       "4                                                          \n",
       "...                                                  ...   \n",
       "64638  Hypothetical mechanism of activation of ENaC b...   \n",
       "64639  A schematic diagram of a proposed working mode...   \n",
       "64640  Models for nucleation of centrosomal and kinet...   \n",
       "64641  A simplified TGFŒ≤ pathway leading to p21 expr...   \n",
       "64642  Hypothetical model of EGFR and ANXA3 to fluoro...   \n",
       "\n",
       "                                              papertitle  \\\n",
       "0      FTY720-induced endocytosis of yeast and human ...   \n",
       "1      Angiotensin II type 2 receptor regulates ROMK-...   \n",
       "2      The role of astrocytes in amyloid production a...   \n",
       "3      Glucolipotoxicity Impairs Ceramide Flow from t...   \n",
       "4      NeuA sialic acid O-acetylesterase activity mod...   \n",
       "...                                                  ...   \n",
       "64638  Prostasin interacts with the epithelial Na+ ch...   \n",
       "64639  Apolipoprotein E mediates sulfatide depletion ...   \n",
       "64640  Aurora kinase inhibitors reveal mechanisms of ...   \n",
       "64641  Relevance of Non-Targeted Effects for Radiothe...   \n",
       "64642  Construction of possible integrated predictive...   \n",
       "\n",
       "                                                 caption  \\\n",
       "0      Model of FTY720-induced transporter endocytosi...   \n",
       "1      Proposed signaling pathway by which the stimul...   \n",
       "2      AŒ≤ production. In the amyloidogenic pathway (...   \n",
       "3      Glucolipotoxicity impairs CERT- and vesicular-...   \n",
       "4      Bacterial Sia biosynthesis can be divided into...   \n",
       "...                                                  ...   \n",
       "64638  Hypothetical mechanism of activation of ENaC b...   \n",
       "64639  A schematic diagram of a proposed working mode...   \n",
       "64640  Models for nucleation of centrosomal and kinet...   \n",
       "64641  A simplified TGFŒ≤ pathway leading to p21 expr...   \n",
       "64642  Hypothetical model of EGFR and ANXA3 to fluoro...   \n",
       "\n",
       "                                                 figlink       pmcid  \\\n",
       "0                  /pmc/articles/PMC5653847/figure/Fig8/  PMC5653847   \n",
       "1                   /pmc/articles/PMC4187043/figure/F13/  PMC4187043   \n",
       "2          /pmc/articles/PMC5746550/figure/RSOB170228F1/  PMC5746550   \n",
       "3      /pmc/articles/PMC4211692/figure/pone-0110875-g...  PMC4211692   \n",
       "4                    /pmc/articles/PMC2588433/figure/F8/  PMC2588433   \n",
       "...                                                  ...         ...   \n",
       "64638                /pmc/articles/PMC4216988/figure/F6/  PMC4216988   \n",
       "64639                /pmc/articles/PMC2873070/figure/F5/  PMC2873070   \n",
       "64640             /pmc/articles/PMC3651446/figure/fig06/  PMC3651446   \n",
       "64641  /pmc/articles/PMC6770832/figure/cancers-11-012...  PMC6770832   \n",
       "64642              /pmc/articles/PMC4609065/figure/Fig5/  PMC4609065   \n",
       "\n",
       "                             filename  \\\n",
       "0      41598_2017_14124_Fig8_HTML.jpg   \n",
       "1                zh20191474070013.jpg   \n",
       "2                rsob-7-170228-g1.jpg   \n",
       "3               pone.0110875.g008.jpg   \n",
       "4                    nihms78212f8.jpg   \n",
       "...                               ...   \n",
       "64638            zh20221474360006.jpg   \n",
       "64639               nihms128887f5.jpg   \n",
       "64640        pnas.1220523110fig06.jpg   \n",
       "64641       cancers-11-01236-g005.jpg   \n",
       "64642   12885_2015_1721_Fig5_HTML.jpg   \n",
       "\n",
       "                                                source_f type.man  \\\n",
       "0      ../data/images/PMC5653847__41598_2017_14124_Fi...       NA   \n",
       "1        ../data/images/PMC4187043__zh20191474070013.jpg       NA   \n",
       "2        ../data/images/PMC5746550__rsob-7-170228-g1.jpg       NA   \n",
       "3       ../data/images/PMC4211692__pone.0110875.g008.jpg       NA   \n",
       "4            ../data/images/PMC2588433__nihms78212f8.jpg       NA   \n",
       "...                                                  ...      ...   \n",
       "64638    ../data/images/PMC4216988__zh20221474360006.jpg  pathway   \n",
       "64639       ../data/images/PMC2873070__nihms128887f5.jpg  pathway   \n",
       "64640  ../data/images/PMC3651446__pnas.1220523110fig0...  pathway   \n",
       "64641  ../data/images/PMC6770832__cancers-11-01236-g0...  pathway   \n",
       "64642  ../data/images/PMC4609065__12885_2015_1721_Fig...  pathway   \n",
       "\n",
       "       automl_index                                             figurl  \n",
       "0              3012  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...  \n",
       "1              4323  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...  \n",
       "2              6334  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5...  \n",
       "3              3808  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...  \n",
       "4              3790  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "...             ...                                                ...  \n",
       "64638         77324  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...  \n",
       "64639         78813  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "64640        114977  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  \n",
       "64641         77606  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...  \n",
       "64642         84038  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4...  \n",
       "\n",
       "[64643 rows x 16 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# figures\n",
    "pfocr_figures_rds_url = (\n",
    "    \"https://www.dropbox.com/s/qhc33zho78rnaoj/pfocr_figures.rds?dl=1\"\n",
    ")\n",
    "\n",
    "with tempfile.NamedTemporaryFile(suffix=\".rds\") as f:\n",
    "    pfocr_figures_rds_path = f.name\n",
    "    with requests.get(pfocr_figures_rds_url, stream=True) as r:\n",
    "        for chunk in r.iter_content(chunk_size=128):\n",
    "            f.write(chunk)\n",
    "        f.seek(0)\n",
    "    pfocr_figures_df = rds2pandas(pfocr_figures_rds_path)\n",
    "pfocr_figures_df[\"figurl\"] = (\n",
    "    \"https://www.ncbi.nlm.nih.gov/pmc/articles/\"\n",
    "    + pfocr_figures_df[\"pmcid\"]\n",
    "    + \"/bin/\"\n",
    "    + pfocr_figures_df[\"filename\"]\n",
    ")\n",
    "pfocr_figures_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>Ga12,Gaq</td>\n",
       "      <td>G-ALPHA-q</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>GNAQ</td>\n",
       "      <td>2776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>Etk</td>\n",
       "      <td>ETK</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>BMX</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>FAK</td>\n",
       "      <td>FAK</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>5747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>AR*</td>\n",
       "      <td>AR</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>AR</td>\n",
       "      <td>367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>(Src</td>\n",
       "      <td>SRC</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>SRC</td>\n",
       "      <td>6714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112546</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>MEK-2</td>\n",
       "      <td>MEK2</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>MAP2K2</td>\n",
       "      <td>5605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112547</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>HRAS</td>\n",
       "      <td>3265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112548</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112549</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112550</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>SOS-1</td>\n",
       "      <td>SOS1</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>SOS1</td>\n",
       "      <td>6654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1112551 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               figid      pmcid      word     symbol  \\\n",
       "0        PMC100003__mb2410470011.jpg  PMC100003  Ga12,Gaq  G-ALPHA-q   \n",
       "1        PMC100003__mb2410470011.jpg  PMC100003       Etk        ETK   \n",
       "2        PMC100003__mb2410470011.jpg  PMC100003       FAK        FAK   \n",
       "3        PMC100003__mb2410470011.jpg  PMC100003       AR*         AR   \n",
       "4        PMC100003__mb2410470011.jpg  PMC100003      (Src        SRC   \n",
       "...                              ...        ...       ...        ...   \n",
       "1112546   PMC99976__mb2310138007.jpg   PMC99976     MEK-2       MEK2   \n",
       "1112547   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112548   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112549   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112550   PMC99976__mb2310138007.jpg   PMC99976     SOS-1       SOS1   \n",
       "\n",
       "                     source hgnc_symbol  entrez  \n",
       "0         hgnc_alias_symbol        GNAQ    2776  \n",
       "1         hgnc_alias_symbol         BMX     660  \n",
       "2         hgnc_alias_symbol        PTK2    5747  \n",
       "3               hgnc_symbol          AR     367  \n",
       "4               hgnc_symbol         SRC    6714  \n",
       "...                     ...         ...     ...  \n",
       "1112546   hgnc_alias_symbol      MAP2K2    5605  \n",
       "1112547  bioentities_symbol        HRAS    3265  \n",
       "1112548  bioentities_symbol        KRAS    3845  \n",
       "1112549  bioentities_symbol        NRAS    4893  \n",
       "1112550         hgnc_symbol        SOS1    6654  \n",
       "\n",
       "[1112551 rows x 7 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes\n",
    "pfocr_genes_rds_url = \"https://www.dropbox.com/s/alf7auvxve36oer/pfocr_genes.rds?dl=1\"\n",
    "\n",
    "with tempfile.NamedTemporaryFile(suffix=\".rds\") as f:\n",
    "    pfocr_genes_rds_path = f.name\n",
    "    with requests.get(pfocr_genes_rds_url, stream=True) as r:\n",
    "        for chunk in r.iter_content(chunk_size=128):\n",
    "            f.write(chunk)\n",
    "        f.seek(0)\n",
    "    pfocr_genes_df = rds2pandas(pfocr_genes_rds_path)\n",
    "pfocr_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>nobe_count</th>\n",
       "      <th>entrez_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC100005__mb2410575011.jpg</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC100008__mb2411709009.jpg</td>\n",
       "      <td>8</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC101225__1475-2867-1-1-1.jpg</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC101242__gkf20707.jpg</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58957</th>\n",
       "      <td>PMC99889__mb2110211013.jpg</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58958</th>\n",
       "      <td>PMC99903__mb2110106008.jpg</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58959</th>\n",
       "      <td>PMC99957__mb2310813001.jpg</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58960</th>\n",
       "      <td>PMC99957__mb2310813011.jpg</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58961</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58962 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                figid  nobe_count  entrez_count\n",
       "0         PMC100003__mb2410470011.jpg           5             5\n",
       "1         PMC100005__mb2410575011.jpg           6            16\n",
       "2         PMC100008__mb2411709009.jpg           8            52\n",
       "3      PMC101225__1475-2867-1-1-1.jpg           1             1\n",
       "4             PMC101242__gkf20707.jpg          13            13\n",
       "...                               ...         ...           ...\n",
       "58957      PMC99889__mb2110211013.jpg          12            22\n",
       "58958      PMC99903__mb2110106008.jpg           1             1\n",
       "58959      PMC99957__mb2310813001.jpg          10            10\n",
       "58960      PMC99957__mb2310813011.jpg           2             2\n",
       "58961      PMC99976__mb2310138007.jpg           4            14\n",
       "\n",
       "[58962 rows x 3 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nobe_counts\n",
    "pfocr_nobe_counts_rds_url = (\n",
    "    \"https://www.dropbox.com/s/pd8n5hs4vrvb4st/pfocr_nobe_counts.rds?dl=1\"\n",
    ")\n",
    "\n",
    "with tempfile.NamedTemporaryFile(suffix=\".rds\") as f:\n",
    "    pfocr_nobe_counts_rds_path = f.name\n",
    "    with requests.get(pfocr_nobe_counts_rds_url, stream=True) as r:\n",
    "        for chunk in r.iter_content(chunk_size=128):\n",
    "            f.write(chunk)\n",
    "        f.seek(0)\n",
    "    pfocr_nobe_counts_df = rds2pandas(pfocr_nobe_counts_rds_path)\n",
    "pfocr_nobe_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>number</th>\n",
       "      <th>reftext</th>\n",
       "      <th>year</th>\n",
       "      <th>pathway_score</th>\n",
       "      <th>pmc_ranked_result_index</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>papertitle</th>\n",
       "      <th>caption</th>\n",
       "      <th>figlink</th>\n",
       "      <th>...</th>\n",
       "      <th>type.man</th>\n",
       "      <th>automl_index</th>\n",
       "      <th>figurl</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "      <th>nobe_count</th>\n",
       "      <th>entrez_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC1201394__nihms3677f4.jpg</td>\n",
       "      <td>Fig. 4</td>\n",
       "      <td>Marissa J. Carstens, et al. J Biol Chem. ;279(...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.894122</td>\n",
       "      <td>38017</td>\n",
       "      <td></td>\n",
       "      <td>Cell Cycle Arrest and Cell Death Are Controlle...</td>\n",
       "      <td>A, Mdm2 is negatively regulated by p19Arf. As ...</td>\n",
       "      <td>/pmc/articles/PMC1201394/figure/F4/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>39679</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>(Cdkn2a</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC1201394__nihms3677f4.jpg</td>\n",
       "      <td>Fig. 4</td>\n",
       "      <td>Marissa J. Carstens, et al. J Biol Chem. ;279(...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.894122</td>\n",
       "      <td>38017</td>\n",
       "      <td></td>\n",
       "      <td>Cell Cycle Arrest and Cell Death Are Controlle...</td>\n",
       "      <td>A, Mdm2 is negatively regulated by p19Arf. As ...</td>\n",
       "      <td>/pmc/articles/PMC1201394/figure/F4/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>39679</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>p19/Arf</td>\n",
       "      <td>ARF</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC1201394__nihms3677f4.jpg</td>\n",
       "      <td>Fig. 4</td>\n",
       "      <td>Marissa J. Carstens, et al. J Biol Chem. ;279(...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.894122</td>\n",
       "      <td>38017</td>\n",
       "      <td></td>\n",
       "      <td>Cell Cycle Arrest and Cell Death Are Controlle...</td>\n",
       "      <td>A, Mdm2 is negatively regulated by p19Arf. As ...</td>\n",
       "      <td>/pmc/articles/PMC1201394/figure/F4/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>39679</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>G1/S</td>\n",
       "      <td>Gs</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>GNAL</td>\n",
       "      <td>2774</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC1201394__nihms3677f4.jpg</td>\n",
       "      <td>Fig. 4</td>\n",
       "      <td>Marissa J. Carstens, et al. J Biol Chem. ;279(...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.894122</td>\n",
       "      <td>38017</td>\n",
       "      <td></td>\n",
       "      <td>Cell Cycle Arrest and Cell Death Are Controlle...</td>\n",
       "      <td>A, Mdm2 is negatively regulated by p19Arf. As ...</td>\n",
       "      <td>/pmc/articles/PMC1201394/figure/F4/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>39679</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>G1/S</td>\n",
       "      <td>Gs</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>GNAS</td>\n",
       "      <td>2778</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC1201394__nihms3677f4.jpg</td>\n",
       "      <td>Fig. 4</td>\n",
       "      <td>Marissa J. Carstens, et al. J Biol Chem. ;279(...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.894122</td>\n",
       "      <td>38017</td>\n",
       "      <td></td>\n",
       "      <td>Cell Cycle Arrest and Cell Death Are Controlle...</td>\n",
       "      <td>A, Mdm2 is negatively regulated by p19Arf. As ...</td>\n",
       "      <td>/pmc/articles/PMC1201394/figure/F4/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>39679</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "      <td>Mdm2</td>\n",
       "      <td>MDM2</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>MDM2</td>\n",
       "      <td>4193</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112546</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td>Fig. 2</td>\n",
       "      <td>Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.828043</td>\n",
       "      <td>65444</td>\n",
       "      <td></td>\n",
       "      <td>Intracellular sensing of viral genomes and vir...</td>\n",
       "      <td>Schematic presentation of positive and negativ...</td>\n",
       "      <td>/pmc/articles/PMC6906418/figure/Fig2/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>50139</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>TRIM41</td>\n",
       "      <td>TRIM41</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>TRIM41</td>\n",
       "      <td>90933</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112547</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td>Fig. 2</td>\n",
       "      <td>Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.828043</td>\n",
       "      <td>65444</td>\n",
       "      <td></td>\n",
       "      <td>Intracellular sensing of viral genomes and vir...</td>\n",
       "      <td>Schematic presentation of positive and negativ...</td>\n",
       "      <td>/pmc/articles/PMC6906418/figure/Fig2/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>50139</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>RNF185</td>\n",
       "      <td>RNF185</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>RNF185</td>\n",
       "      <td>91445</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112548</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td>Fig. 2</td>\n",
       "      <td>Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.828043</td>\n",
       "      <td>65444</td>\n",
       "      <td></td>\n",
       "      <td>Intracellular sensing of viral genomes and vir...</td>\n",
       "      <td>Schematic presentation of positive and negativ...</td>\n",
       "      <td>/pmc/articles/PMC6906418/figure/Fig2/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>50139</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>CGAS</td>\n",
       "      <td>CGAS</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>CGAS</td>\n",
       "      <td>115004</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112549</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td>Fig. 2</td>\n",
       "      <td>Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.828043</td>\n",
       "      <td>65444</td>\n",
       "      <td></td>\n",
       "      <td>Intracellular sensing of viral genomes and vir...</td>\n",
       "      <td>Schematic presentation of positive and negativ...</td>\n",
       "      <td>/pmc/articles/PMC6906418/figure/Fig2/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>50139</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>TTLL6</td>\n",
       "      <td>TTLL6</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>TTLL6</td>\n",
       "      <td>284076</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112550</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td>Fig. 2</td>\n",
       "      <td>Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...</td>\n",
       "      <td>2019</td>\n",
       "      <td>0.828043</td>\n",
       "      <td>65444</td>\n",
       "      <td></td>\n",
       "      <td>Intracellular sensing of viral genomes and vir...</td>\n",
       "      <td>Schematic presentation of positive and negativ...</td>\n",
       "      <td>/pmc/articles/PMC6906418/figure/Fig2/</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>50139</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...</td>\n",
       "      <td>STING</td>\n",
       "      <td>STING</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>TMEM173</td>\n",
       "      <td>340061</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1112551 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            figid  number  \\\n",
       "0                     PMC1201394__nihms3677f4.jpg  Fig. 4   \n",
       "1                     PMC1201394__nihms3677f4.jpg  Fig. 4   \n",
       "2                     PMC1201394__nihms3677f4.jpg  Fig. 4   \n",
       "3                     PMC1201394__nihms3677f4.jpg  Fig. 4   \n",
       "4                     PMC1201394__nihms3677f4.jpg  Fig. 4   \n",
       "...                                           ...     ...   \n",
       "1112546  PMC6906418__12276_2019_299_Fig2_HTML.jpg  Fig. 2   \n",
       "1112547  PMC6906418__12276_2019_299_Fig2_HTML.jpg  Fig. 2   \n",
       "1112548  PMC6906418__12276_2019_299_Fig2_HTML.jpg  Fig. 2   \n",
       "1112549  PMC6906418__12276_2019_299_Fig2_HTML.jpg  Fig. 2   \n",
       "1112550  PMC6906418__12276_2019_299_Fig2_HTML.jpg  Fig. 2   \n",
       "\n",
       "                                                   reftext  year  \\\n",
       "0        Marissa J. Carstens, et al. J Biol Chem. ;279(...    NA   \n",
       "1        Marissa J. Carstens, et al. J Biol Chem. ;279(...    NA   \n",
       "2        Marissa J. Carstens, et al. J Biol Chem. ;279(...    NA   \n",
       "3        Marissa J. Carstens, et al. J Biol Chem. ;279(...    NA   \n",
       "4        Marissa J. Carstens, et al. J Biol Chem. ;279(...    NA   \n",
       "...                                                    ...   ...   \n",
       "1112546  Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...  2019   \n",
       "1112547  Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...  2019   \n",
       "1112548  Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...  2019   \n",
       "1112549  Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...  2019   \n",
       "1112550  Hyun-Cheol Lee, et al. Exp Mol Med. 2019 Dec;5...  2019   \n",
       "\n",
       "         pathway_score  pmc_ranked_result_index figtitle  \\\n",
       "0             0.894122                    38017            \n",
       "1             0.894122                    38017            \n",
       "2             0.894122                    38017            \n",
       "3             0.894122                    38017            \n",
       "4             0.894122                    38017            \n",
       "...                ...                      ...      ...   \n",
       "1112546       0.828043                    65444            \n",
       "1112547       0.828043                    65444            \n",
       "1112548       0.828043                    65444            \n",
       "1112549       0.828043                    65444            \n",
       "1112550       0.828043                    65444            \n",
       "\n",
       "                                                papertitle  \\\n",
       "0        Cell Cycle Arrest and Cell Death Are Controlle...   \n",
       "1        Cell Cycle Arrest and Cell Death Are Controlle...   \n",
       "2        Cell Cycle Arrest and Cell Death Are Controlle...   \n",
       "3        Cell Cycle Arrest and Cell Death Are Controlle...   \n",
       "4        Cell Cycle Arrest and Cell Death Are Controlle...   \n",
       "...                                                    ...   \n",
       "1112546  Intracellular sensing of viral genomes and vir...   \n",
       "1112547  Intracellular sensing of viral genomes and vir...   \n",
       "1112548  Intracellular sensing of viral genomes and vir...   \n",
       "1112549  Intracellular sensing of viral genomes and vir...   \n",
       "1112550  Intracellular sensing of viral genomes and vir...   \n",
       "\n",
       "                                                   caption  \\\n",
       "0        A, Mdm2 is negatively regulated by p19Arf. As ...   \n",
       "1        A, Mdm2 is negatively regulated by p19Arf. As ...   \n",
       "2        A, Mdm2 is negatively regulated by p19Arf. As ...   \n",
       "3        A, Mdm2 is negatively regulated by p19Arf. As ...   \n",
       "4        A, Mdm2 is negatively regulated by p19Arf. As ...   \n",
       "...                                                    ...   \n",
       "1112546  Schematic presentation of positive and negativ...   \n",
       "1112547  Schematic presentation of positive and negativ...   \n",
       "1112548  Schematic presentation of positive and negativ...   \n",
       "1112549  Schematic presentation of positive and negativ...   \n",
       "1112550  Schematic presentation of positive and negativ...   \n",
       "\n",
       "                                       figlink  ... type.man automl_index  \\\n",
       "0          /pmc/articles/PMC1201394/figure/F4/  ...       NA        39679   \n",
       "1          /pmc/articles/PMC1201394/figure/F4/  ...       NA        39679   \n",
       "2          /pmc/articles/PMC1201394/figure/F4/  ...       NA        39679   \n",
       "3          /pmc/articles/PMC1201394/figure/F4/  ...       NA        39679   \n",
       "4          /pmc/articles/PMC1201394/figure/F4/  ...       NA        39679   \n",
       "...                                        ...  ...      ...          ...   \n",
       "1112546  /pmc/articles/PMC6906418/figure/Fig2/  ...       NA        50139   \n",
       "1112547  /pmc/articles/PMC6906418/figure/Fig2/  ...       NA        50139   \n",
       "1112548  /pmc/articles/PMC6906418/figure/Fig2/  ...       NA        50139   \n",
       "1112549  /pmc/articles/PMC6906418/figure/Fig2/  ...       NA        50139   \n",
       "1112550  /pmc/articles/PMC6906418/figure/Fig2/  ...       NA        50139   \n",
       "\n",
       "                                                    figurl     word  symbol  \\\n",
       "0        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  (Cdkn2a  CDKN2A   \n",
       "1        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  p19/Arf     ARF   \n",
       "2        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...     G1/S      Gs   \n",
       "3        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...     G1/S      Gs   \n",
       "4        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...     Mdm2    MDM2   \n",
       "...                                                    ...      ...     ...   \n",
       "1112546  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   TRIM41  TRIM41   \n",
       "1112547  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...   RNF185  RNF185   \n",
       "1112548  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...     CGAS    CGAS   \n",
       "1112549  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...    TTLL6   TTLL6   \n",
       "1112550  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6...    STING   STING   \n",
       "\n",
       "                     source hgnc_symbol  entrez nobe_count entrez_count  \n",
       "0               hgnc_symbol      CDKN2A    1029          6            7  \n",
       "1         hgnc_alias_symbol      CDKN2A    1029          6            7  \n",
       "2        bioentities_symbol        GNAL    2774          6            7  \n",
       "3        bioentities_symbol        GNAS    2778          6            7  \n",
       "4               hgnc_symbol        MDM2    4193          6            7  \n",
       "...                     ...         ...     ...        ...          ...  \n",
       "1112546         hgnc_symbol      TRIM41   90933         24           28  \n",
       "1112547         hgnc_symbol      RNF185   91445         24           28  \n",
       "1112548         hgnc_symbol        CGAS  115004         24           28  \n",
       "1112549         hgnc_symbol       TTLL6  284076         24           28  \n",
       "1112550   hgnc_alias_symbol     TMEM173  340061         24           28  \n",
       "\n",
       "[1112551 rows x 23 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge the dfs to get pfocr analysis set\n",
    "pfocr_df = (\n",
    "    (\n",
    "        pfocr_figures_df.join(\n",
    "            pfocr_genes_df.drop(columns=[\"pmcid\"]).set_index(\"figid\"),\n",
    "            on=\"figid\",\n",
    "            how=\"inner\",\n",
    "        ).join(\n",
    "            pfocr_nobe_counts_df.set_index(\"figid\"), on=\"figid\", how=\"inner\",\n",
    "        )\n",
    "    )\n",
    "    .sort_values([\"year\", \"pmcid\", \"figid\", \"entrez\"])\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "pfocr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58962\n"
     ]
    }
   ],
   "source": [
    "gene_counts_df = (\n",
    "    pfocr_df[[\"figid\", \"entrez\"]]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"figid\")\n",
    "    .count()\n",
    "    .rename(columns={\"entrez\": \"unique_gene_count\"})\n",
    ")\n",
    "print(len(gene_counts_df))\n",
    "pfocr_df = pfocr_df.join(gene_counts_df, on=\"figid\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Limit to figures having 10 or more unique genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PMC1249490__nihms5296f3.jpg</td>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>2774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PMC1249490__nihms5296f3.jpg</td>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>2778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PMC1249490__nihms5296f3.jpg</td>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>5330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PMC1249490__nihms5296f3.jpg</td>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>5331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PMC1249490__nihms5296f3.jpg</td>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>5332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112546</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td></td>\n",
       "      <td>90933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112547</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td></td>\n",
       "      <td>91445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112548</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td></td>\n",
       "      <td>115004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112549</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td></td>\n",
       "      <td>284076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112550</th>\n",
       "      <td>PMC6906418__12276_2019_299_Fig2_HTML.jpg</td>\n",
       "      <td></td>\n",
       "      <td>340061</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>969257 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            figid  \\\n",
       "16                    PMC1249490__nihms5296f3.jpg   \n",
       "17                    PMC1249490__nihms5296f3.jpg   \n",
       "18                    PMC1249490__nihms5296f3.jpg   \n",
       "19                    PMC1249490__nihms5296f3.jpg   \n",
       "20                    PMC1249490__nihms5296f3.jpg   \n",
       "...                                           ...   \n",
       "1112546  PMC6906418__12276_2019_299_Fig2_HTML.jpg   \n",
       "1112547  PMC6906418__12276_2019_299_Fig2_HTML.jpg   \n",
       "1112548  PMC6906418__12276_2019_299_Fig2_HTML.jpg   \n",
       "1112549  PMC6906418__12276_2019_299_Fig2_HTML.jpg   \n",
       "1112550  PMC6906418__12276_2019_299_Fig2_HTML.jpg   \n",
       "\n",
       "                                                  figtitle  entrez  \n",
       "16       The primary targets of lithium in the cell are...    2774  \n",
       "17       The primary targets of lithium in the cell are...    2778  \n",
       "18       The primary targets of lithium in the cell are...    5330  \n",
       "19       The primary targets of lithium in the cell are...    5331  \n",
       "20       The primary targets of lithium in the cell are...    5332  \n",
       "...                                                    ...     ...  \n",
       "1112546                                                      90933  \n",
       "1112547                                                      91445  \n",
       "1112548                                                     115004  \n",
       "1112549                                                     284076  \n",
       "1112550                                                     340061  \n",
       "\n",
       "[969257 rows x 3 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff_pfocr_df = pfocr_df[pfocr_df[\"unique_gene_count\"] >= 10][\n",
    "    [\"figid\", \"figtitle\", \"entrez\"]\n",
    "].drop_duplicates()\n",
    "cutoff_pfocr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32277\n",
      "23450\n",
      "13153\n"
     ]
    }
   ],
   "source": [
    "print(len(cutoff_pfocr_df[\"figid\"].unique()))\n",
    "print(len(cutoff_pfocr_df[\"figtitle\"].unique()))\n",
    "print(len(cutoff_pfocr_df[\"entrez\"].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export GMT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[GMT format description](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29).\n",
    "\n",
    "Sample of what we want:\n",
    "```\n",
    "PMC100005__mb2410575011.jpg     Formation of TrkA–src-aPKC complexes    5293    5294    25759   6714    8503    4914    10818   399694  30849   7189      23533   5290    6464    146850  53358   5291\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: there are some characters that could mess up the parsing of a GMT. For example, a `%` or a `\\t` would be parsed as a delimiter. Also, if a doublequote `\"` can be used to enclose fields with characters like spaces, then a title containing a doublequote would mess up a parser.\n",
    "\n",
    "In those cases, do we want to escape the special characters or remove them?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genes_to_gmt(df):\n",
    "    return df[\"entrez\"].astype(\"str\").str.cat(sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>genes_str</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC1249490__nihms5296f3.jpg</th>\n",
       "      <td>The primary targets of lithium in the cell are...</td>\n",
       "      <td>[2774, 2778, 5330, 5331, 5332, 5333, 5335, 533...</td>\n",
       "      <td>2774\\t2778\\t5330\\t5331\\t5332\\t5333\\t5335\\t5336...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1307498__nihms5498f3.jpg</th>\n",
       "      <td>Schematic model of insulin-stimulated glucose ...</td>\n",
       "      <td>[207, 208, 3667, 5163, 5164, 5165, 5166, 5290,...</td>\n",
       "      <td>207\\t208\\t3667\\t5163\\t5164\\t5165\\t5166\\t5290\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1307511__nihms2079f3.jpg</th>\n",
       "      <td></td>\n",
       "      <td>[317, 834, 835, 836, 837, 838, 839, 840, 841, ...</td>\n",
       "      <td>317\\t834\\t835\\t836\\t837\\t838\\t839\\t840\\t841\\t8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1351030__nihms2404f8.jpg</th>\n",
       "      <td>A model of signaling pathway from oncogenic H-...</td>\n",
       "      <td>[595, 688, 1432, 1958, 3265, 5594, 5595, 5599,...</td>\n",
       "      <td>595\\t688\\t1432\\t1958\\t3265\\t5594\\t5595\\t5599\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1352153__nihms-7536-0007.jpg</th>\n",
       "      <td>Model for GalR1 neuroplasticity within the LC</td>\n",
       "      <td>[466, 1385, 1386, 1390, 2587, 4236, 5566, 5567...</td>\n",
       "      <td>466\\t1385\\t1386\\t1390\\t2587\\t4236\\t5566\\t5567\\...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6904862__WJSC-11-1104-g003.jpg</th>\n",
       "      <td>Biomechanical stress stimulates the mechanosen...</td>\n",
       "      <td>[207, 208, 1432, 5594, 5595, 5599, 5600, 5601,...</td>\n",
       "      <td>207\\t208\\t1432\\t5594\\t5595\\t5599\\t5600\\t5601\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6904864__WJSC-11-1084-g001.jpg</th>\n",
       "      <td>Small molecules guiding mesenchymal stem cell ...</td>\n",
       "      <td>[207, 208, 649, 650, 651, 652, 653, 654, 655, ...</td>\n",
       "      <td>207\\t208\\t649\\t650\\t651\\t652\\t653\\t654\\t655\\t6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6905007__40246_2019_252_Fig2_HTML.jpg</th>\n",
       "      <td>LRF/ZBTB7A‚Äôs silencing compromises Warburg e...</td>\n",
       "      <td>[355, 596, 834, 835, 836, 837, 838, 839, 840, ...</td>\n",
       "      <td>355\\t596\\t834\\t835\\t836\\t837\\t838\\t839\\t840\\t8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6906418__12276_2019_299_Fig1_HTML.jpg</th>\n",
       "      <td></td>\n",
       "      <td>[834, 835, 836, 837, 838, 839, 840, 841, 842, ...</td>\n",
       "      <td>834\\t835\\t836\\t837\\t838\\t839\\t840\\t841\\t842\\t8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC6906418__12276_2019_299_Fig2_HTML.jpg</th>\n",
       "      <td></td>\n",
       "      <td>[207, 208, 1147, 3439, 3551, 3661, 3665, 4790,...</td>\n",
       "      <td>207\\t208\\t1147\\t3439\\t3551\\t3661\\t3665\\t4790\\t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32277 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                   figtitle  \\\n",
       "figid                                                                                         \n",
       "PMC1249490__nihms5296f3.jpg               The primary targets of lithium in the cell are...   \n",
       "PMC1307498__nihms5498f3.jpg               Schematic model of insulin-stimulated glucose ...   \n",
       "PMC1307511__nihms2079f3.jpg                                                                   \n",
       "PMC1351030__nihms2404f8.jpg               A model of signaling pathway from oncogenic H-...   \n",
       "PMC1352153__nihms-7536-0007.jpg               Model for GalR1 neuroplasticity within the LC   \n",
       "...                                                                                     ...   \n",
       "PMC6904862__WJSC-11-1104-g003.jpg         Biomechanical stress stimulates the mechanosen...   \n",
       "PMC6904864__WJSC-11-1084-g001.jpg         Small molecules guiding mesenchymal stem cell ...   \n",
       "PMC6905007__40246_2019_252_Fig2_HTML.jpg  LRF/ZBTB7A‚Äôs silencing compromises Warburg e...   \n",
       "PMC6906418__12276_2019_299_Fig1_HTML.jpg                                                      \n",
       "PMC6906418__12276_2019_299_Fig2_HTML.jpg                                                      \n",
       "\n",
       "                                                                                      genes  \\\n",
       "figid                                                                                         \n",
       "PMC1249490__nihms5296f3.jpg               [2774, 2778, 5330, 5331, 5332, 5333, 5335, 533...   \n",
       "PMC1307498__nihms5498f3.jpg               [207, 208, 3667, 5163, 5164, 5165, 5166, 5290,...   \n",
       "PMC1307511__nihms2079f3.jpg               [317, 834, 835, 836, 837, 838, 839, 840, 841, ...   \n",
       "PMC1351030__nihms2404f8.jpg               [595, 688, 1432, 1958, 3265, 5594, 5595, 5599,...   \n",
       "PMC1352153__nihms-7536-0007.jpg           [466, 1385, 1386, 1390, 2587, 4236, 5566, 5567...   \n",
       "...                                                                                     ...   \n",
       "PMC6904862__WJSC-11-1104-g003.jpg         [207, 208, 1432, 5594, 5595, 5599, 5600, 5601,...   \n",
       "PMC6904864__WJSC-11-1084-g001.jpg         [207, 208, 649, 650, 651, 652, 653, 654, 655, ...   \n",
       "PMC6905007__40246_2019_252_Fig2_HTML.jpg  [355, 596, 834, 835, 836, 837, 838, 839, 840, ...   \n",
       "PMC6906418__12276_2019_299_Fig1_HTML.jpg  [834, 835, 836, 837, 838, 839, 840, 841, 842, ...   \n",
       "PMC6906418__12276_2019_299_Fig2_HTML.jpg  [207, 208, 1147, 3439, 3551, 3661, 3665, 4790,...   \n",
       "\n",
       "                                                                                  genes_str  \n",
       "figid                                                                                        \n",
       "PMC1249490__nihms5296f3.jpg               2774\\t2778\\t5330\\t5331\\t5332\\t5333\\t5335\\t5336...  \n",
       "PMC1307498__nihms5498f3.jpg               207\\t208\\t3667\\t5163\\t5164\\t5165\\t5166\\t5290\\t...  \n",
       "PMC1307511__nihms2079f3.jpg               317\\t834\\t835\\t836\\t837\\t838\\t839\\t840\\t841\\t8...  \n",
       "PMC1351030__nihms2404f8.jpg               595\\t688\\t1432\\t1958\\t3265\\t5594\\t5595\\t5599\\t...  \n",
       "PMC1352153__nihms-7536-0007.jpg           466\\t1385\\t1386\\t1390\\t2587\\t4236\\t5566\\t5567\\...  \n",
       "...                                                                                     ...  \n",
       "PMC6904862__WJSC-11-1104-g003.jpg         207\\t208\\t1432\\t5594\\t5595\\t5599\\t5600\\t5601\\t...  \n",
       "PMC6904864__WJSC-11-1084-g001.jpg         207\\t208\\t649\\t650\\t651\\t652\\t653\\t654\\t655\\t6...  \n",
       "PMC6905007__40246_2019_252_Fig2_HTML.jpg  355\\t596\\t834\\t835\\t836\\t837\\t838\\t839\\t840\\t8...  \n",
       "PMC6906418__12276_2019_299_Fig1_HTML.jpg  834\\t835\\t836\\t837\\t838\\t839\\t840\\t841\\t842\\t8...  \n",
       "PMC6906418__12276_2019_299_Fig2_HTML.jpg  207\\t208\\t1147\\t3439\\t3551\\t3661\\t3665\\t4790\\t...  \n",
       "\n",
       "[32277 rows x 3 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gmt_df = (\n",
    "    cutoff_pfocr_df[[\"figid\", \"figtitle\"]]\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"figid\")\n",
    "    .copy(deep=True)\n",
    ")\n",
    "gmt_df[\"genes\"] = (\n",
    "    cutoff_pfocr_df.set_index(\"figid\")\n",
    "    .groupby(\"figid\")\n",
    "    .apply(lambda x: x[\"entrez\"].astype(\"str\").to_list())\n",
    ")\n",
    "gmt_df[\"genes_str\"] = (\n",
    "    cutoff_pfocr_df.set_index(\"figid\").groupby(\"figid\").apply(genes_to_gmt)\n",
    ")\n",
    "gmt_df = gmt_df.fillna(\"\")\n",
    "\n",
    "gmt_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's export the GMT file. The paths match AR's system, but you'll need to update them if yours aren't the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output saved: /home/ariutta/Dropbox (Gladstone)/Documents/PFOCR_25Years/pfocr_curated.gmt\n"
     ]
    }
   ],
   "source": [
    "dropbox_dir = Path(\"~/Dropbox (Gladstone)\").expanduser().resolve()\n",
    "paper_dir = dropbox_dir.joinpath(\"Documents/PFOCR_25Years\").expanduser().resolve()\n",
    "data_dir = paper_dir\n",
    "# data_dir = paper_dir.joinpath(\"raw\")\n",
    "\n",
    "if not data_dir.exists():\n",
    "    tmp_dir = tempfile.TemporaryDirectory()\n",
    "    # can explicitly remove tmp_dir with tmp_dir.cleanup()\n",
    "    data_dir = Path(tmp_dir.name)\n",
    "\n",
    "gmt_path = data_dir.joinpath(\"pfocr_curated.gmt\")\n",
    "\n",
    "with open(gmt_path, \"w\") as gmt_f:\n",
    "    for figid, row in gmt_df.iterrows():\n",
    "        gmt_f.write(\"\\t\".join([figid, row[\"figtitle\"], row[\"genes_str\"],]))\n",
    "        gmt_f.write(\"\\n\")\n",
    "\n",
    "print(f\"Output saved: {gmt_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "ar_gmt_df = gmt_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC100005__mb2410575011.jpg</td>\n",
       "      <td>Formation of TrkA–src-aPKC complexes</td>\n",
       "      <td>[5293, 5294, 25759, 6714, 8503, 4914, 10818, 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC100008__mb2411709009.jpg</td>\n",
       "      <td>Cooperation between Beta-AR and B2R signaling ...</td>\n",
       "      <td>[5331, 5576, 51196, 23236, 5567, 5566, 107, 11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC101242__gkf20707.jpg</td>\n",
       "      <td>Up-regulation of RAD51 in 8-MOP-treated cells</td>\n",
       "      <td>[5883, 5980, 11200, 5893, 10111, 5888, 546, 47...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC101600__ac1100359003.jpg</td>\n",
       "      <td>Hypothetical biosynthetic pathway of coumermyc...</td>\n",
       "      <td>[125965, 84701, 1327, 10063, 1352, 1353, 1351,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC102186__mb2400926009.jpg</td>\n",
       "      <td>Ras-mediated signaling pathways</td>\n",
       "      <td>[6300, 5602, 6300, 5879, 5058, 5062, 57144, 58...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32274</th>\n",
       "      <td>PMC99537__jb1610128005.jpg</td>\n",
       "      <td>Basic metabolic pathways in C</td>\n",
       "      <td>[23038, 4236, 10603, 788, 5885, 9126, 6576, 12...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32275</th>\n",
       "      <td>PMC99857__mb2010895009.jpg</td>\n",
       "      <td>The B(a)P-induced apoptotic signaling pathway,...</td>\n",
       "      <td>[9459, 64374, 837, 100506742, 23581, 834, 835,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32276</th>\n",
       "      <td>PMC99889__mb2110211013.jpg</td>\n",
       "      <td>Erk activation following TCR stimulation</td>\n",
       "      <td>[5594, 5595, 2885, 7535, 5604, 27040, 5327, 39...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32277</th>\n",
       "      <td>PMC99957__mb2310813001.jpg</td>\n",
       "      <td>Purine metabolism in S</td>\n",
       "      <td>[23038, 4236, 355, 22978, 5471, 2987, 1015, 83...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32278</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>EGL-15/FGFR signaling pathway</td>\n",
       "      <td>[1432, 5594, 5595, 5599, 5600, 5601, 5602, 560...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32279 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             figid  \\\n",
       "0      PMC100005__mb2410575011.jpg   \n",
       "1      PMC100008__mb2411709009.jpg   \n",
       "2          PMC101242__gkf20707.jpg   \n",
       "3      PMC101600__ac1100359003.jpg   \n",
       "4      PMC102186__mb2400926009.jpg   \n",
       "...                            ...   \n",
       "32274   PMC99537__jb1610128005.jpg   \n",
       "32275   PMC99857__mb2010895009.jpg   \n",
       "32276   PMC99889__mb2110211013.jpg   \n",
       "32277   PMC99957__mb2310813001.jpg   \n",
       "32278   PMC99976__mb2310138007.jpg   \n",
       "\n",
       "                                                figtitle  \\\n",
       "0                   Formation of TrkA–src-aPKC complexes   \n",
       "1      Cooperation between Beta-AR and B2R signaling ...   \n",
       "2          Up-regulation of RAD51 in 8-MOP-treated cells   \n",
       "3      Hypothetical biosynthetic pathway of coumermyc...   \n",
       "4                        Ras-mediated signaling pathways   \n",
       "...                                                  ...   \n",
       "32274                      Basic metabolic pathways in C   \n",
       "32275  The B(a)P-induced apoptotic signaling pathway,...   \n",
       "32276           Erk activation following TCR stimulation   \n",
       "32277                             Purine metabolism in S   \n",
       "32278                      EGL-15/FGFR signaling pathway   \n",
       "\n",
       "                                                   genes  \n",
       "0      [5293, 5294, 25759, 6714, 8503, 4914, 10818, 3...  \n",
       "1      [5331, 5576, 51196, 23236, 5567, 5566, 107, 11...  \n",
       "2      [5883, 5980, 11200, 5893, 10111, 5888, 546, 47...  \n",
       "3      [125965, 84701, 1327, 10063, 1352, 1353, 1351,...  \n",
       "4      [6300, 5602, 6300, 5879, 5058, 5062, 57144, 58...  \n",
       "...                                                  ...  \n",
       "32274  [23038, 4236, 10603, 788, 5885, 9126, 6576, 12...  \n",
       "32275  [9459, 64374, 837, 100506742, 23581, 834, 835,...  \n",
       "32276  [5594, 5595, 2885, 7535, 5604, 27040, 5327, 39...  \n",
       "32277  [23038, 4236, 355, 22978, 5471, 2987, 1015, 83...  \n",
       "32278  [1432, 5594, 5595, 5599, 5600, 5601, 5602, 560...  \n",
       "\n",
       "[32279 rows x 3 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ap_gmt_path = paper_dir.joinpath(\"raw/pfocr_curated_ap.gmt\")\n",
    "ap_gmt_data = list()\n",
    "with open(ap_gmt_path, \"r\") as gmt_f:\n",
    "    for line in gmt_f:\n",
    "        chunks = line.strip().split(\"\\t\")\n",
    "        figid = chunks.pop(0)\n",
    "        figtitle = chunks.pop(0)\n",
    "        genes = chunks\n",
    "        ap_gmt_data.append({\"figid\": figid, \"figtitle\": figtitle, \"genes\": genes})\n",
    "ap_gmt_df = pd.DataFrame.from_records(ap_gmt_data)\n",
    "ap_gmt_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results from AP and from AR are not identical matches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "311\n",
      "309\n"
     ]
    }
   ],
   "source": [
    "weird_ap = set(ap_gmt_df[\"figid\"]) - set(ar_gmt_df[\"figid\"])\n",
    "print(len(weird_ap))\n",
    "weird_ar = set(ar_gmt_df[\"figid\"]) - set(ap_gmt_df[\"figid\"])\n",
    "print(len(weird_ar))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample of results in AP but not in AR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>Effect of NS5A on the Beta-catenin signaling p...</td>\n",
       "      <td>[7008, 2932, 572, 9733, 1499, 2308, 10000, 207...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PMC107855__ii0180505007.jpg</td>\n",
       "      <td>Influence of OdDHL on the immune system</td>\n",
       "      <td>[3592, 3576, 3458, 3664, 3553, 3593, 3439, 712...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>PMC140507__gkg141f1.jpg</td>\n",
       "      <td>The PI 3-kinase pathway by RNAi</td>\n",
       "      <td>[5601, 5599, 207, 5602, 5601, 5163, 5599, 5728...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>PMC1435660__75f8c.jpg</td>\n",
       "      <td>Hypothetical model for HHT</td>\n",
       "      <td>[7043, 94, 7046, 94, 7042, 2022, 51175, 7046, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>PMC1544360__nihms10600f7.jpg</td>\n",
       "      <td>Simplified repression model of angptl3 promote...</td>\n",
       "      <td>[6256, 9612, 6257, 6258, 10062, 9612, 6256, 87...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                figid  \\\n",
       "15   PMC1069556__zjv0080560770006.jpg   \n",
       "19        PMC107855__ii0180505007.jpg   \n",
       "263           PMC140507__gkg141f1.jpg   \n",
       "290             PMC1435660__75f8c.jpg   \n",
       "405      PMC1544360__nihms10600f7.jpg   \n",
       "\n",
       "                                              figtitle  \\\n",
       "15   Effect of NS5A on the Beta-catenin signaling p...   \n",
       "19             Influence of OdDHL on the immune system   \n",
       "263                    The PI 3-kinase pathway by RNAi   \n",
       "290                         Hypothetical model for HHT   \n",
       "405  Simplified repression model of angptl3 promote...   \n",
       "\n",
       "                                                 genes  \n",
       "15   [7008, 2932, 572, 9733, 1499, 2308, 10000, 207...  \n",
       "19   [3592, 3576, 3458, 3664, 3553, 3593, 3439, 712...  \n",
       "263  [5601, 5599, 207, 5602, 5601, 5163, 5599, 5728...  \n",
       "290  [7043, 94, 7046, 94, 7042, 2022, 51175, 7046, ...  \n",
       "405  [6256, 9612, 6257, 6258, 10062, 9612, 6256, 87...  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ap_gmt_df[ap_gmt_df[\"figid\"].isin(weird_ap)].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample of results in AR but not in AP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>genes_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PMC2254938__nihms-38751-f0001.jpg</td>\n",
       "      <td></td>\n",
       "      <td>[207, 208, 3265, 3667, 3845, 4893, 8471, 8660,...</td>\n",
       "      <td>207\\t208\\t3265\\t3667\\t3845\\t4893\\t8471\\t8660\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PMC2398704__nihms-47534-f0003.jpg</td>\n",
       "      <td>Variation the insulin/insulin-like (IIS) pathw...</td>\n",
       "      <td>[207, 948, 3479, 3667, 5163, 5164, 5165, 5166,...</td>\n",
       "      <td>207\\t948\\t3479\\t3667\\t5163\\t5164\\t5165\\t5166\\t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>PMC2529154__nihms62767f4.jpg</td>\n",
       "      <td>Proposed biosynthetic pathway for MDP chromoph...</td>\n",
       "      <td>[712, 713, 714, 715, 716, 717, 718, 727, 729, ...</td>\n",
       "      <td>712\\t713\\t714\\t715\\t716\\t717\\t718\\t727\\t729\\t7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>PMC2585776__nihms70500f3.jpg</td>\n",
       "      <td></td>\n",
       "      <td>[1432, 2353, 5594, 5595, 5599, 5600, 5601, 560...</td>\n",
       "      <td>1432\\t2353\\t5594\\t5595\\t5599\\t5600\\t5601\\t5602...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>PMC2585776__nihms70500f5.jpg</td>\n",
       "      <td></td>\n",
       "      <td>[1432, 2353, 5594, 5595, 5599, 5600, 5601, 560...</td>\n",
       "      <td>1432\\t2353\\t5594\\t5595\\t5599\\t5600\\t5601\\t5602...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 figid  \\\n",
       "151  PMC2254938__nihms-38751-f0001.jpg   \n",
       "194  PMC2398704__nihms-47534-f0003.jpg   \n",
       "241       PMC2529154__nihms62767f4.jpg   \n",
       "300       PMC2585776__nihms70500f3.jpg   \n",
       "301       PMC2585776__nihms70500f5.jpg   \n",
       "\n",
       "                                              figtitle  \\\n",
       "151                                                      \n",
       "194  Variation the insulin/insulin-like (IIS) pathw...   \n",
       "241  Proposed biosynthetic pathway for MDP chromoph...   \n",
       "300                                                      \n",
       "301                                                      \n",
       "\n",
       "                                                 genes  \\\n",
       "151  [207, 208, 3265, 3667, 3845, 4893, 8471, 8660,...   \n",
       "194  [207, 948, 3479, 3667, 5163, 5164, 5165, 5166,...   \n",
       "241  [712, 713, 714, 715, 716, 717, 718, 727, 729, ...   \n",
       "300  [1432, 2353, 5594, 5595, 5599, 5600, 5601, 560...   \n",
       "301  [1432, 2353, 5594, 5595, 5599, 5600, 5601, 560...   \n",
       "\n",
       "                                             genes_str  \n",
       "151  207\\t208\\t3265\\t3667\\t3845\\t4893\\t8471\\t8660\\t...  \n",
       "194  207\\t948\\t3479\\t3667\\t5163\\t5164\\t5165\\t5166\\t...  \n",
       "241  712\\t713\\t714\\t715\\t716\\t717\\t718\\t727\\t729\\t7...  \n",
       "300  1432\\t2353\\t5594\\t5595\\t5599\\t5600\\t5601\\t5602...  \n",
       "301  1432\\t2353\\t5594\\t5595\\t5599\\t5600\\t5601\\t5602...  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ar_gmt_df[ar_gmt_df[\"figid\"].isin(weird_ar)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "385"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ap_gmt_df[ap_gmt_df[\"figid\"].isin(weird_ap)][\"genes\"].apply(len).max()\n",
    "ap_gmt_df[\"genes\"].apply(len).max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results from AP include some cases where entrez ids are duplicated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>unique_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC101242__gkf20707.jpg</td>\n",
       "      <td>Up-regulation of RAD51 in 8-MOP-treated cells</td>\n",
       "      <td>[5883, 5980, 11200, 5893, 10111, 5888, 546, 47...</td>\n",
       "      <td>[2521, 546, 545, 10111, 5980, 5893, 472, 5883,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC102186__mb2400926009.jpg</td>\n",
       "      <td>Ras-mediated signaling pathways</td>\n",
       "      <td>[6300, 5602, 6300, 5879, 5058, 5062, 57144, 58...</td>\n",
       "      <td>[391, 5062, 5605, 3845, 10298, 56924, 3725, 56...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PMC104390__mb0691873009.jpg</td>\n",
       "      <td>Proposed model for translational homeostasis i...</td>\n",
       "      <td>[5293, 8503, 2475, 5291, 5464, 30849, 146850, ...</td>\n",
       "      <td>[5290, 208, 8503, 5464, 23533, 207, 5293, 5291...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PMC1052008__JCI0524178.f2.jpg</td>\n",
       "      <td>Gq/11-activated pathways in maladaptive hypert...</td>\n",
       "      <td>[5331, 5534, 5332, 5579, 23683, 5587, 5533, 55...</td>\n",
       "      <td>[5590, 5330, 5583, 5579, 5532, 5578, 23683, 55...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PMC106275__am0681773006.jpg</td>\n",
       "      <td>Analysis of the region between styR and styA</td>\n",
       "      <td>[55811, 196883, 107, 108, 112, 115, 114, 113, ...</td>\n",
       "      <td>[113, 90527, 196883, 107, 108, 55811, 112, 263...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            figid  \\\n",
       "2         PMC101242__gkf20707.jpg   \n",
       "4     PMC102186__mb2400926009.jpg   \n",
       "5     PMC104390__mb0691873009.jpg   \n",
       "9   PMC1052008__JCI0524178.f2.jpg   \n",
       "12    PMC106275__am0681773006.jpg   \n",
       "\n",
       "                                             figtitle  \\\n",
       "2       Up-regulation of RAD51 in 8-MOP-treated cells   \n",
       "4                     Ras-mediated signaling pathways   \n",
       "5   Proposed model for translational homeostasis i...   \n",
       "9   Gq/11-activated pathways in maladaptive hypert...   \n",
       "12       Analysis of the region between styR and styA   \n",
       "\n",
       "                                                genes  \\\n",
       "2   [5883, 5980, 11200, 5893, 10111, 5888, 546, 47...   \n",
       "4   [6300, 5602, 6300, 5879, 5058, 5062, 57144, 58...   \n",
       "5   [5293, 8503, 2475, 5291, 5464, 30849, 146850, ...   \n",
       "9   [5331, 5534, 5332, 5579, 23683, 5587, 5533, 55...   \n",
       "12  [55811, 196883, 107, 108, 112, 115, 114, 113, ...   \n",
       "\n",
       "                                         unique_genes  \n",
       "2   [2521, 546, 545, 10111, 5980, 5893, 472, 5883,...  \n",
       "4   [391, 5062, 5605, 3845, 10298, 56924, 3725, 56...  \n",
       "5   [5290, 208, 8503, 5464, 23533, 207, 5293, 5291...  \n",
       "9   [5590, 5330, 5583, 5579, 5532, 5578, 23683, 55...  \n",
       "12  [113, 90527, 196883, 107, 108, 55811, 112, 263...  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ap_gmt_df[\"unique_genes\"] = ap_gmt_df[\"genes\"].apply(lambda x: list(set(x)))\n",
    "ap_gmt_df[ap_gmt_df[\"genes\"].apply(len) != ap_gmt_df[\"unique_genes\"].apply(len)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10111',\n",
       " '11200',\n",
       " '2521',\n",
       " '472',\n",
       " '545',\n",
       " '546',\n",
       " '5883',\n",
       " '5884',\n",
       " '5884',\n",
       " '5888',\n",
       " '5893',\n",
       " '5980',\n",
       " '7376']"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(\n",
    "    ap_gmt_df[ap_gmt_df[\"genes\"].apply(len) != ap_gmt_df[\"unique_genes\"].apply(len)][\n",
    "        \"genes\"\n",
    "    ].iat[0]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following should have no rows, because the results from AR shouldn't have any duplicates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>genes_str</th>\n",
       "      <th>unique_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [figid, figtitle, genes, genes_str, unique_genes]\n",
       "Index: []"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ar_gmt_df[\"unique_genes\"] = ar_gmt_df[\"genes\"].apply(lambda x: list(set(x)))\n",
    "ar_gmt_df[ar_gmt_df[\"genes\"].apply(len) != ar_gmt_df[\"unique_genes\"].apply(len)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are `309` hits in AR that are missing from AP, e.g., the figure with figid `PMC4929164__294_2016_565_Fig2_HTML.jpg`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "309"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(weird_ar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PMC4929164__294_2016_565_Fig2_HTML.jpg'"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(weird_ar)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>661217</th>\n",
       "      <td>PMC4929164__294_2016_565_Fig2_HTML.jpg</td>\n",
       "      <td>PMC4929164</td>\n",
       "      <td>(CDK)</td>\n",
       "      <td>CDK</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CDK11A</td>\n",
       "      <td>728642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661218</th>\n",
       "      <td>PMC4929164__294_2016_565_Fig2_HTML.jpg</td>\n",
       "      <td>PMC4929164</td>\n",
       "      <td>(CDK)</td>\n",
       "      <td>CDK</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CDK8</td>\n",
       "      <td>1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661219</th>\n",
       "      <td>PMC4929164__294_2016_565_Fig2_HTML.jpg</td>\n",
       "      <td>PMC4929164</td>\n",
       "      <td>(CDK)</td>\n",
       "      <td>CDK</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CDK7</td>\n",
       "      <td>1022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661220</th>\n",
       "      <td>PMC4929164__294_2016_565_Fig2_HTML.jpg</td>\n",
       "      <td>PMC4929164</td>\n",
       "      <td>(CDK)</td>\n",
       "      <td>CDK</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CDK3</td>\n",
       "      <td>1018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661221</th>\n",
       "      <td>PMC4929164__294_2016_565_Fig2_HTML.jpg</td>\n",
       "      <td>PMC4929164</td>\n",
       "      <td>(CDK)</td>\n",
       "      <td>CDK</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CDKL5</td>\n",
       "      <td>6792</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         figid       pmcid   word symbol  \\\n",
       "661217  PMC4929164__294_2016_565_Fig2_HTML.jpg  PMC4929164  (CDK)    CDK   \n",
       "661218  PMC4929164__294_2016_565_Fig2_HTML.jpg  PMC4929164  (CDK)    CDK   \n",
       "661219  PMC4929164__294_2016_565_Fig2_HTML.jpg  PMC4929164  (CDK)    CDK   \n",
       "661220  PMC4929164__294_2016_565_Fig2_HTML.jpg  PMC4929164  (CDK)    CDK   \n",
       "661221  PMC4929164__294_2016_565_Fig2_HTML.jpg  PMC4929164  (CDK)    CDK   \n",
       "\n",
       "                    source hgnc_symbol  entrez  \n",
       "661217  bioentities_symbol      CDK11A  728642  \n",
       "661218  bioentities_symbol        CDK8    1024  \n",
       "661219  bioentities_symbol        CDK7    1022  \n",
       "661220  bioentities_symbol        CDK3    1018  \n",
       "661221  bioentities_symbol       CDKL5    6792  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_genes_df[pfocr_genes_df[\"figid\"] == list(weird_ar)[0]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results from AP don't even contain anything for the paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>unique_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [figid, figtitle, genes, unique_genes]\n",
       "Index: []"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ap_gmt_df[ap_gmt_df[\"figid\"].str.contains(\"PMC4929164\", regex=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are `311` hits in AP that are missing from AR, e.g., the figure with figid `PMC5752509__oncotarget-08-109135-g007.jpg`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "311"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(weird_ap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PMC5752509__oncotarget-08-109135-g007.jpg'"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(weird_ap)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>847113</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>RAR</td>\n",
       "      <td>RAR</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>RARB</td>\n",
       "      <td>5915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847114</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>RAR</td>\n",
       "      <td>RAR</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>RARG</td>\n",
       "      <td>5916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847115</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>P38</td>\n",
       "      <td>p38</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>MAPK13</td>\n",
       "      <td>5603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847116</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>P38</td>\n",
       "      <td>p38</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>MAPK12</td>\n",
       "      <td>6300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847117</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>P38</td>\n",
       "      <td>p38</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>MAPK14</td>\n",
       "      <td>1432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847118</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>P38</td>\n",
       "      <td>p38</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>MAPK11</td>\n",
       "      <td>5600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847119</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>RARa</td>\n",
       "      <td>RARA</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>RARA</td>\n",
       "      <td>5914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847120</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>Src</td>\n",
       "      <td>SRC</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>SRC</td>\n",
       "      <td>6714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847121</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>RAR</td>\n",
       "      <td>RAR</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>RARA</td>\n",
       "      <td>5914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847122</th>\n",
       "      <td>PMC5752509__oncotarget-08-109135-g007.jpg</td>\n",
       "      <td>PMC5752509</td>\n",
       "      <td>GPR40</td>\n",
       "      <td>GPR40</td>\n",
       "      <td>hgnc_prev_symbol</td>\n",
       "      <td>FFAR1</td>\n",
       "      <td>2864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            figid       pmcid   word symbol  \\\n",
       "847113  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    RAR    RAR   \n",
       "847114  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    RAR    RAR   \n",
       "847115  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    P38    p38   \n",
       "847116  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    P38    p38   \n",
       "847117  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    P38    p38   \n",
       "847118  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    P38    p38   \n",
       "847119  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509   RARa   RARA   \n",
       "847120  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    Src    SRC   \n",
       "847121  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509    RAR    RAR   \n",
       "847122  PMC5752509__oncotarget-08-109135-g007.jpg  PMC5752509  GPR40  GPR40   \n",
       "\n",
       "                    source hgnc_symbol  entrez  \n",
       "847113  bioentities_symbol        RARB    5915  \n",
       "847114  bioentities_symbol        RARG    5916  \n",
       "847115  bioentities_symbol      MAPK13    5603  \n",
       "847116  bioentities_symbol      MAPK12    6300  \n",
       "847117  bioentities_symbol      MAPK14    1432  \n",
       "847118  bioentities_symbol      MAPK11    5600  \n",
       "847119         hgnc_symbol        RARA    5914  \n",
       "847120         hgnc_symbol         SRC    6714  \n",
       "847121  bioentities_symbol        RARA    5914  \n",
       "847122    hgnc_prev_symbol       FFAR1    2864  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(pfocr_genes_df[pfocr_genes_df[\"figid\"] == list(weird_ap)[0]]))\n",
    "pfocr_genes_df[pfocr_genes_df[\"figid\"] == list(weird_ap)[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results from AR don't even contain anything for the paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>genes</th>\n",
       "      <th>genes_str</th>\n",
       "      <th>unique_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [figid, figtitle, genes, genes_str, unique_genes]\n",
       "Index: []"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ar_gmt_df[ar_gmt_df[\"figid\"].str.contains(\"PMC5752509\", regex=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3304\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>PMC1069556</td>\n",
       "      <td>Tef</td>\n",
       "      <td>TEF</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>TEF</td>\n",
       "      <td>7008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>PMC1069556</td>\n",
       "      <td>GSK-3B</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>2932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>PMC1069556</td>\n",
       "      <td>Bad</td>\n",
       "      <td>BAD</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>BAD</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>PMC1069556</td>\n",
       "      <td>p110</td>\n",
       "      <td>p110</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>SART3</td>\n",
       "      <td>9733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>PMC1069556__zjv0080560770006.jpg</td>\n",
       "      <td>PMC1069556</td>\n",
       "      <td>Bcatenin</td>\n",
       "      <td>BCATENIN</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112145</th>\n",
       "      <td>PMC98722__ii1010637004.jpg</td>\n",
       "      <td>PMC98722</td>\n",
       "      <td>actin</td>\n",
       "      <td>Actin</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112146</th>\n",
       "      <td>PMC98722__ii1010637004.jpg</td>\n",
       "      <td>PMC98722</td>\n",
       "      <td>actin</td>\n",
       "      <td>Actin</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>ACTG2</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112147</th>\n",
       "      <td>PMC98722__ii1010637004.jpg</td>\n",
       "      <td>PMC98722</td>\n",
       "      <td>Cdc42</td>\n",
       "      <td>CDC42</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>CDC42</td>\n",
       "      <td>998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112148</th>\n",
       "      <td>PMC98722__ii1010637004.jpg</td>\n",
       "      <td>PMC98722</td>\n",
       "      <td>N-WASP</td>\n",
       "      <td>N-WASP</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>WASL</td>\n",
       "      <td>8976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112149</th>\n",
       "      <td>PMC98722__ii1010637004.jpg</td>\n",
       "      <td>PMC98722</td>\n",
       "      <td>(WASP)</td>\n",
       "      <td>WASP</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>WAS</td>\n",
       "      <td>7454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3304 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    figid       pmcid      word    symbol  \\\n",
       "452      PMC1069556__zjv0080560770006.jpg  PMC1069556       Tef       TEF   \n",
       "453      PMC1069556__zjv0080560770006.jpg  PMC1069556    GSK-3B     GSK3B   \n",
       "454      PMC1069556__zjv0080560770006.jpg  PMC1069556       Bad       BAD   \n",
       "455      PMC1069556__zjv0080560770006.jpg  PMC1069556      p110      p110   \n",
       "456      PMC1069556__zjv0080560770006.jpg  PMC1069556  Bcatenin  BCATENIN   \n",
       "...                                   ...         ...       ...       ...   \n",
       "1112145        PMC98722__ii1010637004.jpg    PMC98722     actin     Actin   \n",
       "1112146        PMC98722__ii1010637004.jpg    PMC98722     actin     Actin   \n",
       "1112147        PMC98722__ii1010637004.jpg    PMC98722     Cdc42     CDC42   \n",
       "1112148        PMC98722__ii1010637004.jpg    PMC98722    N-WASP    N-WASP   \n",
       "1112149        PMC98722__ii1010637004.jpg    PMC98722    (WASP)      WASP   \n",
       "\n",
       "                     source hgnc_symbol  entrez  \n",
       "452             hgnc_symbol         TEF    7008  \n",
       "453             hgnc_symbol       GSK3B    2932  \n",
       "454             hgnc_symbol         BAD     572  \n",
       "455       hgnc_alias_symbol       SART3    9733  \n",
       "456      bioentities_symbol      CTNNB1    1499  \n",
       "...                     ...         ...     ...  \n",
       "1112145  bioentities_symbol       ACTG1      71  \n",
       "1112146  bioentities_symbol       ACTG2      72  \n",
       "1112147         hgnc_symbol       CDC42     998  \n",
       "1112148   hgnc_alias_symbol        WASL    8976  \n",
       "1112149   hgnc_alias_symbol         WAS    7454  \n",
       "\n",
       "[3304 rows x 7 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(pfocr_genes_df[pfocr_genes_df[\"figid\"].isin(weird_ap)]))\n",
    "pfocr_genes_df[pfocr_genes_df[\"figid\"].isin(weird_ap)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>Ga12,Gaq</td>\n",
       "      <td>G-ALPHA-q</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>GNAQ</td>\n",
       "      <td>2776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>Etk</td>\n",
       "      <td>ETK</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>BMX</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>FAK</td>\n",
       "      <td>FAK</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>5747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>AR*</td>\n",
       "      <td>AR</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>AR</td>\n",
       "      <td>367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC100003__mb2410470011.jpg</td>\n",
       "      <td>PMC100003</td>\n",
       "      <td>(Src</td>\n",
       "      <td>SRC</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>SRC</td>\n",
       "      <td>6714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112546</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>MEK-2</td>\n",
       "      <td>MEK2</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>MAP2K2</td>\n",
       "      <td>5605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112547</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>HRAS</td>\n",
       "      <td>3265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112548</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112549</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>RAS</td>\n",
       "      <td>RAS</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>4893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112550</th>\n",
       "      <td>PMC99976__mb2310138007.jpg</td>\n",
       "      <td>PMC99976</td>\n",
       "      <td>SOS-1</td>\n",
       "      <td>SOS1</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>SOS1</td>\n",
       "      <td>6654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1112551 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               figid      pmcid      word     symbol  \\\n",
       "0        PMC100003__mb2410470011.jpg  PMC100003  Ga12,Gaq  G-ALPHA-q   \n",
       "1        PMC100003__mb2410470011.jpg  PMC100003       Etk        ETK   \n",
       "2        PMC100003__mb2410470011.jpg  PMC100003       FAK        FAK   \n",
       "3        PMC100003__mb2410470011.jpg  PMC100003       AR*         AR   \n",
       "4        PMC100003__mb2410470011.jpg  PMC100003      (Src        SRC   \n",
       "...                              ...        ...       ...        ...   \n",
       "1112546   PMC99976__mb2310138007.jpg   PMC99976     MEK-2       MEK2   \n",
       "1112547   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112548   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112549   PMC99976__mb2310138007.jpg   PMC99976       RAS        RAS   \n",
       "1112550   PMC99976__mb2310138007.jpg   PMC99976     SOS-1       SOS1   \n",
       "\n",
       "                     source hgnc_symbol  entrez  \n",
       "0         hgnc_alias_symbol        GNAQ    2776  \n",
       "1         hgnc_alias_symbol         BMX     660  \n",
       "2         hgnc_alias_symbol        PTK2    5747  \n",
       "3               hgnc_symbol          AR     367  \n",
       "4               hgnc_symbol         SRC    6714  \n",
       "...                     ...         ...     ...  \n",
       "1112546   hgnc_alias_symbol      MAP2K2    5605  \n",
       "1112547  bioentities_symbol        HRAS    3265  \n",
       "1112548  bioentities_symbol        KRAS    3845  \n",
       "1112549  bioentities_symbol        NRAS    4893  \n",
       "1112550         hgnc_symbol        SOS1    6654  \n",
       "\n",
       "[1112551 rows x 7 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_df[pfocr_df[\"figid\"].isin(weird_ap)][\"unique_gene_count\"].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_genes_df[pfocr_genes_df[\"figid\"].isin(weird_ap)].groupby(\"figid\")[\n",
    "    \"entrez\"\n",
    "].count().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "figid\n",
       "PMC1069556__zjv0080560770006.jpg              10\n",
       "PMC4800799__zmb9991011760007.jpg              10\n",
       "PMC4815360__pr.115.011833f3.jpg               10\n",
       "PMC4820813__kcll-05-04-1136374-g005.jpg       10\n",
       "PMC4825016__zbc0171641000007.jpg              10\n",
       "                                              ..\n",
       "PMC5735158__41598_2017_16627_Fig6_HTML.jpg    15\n",
       "PMC5915505__441_2018_2801_Fig2_HTML.jpg       15\n",
       "PMC6272071__IJBMS-21-911-g007.jpg             15\n",
       "PMC5045067__hp-2-197Fig1.jpg                  15\n",
       "PMC3151466__nihms-315348-f0002.jpg            17\n",
       "Name: entrez, Length: 311, dtype: int64"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df = (\n",
    "    pfocr_genes_df[pfocr_genes_df[\"figid\"].isin(weird_ap)]\n",
    "    .groupby(\"figid\")[\"entrez\"]\n",
    "    .count()\n",
    ")\n",
    "counts_df.sort_values()\n",
    "# counts_df.max()\n",
    "# counts_df[counts_df[\"entrez\"] == 17]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unique_gene_count    9\n",
       "dtype: int64"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_df[pfocr_df[\"figid\"].isin(weird_ap)][[\"unique_gene_count\"]].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>unique_gene_count</th>\n",
       "      <th>entrez</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [figid, unique_gene_count, entrez]\n",
       "Index: []"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_df[(pfocr_df[\"figid\"].isin(weird_ap)) & (pfocr_df[\"unique_gene_count\"] == 17)][\n",
    "    [\"figid\", \"unique_gene_count\", \"entrez\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>number</th>\n",
       "      <th>reftext</th>\n",
       "      <th>year</th>\n",
       "      <th>pathway_score</th>\n",
       "      <th>pmc_ranked_result_index</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>papertitle</th>\n",
       "      <th>caption</th>\n",
       "      <th>figlink</th>\n",
       "      <th>...</th>\n",
       "      <th>automl_index</th>\n",
       "      <th>figurl</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "      <th>nobe_count</th>\n",
       "      <th>entrez_count</th>\n",
       "      <th>unique_gene_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64942</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>1869</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64943</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>1869</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64944</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>1870</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64945</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>1870</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64946</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F3</td>\n",
       "      <td>E2F3</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F3</td>\n",
       "      <td>1871</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64947</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F3</td>\n",
       "      <td>1871</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64948</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F4</td>\n",
       "      <td>1874</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64949</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F4</td>\n",
       "      <td>E2F4</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F4</td>\n",
       "      <td>1874</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64950</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F5</td>\n",
       "      <td>1875</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64951</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F5</td>\n",
       "      <td>E2F5</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F5</td>\n",
       "      <td>1875</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64952</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>1876</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64953</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>1876</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64954</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>p107</td>\n",
       "      <td>p107</td>\n",
       "      <td>hgnc_alias_symbol</td>\n",
       "      <td>RBL1</td>\n",
       "      <td>5933</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64955</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>79733</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64956</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>79733</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64957</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F7</td>\n",
       "      <td>E2F7</td>\n",
       "      <td>hgnc_symbol</td>\n",
       "      <td>E2F7</td>\n",
       "      <td>144455</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64958</th>\n",
       "      <td>PMC3151466__nihms-315348-f0002.jpg</td>\n",
       "      <td>Fig. (2)</td>\n",
       "      <td>Wei Du, et al. Curr Drug Targets. ;10(7):581-589.</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.910064</td>\n",
       "      <td>190619</td>\n",
       "      <td></td>\n",
       "      <td>The Rb Pathway and Cancer Therapeutics.</td>\n",
       "      <td>In mammals the E2F family is composed of eight...</td>\n",
       "      <td>/pmc/articles/PMC3151466/figure/F2/</td>\n",
       "      <td>...</td>\n",
       "      <td>35244</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...</td>\n",
       "      <td>E2F</td>\n",
       "      <td>E2F</td>\n",
       "      <td>bioentities_symbol</td>\n",
       "      <td>E2F7</td>\n",
       "      <td>144455</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    figid    number  \\\n",
       "64942  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64943  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64944  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64945  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64946  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64947  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64948  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64949  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64950  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64951  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64952  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64953  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64954  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64955  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64956  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64957  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "64958  PMC3151466__nihms-315348-f0002.jpg  Fig. (2)   \n",
       "\n",
       "                                                 reftext year  pathway_score  \\\n",
       "64942  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64943  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64944  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64945  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64946  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64947  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64948  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64949  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64950  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64951  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64952  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64953  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64954  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64955  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64956  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64957  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "64958  Wei Du, et al. Curr Drug Targets. ;10(7):581-589.   NA       0.910064   \n",
       "\n",
       "       pmc_ranked_result_index figtitle  \\\n",
       "64942                   190619            \n",
       "64943                   190619            \n",
       "64944                   190619            \n",
       "64945                   190619            \n",
       "64946                   190619            \n",
       "64947                   190619            \n",
       "64948                   190619            \n",
       "64949                   190619            \n",
       "64950                   190619            \n",
       "64951                   190619            \n",
       "64952                   190619            \n",
       "64953                   190619            \n",
       "64954                   190619            \n",
       "64955                   190619            \n",
       "64956                   190619            \n",
       "64957                   190619            \n",
       "64958                   190619            \n",
       "\n",
       "                                    papertitle  \\\n",
       "64942  The Rb Pathway and Cancer Therapeutics.   \n",
       "64943  The Rb Pathway and Cancer Therapeutics.   \n",
       "64944  The Rb Pathway and Cancer Therapeutics.   \n",
       "64945  The Rb Pathway and Cancer Therapeutics.   \n",
       "64946  The Rb Pathway and Cancer Therapeutics.   \n",
       "64947  The Rb Pathway and Cancer Therapeutics.   \n",
       "64948  The Rb Pathway and Cancer Therapeutics.   \n",
       "64949  The Rb Pathway and Cancer Therapeutics.   \n",
       "64950  The Rb Pathway and Cancer Therapeutics.   \n",
       "64951  The Rb Pathway and Cancer Therapeutics.   \n",
       "64952  The Rb Pathway and Cancer Therapeutics.   \n",
       "64953  The Rb Pathway and Cancer Therapeutics.   \n",
       "64954  The Rb Pathway and Cancer Therapeutics.   \n",
       "64955  The Rb Pathway and Cancer Therapeutics.   \n",
       "64956  The Rb Pathway and Cancer Therapeutics.   \n",
       "64957  The Rb Pathway and Cancer Therapeutics.   \n",
       "64958  The Rb Pathway and Cancer Therapeutics.   \n",
       "\n",
       "                                                 caption  \\\n",
       "64942  In mammals the E2F family is composed of eight...   \n",
       "64943  In mammals the E2F family is composed of eight...   \n",
       "64944  In mammals the E2F family is composed of eight...   \n",
       "64945  In mammals the E2F family is composed of eight...   \n",
       "64946  In mammals the E2F family is composed of eight...   \n",
       "64947  In mammals the E2F family is composed of eight...   \n",
       "64948  In mammals the E2F family is composed of eight...   \n",
       "64949  In mammals the E2F family is composed of eight...   \n",
       "64950  In mammals the E2F family is composed of eight...   \n",
       "64951  In mammals the E2F family is composed of eight...   \n",
       "64952  In mammals the E2F family is composed of eight...   \n",
       "64953  In mammals the E2F family is composed of eight...   \n",
       "64954  In mammals the E2F family is composed of eight...   \n",
       "64955  In mammals the E2F family is composed of eight...   \n",
       "64956  In mammals the E2F family is composed of eight...   \n",
       "64957  In mammals the E2F family is composed of eight...   \n",
       "64958  In mammals the E2F family is composed of eight...   \n",
       "\n",
       "                                   figlink  ... automl_index  \\\n",
       "64942  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64943  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64944  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64945  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64946  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64947  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64948  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64949  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64950  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64951  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64952  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64953  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64954  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64955  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64956  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64957  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "64958  /pmc/articles/PMC3151466/figure/F2/  ...        35244   \n",
       "\n",
       "                                                  figurl  word symbol  \\\n",
       "64942  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F1   E2F1   \n",
       "64943  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64944  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F2   E2F2   \n",
       "64945  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64946  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F3   E2F3   \n",
       "64947  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64948  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64949  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F4   E2F4   \n",
       "64950  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64951  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F5   E2F5   \n",
       "64952  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64953  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F6   E2F6   \n",
       "64954  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  p107   p107   \n",
       "64955  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F8   E2F8   \n",
       "64956  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "64957  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...  E2F7   E2F7   \n",
       "64958  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3...   E2F    E2F   \n",
       "\n",
       "                   source hgnc_symbol  entrez nobe_count entrez_count  \\\n",
       "64942         hgnc_symbol        E2F1    1869         10           17   \n",
       "64943  bioentities_symbol        E2F1    1869         10           17   \n",
       "64944         hgnc_symbol        E2F2    1870         10           17   \n",
       "64945  bioentities_symbol        E2F2    1870         10           17   \n",
       "64946         hgnc_symbol        E2F3    1871         10           17   \n",
       "64947  bioentities_symbol        E2F3    1871         10           17   \n",
       "64948  bioentities_symbol        E2F4    1874         10           17   \n",
       "64949         hgnc_symbol        E2F4    1874         10           17   \n",
       "64950  bioentities_symbol        E2F5    1875         10           17   \n",
       "64951         hgnc_symbol        E2F5    1875         10           17   \n",
       "64952  bioentities_symbol        E2F6    1876         10           17   \n",
       "64953         hgnc_symbol        E2F6    1876         10           17   \n",
       "64954   hgnc_alias_symbol        RBL1    5933         10           17   \n",
       "64955         hgnc_symbol        E2F8   79733         10           17   \n",
       "64956  bioentities_symbol        E2F8   79733         10           17   \n",
       "64957         hgnc_symbol        E2F7  144455         10           17   \n",
       "64958  bioentities_symbol        E2F7  144455         10           17   \n",
       "\n",
       "      unique_gene_count  \n",
       "64942                 9  \n",
       "64943                 9  \n",
       "64944                 9  \n",
       "64945                 9  \n",
       "64946                 9  \n",
       "64947                 9  \n",
       "64948                 9  \n",
       "64949                 9  \n",
       "64950                 9  \n",
       "64951                 9  \n",
       "64952                 9  \n",
       "64953                 9  \n",
       "64954                 9  \n",
       "64955                 9  \n",
       "64956                 9  \n",
       "64957                 9  \n",
       "64958                 9  \n",
       "\n",
       "[17 rows x 24 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_df[pfocr_df[\"figid\"] == \"PMC3151466__nihms-315348-f0002.jpg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_genes_df[pfocr_genes_df[\"figid\"].isin(weird_ap)][\n",
    "    [\"figid\", \"entrez\"]\n",
    "].drop_duplicates().groupby(\"figid\")[\"entrez\"].count().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For GMT files, tabs and percent signs are special characters. Let's check whether we have any that require escaping.\n",
    "\n",
    "No figids or figtitles have tabs, and no figid has a percent sign."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in (\n",
    "    pfocr_df[\n",
    "        pfocr_df[\"figtitle\"].str.contains(\"\\t\")\n",
    "        | pfocr_df[\"figid\"].str.contains(\"\\t\")\n",
    "        | pfocr_df[\"figid\"].str.contains(\"\\%\")\n",
    "    ][[\"figid\", \"figtitle\"]]\n",
    "    .drop_duplicates()\n",
    "    .iterrows()\n",
    "):\n",
    "    figtitle = row[\"figtitle\"]\n",
    "    figid = row[\"figid\"]\n",
    "    print(f\"{figid} '{figtitle}'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some figtitles do have at least one percent sign."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMC2703816__nihms100324f3.jpg\n",
      "Functional annotation of 115 C/EBPŒ≤ target genes presented in pie chart format shows that 74% of target genes are dedicated to signaling, metabolism, and transport, with the remaining 26% having roles in transcription, synaptic transmission, differentiation and proliferation\n",
      "\n",
      "PMC2730981__nihms137404f1.jpg\n",
      "Activation of inflammatory pathway mediated through NF-Œ∫B by life-style related factors such as tobacco, stress, dietary agents, obesity, alcohol, infectious agents, irradiation and environmental stimuli that account for as much as 95% of all cancers\n",
      "\n",
      "PMC2750869__nihms91210f5.jpg\n",
      "Final elaboration of 8, the requisite cyclization precursor, entailed treatment of 2-butanone (21) with (‚àí)-DIPCl and Et3N according to conditions developed by Paterson,, followed by addition of aldehyde (+)-20; cyclization precursor (8) was obtained as an inseparable mixture of diastereomers (5:1) in 86% yield ()\n",
      "\n",
      "PMC3188852__nihms196597f11.jpg\n",
      "In order to construct the quaternary carbon center at C-16, numerous efforts (aldolizations, alkylations, and acylations) were originally carried out, but they were not successful.‚Äì Gratifyingly, it was found that the aldehydic group at C-16 in Na-methylvellosimine could be converted into its related diol after 60 hours in 88% yield via the Tollens reaction by using 37% aqueous formaldehyde (5 equivalents) and KOH (10 equivalents) in methanol\n",
      "\n",
      "PMC3188852__nihms196597f6.jpg\n",
      "The key intermediate, (+)-vellosimine (13) was prepared from the common intermediate, E-ethylidene ketone 10, which was synthesized via the asymmetric Pictect-Spengler reaction, Dieckmann cyclization, and palladium mediated cross coupling in high overall yield.‚Äì As illustrated in , the route began with the readily available (‚àí)-Na-H tetracyclic ketone 19, which was subjected to the conditions of catalytic hydrogenolysis to remove the benzyl function in 92% yield\n",
      "\n",
      "PMC3887475__nihms515312f5.jpg\n",
      "When we conducted further studies to test whether the oxidation pathway could be generalized, we found that cyclopentenones with exocyclic double bonds could be obtained in good yields from several other substrates with electron-withdrawing aromatic substituents (62‚Äì92% yield; )\n",
      "\n",
      "PMC5195866__nihms818513f1.jpg\n",
      "Role of CYP2A6 in the metabolic pathway of nicotine.Up to 80% of a nicotine dose inhaled from cigarette smoke undergoes metabolic inactivation to cotinine\n",
      "\n",
      "PMC152346__008946f5.jpg\n",
      "Simplified Scheme of Predicted Fluxes of C16 and C18 Fatty Acids in Membrane Leaf Glycerolipids of Arabidopsis Wild-Type, fatb-ko, and fatb-ko act1 Mutant Plants.The numbers in the figure represent average values of mol % units of C16 and C18 fatty acids accumulated in membrane glycerolipids\n",
      "\n",
      "PMC523366__pp1363043f05.jpg\n",
      "Metabolic flux map of primary and intermediate metabolic pathways in developing soybean cotyledons cultured on Suc (10% [w/w] U-13C) and Gln\n",
      "\n",
      "PMC1383630__tpc1800523f01.jpg\n",
      "Schematic Representation of Arabidopsis Nonhost Resistance to Bgh.After germination of spores (Sp), formation of appressoria (Ap), and breaching of the host cell wall (CW) by the fungal penetration peg (Pp), ‚àº90% of penetrations are stopped in cell wall extensions called papillae (Pa)\n",
      "\n",
      "PMC1393175__zam0030665160003.jpg\n",
      "Transcriptional response of frc and oxc to pH 5.5, pH 4.5, and 1% (70 mM) ammonium oxalate (pH 6.8) in MRS broth after 30 min\n",
      "\n",
      "PMC2785970__pp1511769f03.jpg\n",
      "Iodine staining of starch content together with quantitative measurements (in mg g‚àí1 fresh weight, and as a % of Columbia-0 in parentheses) from four replicate plants, harvested at the end of the day\n",
      "\n",
      "PMC2813229__1477-5956-7-48-4.jpg\n",
      "Pathway analysis result showing 18 differentially expressed proteins following 50% PH can be sorted into the c-myc pathway\n",
      "\n",
      "PMC2932889__kdj-34-211-g002.jpg\n",
      "A fraction (2-5%) of the glucose entering a cell is directed into the hexosamine biosynthesis pathway (HBP) pathway\n",
      "\n",
      "PMC3109408__PD2011-716859.002.jpg\n",
      "Simplified diagram of Kynurenine pathway: during neuroinflammation, 95% of the dietary tryptophan is metabolized along the KP within the brain\n",
      "\n",
      "PMC3715096__anie0052-5305-f7.jpg\n",
      "Mechanistic rationale for the formation of (¬±)-1 and (¬±)-14: 15 was obtained by the oxidative addition of Pd0 to (¬±)-2; pathway A: C‚ÄìH activation pathway by highly electrophilic Ar‚ÄìPd species; pathway B: Generation of a syn Œ≤-hydride (Hb) by a Pd-enolate; Pathway C: Formation of (¬±)-iso-linoxepin 14 by a syn Œ≤-hydride elimination of Ha; conditions: a) PdCl2 (20 mol %), PPh3 (44 mol %) NEt3 (10 equiv), microwave radiation, DMF, 130 ¬∞C, 7 h, 74 %; b) PdCl2 (20 mol %), PPh3 (44 mol %), CsOAc (10.0 equiv), DMF, 75 ¬∞C, 4 h, 78 %\n",
      "\n",
      "PMC3842785__1475-2859-12-109-6.jpg\n",
      "100% 2-13C glycerol ILE shows that glycine and serine are predominantly synthesized from glyoxylate rather than 3-phosphoglycerate\n",
      "\n",
      "PMC4417967__toxins-07-01273-g002.jpg\n",
      "Relative adenine and thymine (%AT) DNA content of ergot alkaloid synthesis (EAS) loci\n",
      "\n",
      "PMC4638047__zjw0111551700007.jpg\n",
      "Pathway analysis highlights APP is a central hub interacting with nearly 40% of candidate molecules in EGS\n",
      "\n",
      "PMC5223135__nutd201649f4.jpg\n",
      "Proposed mechanisms of kefir peptides-mediated modulation of the lipid metabolism pathway in 30% fructose-induced NAFLD mice\n",
      "\n",
      "PMC5505034__13068_2017_868_Fig2_HTML.jpg\n",
      "Genes transcript abundance of reconstructed carbon fixation pathway in Chlorella PY-ZU1 cells cultivated with 15% CO2 gas versus air\n",
      "\n",
      "PMC6404759__nihms-997888-f0001.jpg\n",
      "Major metabolic pathway (97%) in cancer initiation by estrogens\n",
      "\n",
      "PMC5863803__12864_2018_4609_Fig5_HTML.jpg\n",
      "Schematic representation of the expression of genes of glycolysis, pentose phosphate pathway (PPP) and pentose catabolic pathway (PCP) in Aspergillus nidulans after 2¬†h of transfer to 1% (w/v) wheat bran\n",
      "\n",
      "PMC5918149__13568_2018_594_Fig6_HTML.jpg\n",
      "Contribution made by different microbes in consortium CY-1 when 50% phenanthrene was removed\n",
      "\n",
      "PMC6016767__pnas.1721487115fig02.jpg\n",
      "Three TFs‚ÄîCRF4, SNZ, and CDF1‚Äîregulate 53% of the N-uptake/assimilation pathway genes\n",
      "\n",
      "PMC6134795__12916_2018_1136_Fig1_HTML.jpg\n",
      "Composite pathway using the KEGG pathway analysis, a composite pathway was generated based on the canonical mapping of the phosphorylated proteins independently associated with increased % collagen (red) in the liver\n",
      "\n",
      "PMC6471220__ijms-20-01472-g003.jpg\n",
      "Transcriptional effects in zebrafish liver after feeding with 25% of maize substitution\n",
      "\n",
      "PMC6630544__pharmaceuticals-12-00066-g001.jpg\n",
      "Stages of Mtb transmission and infection cycle showing that 95% of infected patients will have the latent form, and 50% of relapse cases will go directly from latent to active form (only 5% for new cases) []\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i, row in (\n",
    "    pfocr_df[pfocr_df[\"figtitle\"].str.contains(\"\\%\")][[\"figid\", \"figtitle\"]]\n",
    "    .drop_duplicates()\n",
    "    .iterrows()\n",
    "):\n",
    "    figtitle = row[\"figtitle\"]\n",
    "    figid = row[\"figid\"]\n",
    "    print(figid)\n",
    "    print(figtitle)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>number</th>\n",
       "      <th>reftext</th>\n",
       "      <th>year</th>\n",
       "      <th>pathway_score</th>\n",
       "      <th>pmc_ranked_result_index</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>papertitle</th>\n",
       "      <th>caption</th>\n",
       "      <th>figlink</th>\n",
       "      <th>...</th>\n",
       "      <th>automl_index</th>\n",
       "      <th>figurl</th>\n",
       "      <th>word</th>\n",
       "      <th>symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>entrez</th>\n",
       "      <th>nobe_count</th>\n",
       "      <th>entrez_count</th>\n",
       "      <th>unique_gene_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [figid, number, reftext, year, pathway_score, pmc_ranked_result_index, figtitle, papertitle, caption, figlink, pmcid, filename, source_f, type.man, automl_index, figurl, word, symbol, source, hgnc_symbol, entrez, nobe_count, entrez_count, unique_gene_count]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfocr_df[pfocr_df[\"figid\"].str.contains(\"\\t\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>figid</th>\n",
       "      <th>number</th>\n",
       "      <th>reftext</th>\n",
       "      <th>year</th>\n",
       "      <th>pathway_score</th>\n",
       "      <th>pmc_ranked_result_index</th>\n",
       "      <th>figtitle</th>\n",
       "      <th>papertitle</th>\n",
       "      <th>caption</th>\n",
       "      <th>figlink</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>filename</th>\n",
       "      <th>source_f</th>\n",
       "      <th>type.man</th>\n",
       "      <th>automl_index</th>\n",
       "      <th>figurl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61692</th>\n",
       "      <td>PMC2703816__nihms100324f3.jpg</td>\n",
       "      <td>Figure 3</td>\n",
       "      <td>N. Kfoury, et al. Mol Cell Neurosci. ;40(3):31...</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.708703</td>\n",
       "      <td>222034</td>\n",
       "      <td>Functional annotation of 115 C/EBPŒ≤ target ge...</td>\n",
       "      <td>Identification of neuronal target genes for CC...</td>\n",
       "      <td>Functional annotation of 115 C/EBPŒ≤ target ge...</td>\n",
       "      <td>/pmc/articles/PMC2703816/figure/F3/</td>\n",
       "      <td>PMC2703816</td>\n",
       "      <td>nihms100324f3.jpg</td>\n",
       "      <td>../data/images/PMC2703816__nihms100324f3.jpg</td>\n",
       "      <td>NA</td>\n",
       "      <td>58133</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               figid    number  \\\n",
       "61692  PMC2703816__nihms100324f3.jpg  Figure 3   \n",
       "\n",
       "                                                 reftext year  pathway_score  \\\n",
       "61692  N. Kfoury, et al. Mol Cell Neurosci. ;40(3):31...   NA       0.708703   \n",
       "\n",
       "       pmc_ranked_result_index  \\\n",
       "61692                   222034   \n",
       "\n",
       "                                                figtitle  \\\n",
       "61692  Functional annotation of 115 C/EBPŒ≤ target ge...   \n",
       "\n",
       "                                              papertitle  \\\n",
       "61692  Identification of neuronal target genes for CC...   \n",
       "\n",
       "                                                 caption  \\\n",
       "61692  Functional annotation of 115 C/EBPŒ≤ target ge...   \n",
       "\n",
       "                                   figlink       pmcid           filename  \\\n",
       "61692  /pmc/articles/PMC2703816/figure/F3/  PMC2703816  nihms100324f3.jpg   \n",
       "\n",
       "                                           source_f type.man  automl_index  \\\n",
       "61692  ../data/images/PMC2703816__nihms100324f3.jpg       NA         58133   \n",
       "\n",
       "                                                  figurl  \n",
       "61692  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A figid that rpy2 may be importing incorrectly. Check the encoding of figtitles.\n",
    "pfocr_figures_df[pfocr_figures_df[\"figid\"] == \"PMC2703816__nihms100324f3.jpg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import rpy2\n",
    "\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "pandas2ri.activate()\n",
    "base = importr(\"base\")\n",
    "readRDS = robjects.r[\"readRDS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rdf2pandas(df):\n",
    "    df_names = base.names(df)\n",
    "    data = dict()\n",
    "    for i in range(len(df_names)):\n",
    "        column = df[i]\n",
    "        if hasattr(column, \"iter_labels\"):\n",
    "            data[df_names[i]] = [x for x in column.iter_labels()]\n",
    "        else:\n",
    "            data[df_names[i]] = [x for x in column]\n",
    "\n",
    "    return pd.DataFrame(data=data)\n",
    "\n",
    "\n",
    "def rds2pandas(f):\n",
    "    return rdf2pandas(readRDS(str(f)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <span>R/rpy2 DataFrame (6 x 16)</span>\n",
       "    <table>\n",
       "      <thead>\n",
       "        <tr>\n",
       "        \n",
       "          <th>figid</th>\n",
       "        \n",
       "          <th>pmcid</th>\n",
       "        \n",
       "          <th>filename</th>\n",
       "        \n",
       "          <th>...</th>\n",
       "        \n",
       "          <th>number</th>\n",
       "        \n",
       "          <th>caption</th>\n",
       "        \n",
       "          <th>organism</th>\n",
       "        \n",
       "        </tr>\n",
       "      </thead>\n",
       "      <tbody>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC5...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC5...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        '4159...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Figu...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Mode...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'zh20...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Fig....\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Prop...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC5...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC5...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'rsob...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Figu...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Aβ p...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'pone...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Figu...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Gluc...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC2...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC2...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'nihm...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Figu...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Bact...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'PMC4...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'emss...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        ...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Figu...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Sche...\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'Homo...\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      </tbody>\n",
       "    </table>\n",
       "    "
      ],
      "text/plain": [
       "R object with classes: ('data.frame',) mapped to:\n",
       "<DataFrame - Python:0x75f3d5d70eb0 / R:0x1eb60358>\n",
       "[StrVector, StrVector, StrVector, StrVector, ..., StrVector, StrVector, StrVector, StrVector]\n",
       "  figid: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5edfbe0 / R:0x151ad988>\n",
       "['PMC5..., 'PMC4..., 'PMC5..., 'PMC4..., 'PMC2..., 'PMC4...]\n",
       "  pmcid: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5fd9f50 / R:0x151ad9f8>\n",
       "['PMC5..., 'PMC4..., 'PMC5..., 'PMC4..., 'PMC2..., 'PMC4...]\n",
       "  filename: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5dd28c0 / R:0x151ada68>\n",
       "['4159..., 'zh20..., 'rsob..., 'pone..., 'nihm..., 'emss...]\n",
       "  year: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5d83be0 / R:0x151adad8>\n",
       "['2017', '2014', '2017', '2014', NA_character_, '2015']\n",
       "...\n",
       "  pmc_ranked_result_index: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5dc9a50 / R:0x151add08>\n",
       "['Mode..., 'Stim..., 'AB p..., 'Gluc..., 'NeuA..., 'Sche...]\n",
       "  figlink: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5fb3640 / R:0x151add78>\n",
       "['Figu..., 'Fig...., 'Figu..., 'Figu..., 'Figu..., 'Figu...]\n",
       "  source_f: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5f3e500 / R:0x151adde8>\n",
       "['Mode..., 'Prop..., 'Aβ p..., 'Gluc..., 'Bact..., 'Sche...]\n",
       "  type.man: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3d5d908c0 / R:0x151ade58>\n",
       "['Homo..., 'Homo..., 'Homo..., 'Homo..., 'Homo..., 'Homo...]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmm_pfocr_figures_df = readRDS(\n",
    "    \"~/Dropbox (Gladstone)/Documents/PFOCR_25Years/pfocr_figures.rds\"\n",
    ")\n",
    "hmm_pfocr_figures_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <span>R/rpy2 DataFrame (3 x 2)</span>\n",
       "    <table>\n",
       "      <thead>\n",
       "        <tr>\n",
       "        \n",
       "          <th>int_values</th>\n",
       "        \n",
       "          <th>str_values</th>\n",
       "        \n",
       "        </tr>\n",
       "      </thead>\n",
       "      <tbody>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        1\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'abc'\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        2\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'def'\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      <tr>\n",
       "      \n",
       "      <td>\n",
       "        3\n",
       "      </td>\n",
       "      \n",
       "      <td>\n",
       "        'ghi'\n",
       "      </td>\n",
       "      \n",
       "      </tr>\n",
       "      \n",
       "      </tbody>\n",
       "    </table>\n",
       "    "
      ],
      "text/plain": [
       "R object with classes: ('data.frame',) mapped to:\n",
       "<DataFrame - Python:0x75f3c8ddcb40 / R:0x36f91d18>\n",
       "[IntVector, StrVector]\n",
       "  int_values: <class 'rpy2.robjects.vectors.IntVector'>\n",
       "  R object with classes: ('integer',) mapped to:\n",
       "<IntVector - Python:0x75f3c8ade320 / R:0x36f91258>\n",
       "[1, 2, 3]\n",
       "  str_values: <class 'rpy2.robjects.vectors.StrVector'>\n",
       "  R object with classes: ('character',) mapped to:\n",
       "<StrVector - Python:0x75f3c8ade3c0 / R:0x1ed36ff8>\n",
       "['abc', 'def', 'ghi']"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_df = ro.DataFrame(\n",
    "    {\n",
    "        \"int_values\": ro.IntVector([1, 2, 3]),\n",
    "        \"str_values\": ro.StrVector([\"abc\", \"def\", \"ghi\"]),\n",
    "    }\n",
    ")\n",
    "\n",
    "r_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Demo from [here](https://rpy2.github.io/doc/v3.4.x/html/generated_rst/pandas.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'rpy2.robjects.conversion' has no attribute 'ri2py_dataframe'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-192-6d2b643804af>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mlocalconverter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault_converter\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mpandas2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconverter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mpd_from_r_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mpd_from_r_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'rpy2.robjects.conversion' has no attribute 'ri2py_dataframe'"
     ]
    }
   ],
   "source": [
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    pd_from_r_df = ro.conversion.ri2py_dataframe(r_df)\n",
    "\n",
    "pd_from_r_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Buffer for this type not yet supported.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-195-dd482ca90ca7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mlocalconverter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdefault_converter\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mpandas2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconverter\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mcv\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mpd_from_r_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpandas2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0mpd_from_r_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/gpnm7i19lpj8p43mjrdw03d0hjalmskl-python3-3.7.5/lib/python3.7/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    838\u001b[0m                             '1 positional argument')\n\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mri2py_dataframe\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPandasDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_items\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPandasDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_items\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/gpnm7i19lpj8p43mjrdw03d0hjalmskl-python3-3.7.5/lib/python3.7/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    838\u001b[0m                             '1 positional argument')\n\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mri2py_vector\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    137\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSexpVector\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_vector\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 139\u001b[0;31m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumpy2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    140\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/gpnm7i19lpj8p43mjrdw03d0hjalmskl-python3-3.7.5/lib/python3.7/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    838\u001b[0m                             '1 positional argument')\n\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/numpy2ri.py\u001b[0m in \u001b[0;36mri2py_sexp\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_sexp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    158\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtypeof\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_vectortypes\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtypeof\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mVECSXP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 159\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    160\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault_converter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/numpy/core/_asarray.py\u001b[0m in \u001b[0;36masarray\u001b[0;34m(a, dtype, order)\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \"\"\"\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morder\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Buffer for this type not yet supported."
     ]
    }
   ],
   "source": [
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects import default_converter\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "with localconverter(default_converter + pandas2ri.converter) as cv:\n",
    "    pd_from_r_df = pandas2ri.ri2py(r_df)\n",
    "pd_from_r_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Buffer for this type not yet supported.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-193-714aeae71077>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpandas2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mri2py_dataframe\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPandasDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_items\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPandasDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_items\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/gpnm7i19lpj8p43mjrdw03d0hjalmskl-python3-3.7.5/lib/python3.7/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    838\u001b[0m                             '1 positional argument')\n\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py\u001b[0m in \u001b[0;36mri2py_vector\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    137\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mregister\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSexpVector\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_vector\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 139\u001b[0;31m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumpy2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    140\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/gpnm7i19lpj8p43mjrdw03d0hjalmskl-python3-3.7.5/lib/python3.7/functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    838\u001b[0m                             '1 positional argument')\n\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0mfuncname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__name__'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'singledispatch function'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/robjects/numpy2ri.py\u001b[0m in \u001b[0;36mri2py_sexp\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mri2py_sexp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    158\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtypeof\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_vectortypes\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtypeof\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mVECSXP\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 159\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    160\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault_converter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mri2py\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/numpy/core/_asarray.py\u001b[0m in \u001b[0;36masarray\u001b[0;34m(a, dtype, order)\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \"\"\"\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morder\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Buffer for this type not yet supported."
     ]
    }
   ],
   "source": [
    "pandas2ri.ri2py_dataframe(r_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'rpy2.situation' from '/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/situation.py'>"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpy2.situation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<function rpy2.robjects.conversion._ri2py(obj)>"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ro.conversion.ri2py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>int_values</th>\n",
       "      <th>str_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>abc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>def</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ghi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   int_values str_values\n",
       "0           1        abc\n",
       "1           2        def\n",
       "2           3        ghi"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd_df = pd.DataFrame({\"int_values\": [1, 2, 3], \"str_values\": [\"abc\", \"def\", \"ghi\"]})\n",
    "\n",
    "pd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'rpy2.robjects.conversion' has no attribute 'py2rpy'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-199-ba05bb0f6f26>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mlocalconverter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefault_converter\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mpandas2ri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconverter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mr_from_pd_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpy2rpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpd_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mr_from_pd_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'rpy2.robjects.conversion' has no attribute 'py2rpy'"
     ]
    }
   ],
   "source": [
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    r_from_pd_df = ro.conversion.py2rpy(pd_df)\n",
    "\n",
    "r_from_pd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'simplegeneric'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m     \u001b[0;32mfrom\u001b[0m \u001b[0mIPython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexternal\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimplegeneric\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mgeneric\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    107\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'IPython.external.simplegeneric'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-200-39bcd566a575>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mfunctools\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpartial\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mrpy2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mipython\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mhtml\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mhtml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhtml_rdataframe\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpartial\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhtml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhtml_rdataframe\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtable_class\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"docutils\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/ipython/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mrmagic\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mload_ipython_extension\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/nix/store/hm475d8nhddsi2cvipkhl8k06cmxn7kh-python3-3.7.5-env/lib/python3.7/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    107\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m     \u001b[0;31m# IPython 4.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 109\u001b[0;31m     \u001b[0;32mfrom\u001b[0m \u001b[0msimplegeneric\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mgeneric\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    110\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    111\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'simplegeneric'"
     ]
    }
   ],
   "source": [
    "from functools import partial\n",
    "from rpy2.ipython import html\n",
    "html.html_rdataframe=partial(html.html_rdataframe, table_class=\"docutils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# /run/user/1000/tmp_yz5_q44/pfocr20200224.gmt"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "encoding": "# -*- coding: utf-8 -*-"
  },
  "kernelspec": {
   "display_name": "Python3 - mypkgs",
   "language": "python",
   "name": "ipython_mypkgs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
